{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Cancer Prediction from Scratch: the good, the bad and the ugly "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The notebook is made to demonstrate the next skills:\n",
    "\n",
    "- Perform proper data preprocessing.\n",
    "- Descriptive Analytics\n",
    "- Predictive Analytics\n",
    "- Perform natural language processing.\n",
    "- Use map and reduce operations.\n",
    "- Use streaming data.\n",
    "- Translate data into a graph and extract information from a network.\n",
    "- Apply machine learning\n",
    "- My ability to detect and handle imabalanced data.\n",
    "- Data Reduction using the PCA algorithm (The Principal Component Analysis)\n",
    "- Data Visualization with Matplotlib\n",
    "\n",
    "*From my point of view the application of graph theory is very handy to improve logistics in companies. Specially if we are talking about analyzing information from point A to point B."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Import Libraries\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\g1238\\AppData\\Local\\conda\\conda\\envs\\DL4Denv\\lib\\site-packages\\nltk\\twitter\\__init__.py:20: UserWarning: The twython library has not been installed. Some functionality from the twitter package will not be available.\n",
      "  warnings.warn(\"The twython library has not been installed. \"\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import sklearn as skl\n",
    "import nltk\n",
    "from nltk import word_tokenize\n",
    "from nltk import WordNetLemmatizer\n",
    "from nltk.corpus import stopwords\n",
    "from nltk.corpus import wordnet\n",
    "from nltk.sentiment import SentimentIntensityAnalyzer\n",
    "import nltk.sentiment.util\n",
    "from nltk.tag import pos_tag\n",
    "import networkx as nx\n",
    "import itertools\n",
    "import random\n",
    "import imblearn\n",
    "import category_encoders as ce\n",
    "from imblearn.over_sampling import SMOTE\n",
    "from imblearn.combine import SMOTEENN, SMOTETomek\n",
    "from sklearn import preprocessing\n",
    "from sklearn.preprocessing import LabelEncoder\n",
    "from imblearn.under_sampling import RandomUnderSampler\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn import svm\n",
    "from sklearn.metrics import confusion_matrix\n",
    "from sklearn.decomposition import PCA\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from sklearn.preprocessing import MinMaxScaler\n",
    "#from pyspark.sql import SparkSession\n",
    "from sklearn.decomposition import TruncatedSVD\n",
    "import seaborn as sb\n",
    "import matplotlib.pyplot as plt\n",
    "from clustergrammer_widget import *\n",
    "import ipywidgets as widgets\n",
    "import matplotlib.pyplot as plt\n",
    "from sklearn.model_selection import KFold \n",
    "import re"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    " # Cancer Gene Detection\n",
    " \n",
    " To start coding. I am going to discuss a little bit about the idea behind the scence. I took this dataset available online which has the interactions within human proteins. I am interested in finding out those proteins which are more related to cancer given certain characteristics of the interaction network, to do this I need to detect whether the description containing the word cancer relates positively or negatively to the protein description, in order to classify them as cancer/no cancer(sentiment analysis). Then I am going to do natural language processing to add certain caracteristics given the protein description as an attribute to the node. Afterwards, I am going to proceed to perform some machine learning algorithm to find out those proteins interactions which tend to be, more exposed to turn into cancer given the characteristics of my network."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Dataset\n",
    "\n",
    "I took the dataset from the next link https://string-db.org/cgi/download?sessionId=%24input-%3E%7BsessionId%7D&species_text=Homo+sapiens, which takes into account only those proteins from Humans. I do have a list of links with the proteins and their interactions, together with their ID's and descriptions, I will try to find out the cancer labels by doing sentiment analysis over the descriptions. \n",
    "   a) First dataset contains the proteins with its descriptions.\n",
    "   b) Second dataset contains the protein interactions. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "#First Clean the file with the information of each protein.\n",
    "proteinsInfoOriginal = open(\"9606.protein.info.v11.0.txt\", \"r\").readlines()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Second dataset with protein interactions\n",
    "proteinNetwork = pd.read_csv('9606.protein.links.detailed.v11.0.txt', sep=' ', header='infer')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### NATURAL LANGUAGE PROCESSING: ASSINGING THE LABELS \n",
    "\n",
    "I am going to use sentiment analysis over the descriptions to try to detect those descriptions, which I consider to be cancer in negative way. Because, the dataset does not have any label. I want to be able to classify my descriptions in malign cancer.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Creating dataset into dataframe.\n",
    "stopwords = set(nltk.corpus.stopwords.words('english'))\n",
    "\n",
    "def edit_line(line):\n",
    "    return str(line).split(\"\\t\")\n",
    "def clean_punc(text):\n",
    "    return re.sub('[^A-Za-z0-9]+',' ',text.lower())\n",
    "def clean_num(text):\n",
    "    return re.sub(r'\\d+','',text)\n",
    "def tokenization(txt):\n",
    "    return word_tokenize(str(txt).lower())\n",
    "def stopwords_txt(text):\n",
    "    return [w for w in text if not w in list(stopwords)]\n",
    "def lemmatization(txt):\n",
    "    return [WordNetLemmatizer().lemmatize(t) for t in txt]\n",
    "def contains_cancer(txt):\n",
    "    if 'cancer' in txt:\n",
    "        label = 1\n",
    "    else: \n",
    "        label = 0\n",
    "    return label\n",
    "# The ugliness in the next three definitions can be explain as the computer does not support to retrieve all at one time.\n",
    "def get_desc_sentiment_neg(desc):\n",
    "    sentiment = SentimentIntensityAnalyzer()\n",
    "    cancer_sent = sentiment.polarity_scores(desc)\n",
    "    return cancer_sent['neg']\n",
    "def get_desc_sentiment_neu(desc):\n",
    "    sentiment = SentimentIntensityAnalyzer()\n",
    "    cancer_sent = sentiment.polarity_scores(desc)\n",
    "    return cancer_sent['neu']\n",
    "def get_desc_sentiment_pos(desc):\n",
    "    sentiment = SentimentIntensityAnalyzer()\n",
    "    cancer_sent = sentiment.polarity_scores(desc)\n",
    "    return cancer_sent['pos']\n",
    "\n",
    "data = list(map(edit_line,proteinsInfoOriginal))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_df = pd.DataFrame(data)\n",
    "proteinDescription = pd.DataFrame(data_df.iloc[1:])\n",
    "proteinDescription.columns = data_df.iloc[0]\n",
    "proteinDescription = proteinDescription.reset_index()\n",
    "del proteinDescription['index']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Start NLP pipeline."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Clean punctuation from text\n",
    "proteinDescription['annotation\\n'] = list(map(clean_punc,proteinDescription['annotation\\n']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Clean numbers from text\n",
    "proteinDescription['annotation\\n'] = list(map(clean_num,proteinDescription['annotation\\n']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Tokenize text\n",
    "proteinDescription['nlp_processing'] = list(map(tokenization,proteinDescription['annotation\\n']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Erase stop words\n",
    "proteinDescription['nlp_processing'] = list(map(stopwords_txt,proteinDescription['nlp_processing']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Apply lemmatization\n",
    "proteinDescription['nlp_processing'] = list(map(lemmatization,proteinDescription['nlp_processing']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get cancer noun\n",
    "proteinDescription['cancer_noun_identification'] = list(map(contains_cancer,proteinDescription['nlp_processing']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# According to the findings only 199 descriptions are related to cancer.\n",
    "cancer_df = proteinDescription[proteinDescription['cancer_noun_identification'] == 1].reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "sentiment = SentimentIntensityAnalyzer()\n",
    "cancer_sent = sentiment.polarity_scores(cancer_df[\"annotation\\n\"][178])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "cancer_df['sent_negative'] = list(map(get_desc_sentiment_neg,cancer_df[\"annotation\\n\"]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "cancer_df['sent_positive'] = list(map(get_desc_sentiment_pos,cancer_df[\"annotation\\n\"]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "cancer_df['sent_neutral'] = list(map(get_desc_sentiment_neu,cancer_df[\"annotation\\n\"]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>protein_external_id</th>\n",
       "      <th>preferred_name</th>\n",
       "      <th>protein_size</th>\n",
       "      <th>annotation</th>\n",
       "      <th>nlp_processing</th>\n",
       "      <th>cancer_noun_identification</th>\n",
       "      <th>sent_negative</th>\n",
       "      <th>sent_positive</th>\n",
       "      <th>sent_neutral</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>102</td>\n",
       "      <td>9606.ENSP00000064778</td>\n",
       "      <td>FAM168A</td>\n",
       "      <td>244</td>\n",
       "      <td>protein fama in cancer context protects cells ...</td>\n",
       "      <td>[protein, fama, cancer, context, protects, cel...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.261</td>\n",
       "      <td>0.050</td>\n",
       "      <td>0.690</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>156</td>\n",
       "      <td>9606.ENSP00000167106</td>\n",
       "      <td>VASH1</td>\n",
       "      <td>365</td>\n",
       "      <td>vasohibin  angiogenesis inhibitor inhibits mig...</td>\n",
       "      <td>[vasohibin, angiogenesis, inhibitor, inhibits,...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.221</td>\n",
       "      <td>0.059</td>\n",
       "      <td>0.720</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>241</td>\n",
       "      <td>9606.ENSP00000201031</td>\n",
       "      <td>TFAP2C</td>\n",
       "      <td>450</td>\n",
       "      <td>transcription factor ap  gamma sequence specif...</td>\n",
       "      <td>[transcription, factor, ap, gamma, sequence, s...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.052</td>\n",
       "      <td>0.037</td>\n",
       "      <td>0.911</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>441</td>\n",
       "      <td>9606.ENSP00000216807</td>\n",
       "      <td>BRMS1L</td>\n",
       "      <td>323</td>\n",
       "      <td>breast cancer metastasis suppressor  like prot...</td>\n",
       "      <td>[breast, cancer, metastasis, suppressor, like,...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.314</td>\n",
       "      <td>0.120</td>\n",
       "      <td>0.566</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>514</td>\n",
       "      <td>9606.ENSP00000219022</td>\n",
       "      <td>OLFM4</td>\n",
       "      <td>510</td>\n",
       "      <td>olfactomedin  may promote proliferation of pan...</td>\n",
       "      <td>[olfactomedin, may, promote, proliferation, pa...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.135</td>\n",
       "      <td>0.160</td>\n",
       "      <td>0.706</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>589</td>\n",
       "      <td>9606.ENSP00000220812</td>\n",
       "      <td>DKK4</td>\n",
       "      <td>224</td>\n",
       "      <td>dickkopf related protein  antagonizes canonica...</td>\n",
       "      <td>[dickkopf, related, protein, antagonizes, cano...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.120</td>\n",
       "      <td>0.080</td>\n",
       "      <td>0.800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>815</td>\n",
       "      <td>9606.ENSP00000226230</td>\n",
       "      <td>TMEM97</td>\n",
       "      <td>176</td>\n",
       "      <td>sigma intracellular receptor  intracellular or...</td>\n",
       "      <td>[sigma, intracellular, receptor, intracellular...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.057</td>\n",
       "      <td>0.086</td>\n",
       "      <td>0.857</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>891</td>\n",
       "      <td>9606.ENSP00000228865</td>\n",
       "      <td>CREBL2</td>\n",
       "      <td>120</td>\n",
       "      <td>camp responsive element binding protein like  ...</td>\n",
       "      <td>[camp, responsive, element, binding, protein, ...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.045</td>\n",
       "      <td>0.128</td>\n",
       "      <td>0.828</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>999</td>\n",
       "      <td>9606.ENSP00000231656</td>\n",
       "      <td>CDX1</td>\n",
       "      <td>265</td>\n",
       "      <td>homeobox protein cdx  plays a role in transcri...</td>\n",
       "      <td>[homeobox, protein, cdx, play, role, transcrip...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.162</td>\n",
       "      <td>0.081</td>\n",
       "      <td>0.756</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1065</td>\n",
       "      <td>9606.ENSP00000233948</td>\n",
       "      <td>WNT6</td>\n",
       "      <td>365</td>\n",
       "      <td>protein wnt  ligand for members of the frizzle...</td>\n",
       "      <td>[protein, wnt, ligand, member, frizzled, famil...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.110</td>\n",
       "      <td>0.037</td>\n",
       "      <td>0.853</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>1121</td>\n",
       "      <td>9606.ENSP00000236671</td>\n",
       "      <td>CTSD</td>\n",
       "      <td>412</td>\n",
       "      <td>cathepsin d acid protease active in intracellu...</td>\n",
       "      <td>[cathepsin, acid, protease, active, intracellu...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.161</td>\n",
       "      <td>0.097</td>\n",
       "      <td>0.742</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>1145</td>\n",
       "      <td>9606.ENSP00000237612</td>\n",
       "      <td>ABCG2</td>\n",
       "      <td>655</td>\n",
       "      <td>atp binding cassette sub family g member  high...</td>\n",
       "      <td>[atp, binding, cassette, sub, family, g, membe...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.066</td>\n",
       "      <td>0.086</td>\n",
       "      <td>0.848</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>1178</td>\n",
       "      <td>9606.ENSP00000238823</td>\n",
       "      <td>FAM98A</td>\n",
       "      <td>518</td>\n",
       "      <td>protein fama positively stimulates prmt induce...</td>\n",
       "      <td>[protein, fama, positively, stimulates, prmt, ...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.092</td>\n",
       "      <td>0.235</td>\n",
       "      <td>0.672</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>1204</td>\n",
       "      <td>9606.ENSP00000239462</td>\n",
       "      <td>TNN</td>\n",
       "      <td>1299</td>\n",
       "      <td>tenascin n extracellular matrix protein that s...</td>\n",
       "      <td>[tenascin, n, extracellular, matrix, protein, ...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.074</td>\n",
       "      <td>0.058</td>\n",
       "      <td>0.869</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>1372</td>\n",
       "      <td>9606.ENSP00000244926</td>\n",
       "      <td>SCGB1D2</td>\n",
       "      <td>90</td>\n",
       "      <td>secretoglobin family d member  may bind androg...</td>\n",
       "      <td>[secretoglobin, family, member, may, bind, and...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.140</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.860</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>1373</td>\n",
       "      <td>9606.ENSP00000244930</td>\n",
       "      <td>SCGB2A1</td>\n",
       "      <td>95</td>\n",
       "      <td>mammaglobin b may bind androgens and other ste...</td>\n",
       "      <td>[mammaglobin, b, may, bind, androgen, steroid,...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.124</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.876</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>1471</td>\n",
       "      <td>9606.ENSP00000247191</td>\n",
       "      <td>DLGAP5</td>\n",
       "      <td>846</td>\n",
       "      <td>disks large associated protein  potential cell...</td>\n",
       "      <td>[disk, large, associated, protein, potential, ...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.084</td>\n",
       "      <td>0.095</td>\n",
       "      <td>0.821</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>1478</td>\n",
       "      <td>9606.ENSP00000247306</td>\n",
       "      <td>CTAG2</td>\n",
       "      <td>210</td>\n",
       "      <td>cancer testis antigen</td>\n",
       "      <td>[cancer, testis, antigen]</td>\n",
       "      <td>1</td>\n",
       "      <td>0.688</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.312</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>1646</td>\n",
       "      <td>9606.ENSP00000251691</td>\n",
       "      <td>KIAA1244</td>\n",
       "      <td>2177</td>\n",
       "      <td>brefeldin a inhibited guanine nucleotide excha...</td>\n",
       "      <td>[brefeldin, inhibited, guanine, nucleotide, ex...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.119</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.881</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>1899</td>\n",
       "      <td>9606.ENSP00000255681</td>\n",
       "      <td>MACROD1</td>\n",
       "      <td>325</td>\n",
       "      <td>o acetyl adp ribose deacetylase macrod removes...</td>\n",
       "      <td>[acetyl, adp, ribose, deacetylase, macrod, rem...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.072</td>\n",
       "      <td>0.928</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>1912</td>\n",
       "      <td>9606.ENSP00000256078</td>\n",
       "      <td>KRAS</td>\n",
       "      <td>189</td>\n",
       "      <td>gtpase kras ras proteins bind gdp gtp and poss...</td>\n",
       "      <td>[gtpase, kras, ra, protein, bind, gdp, gtp, po...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.133</td>\n",
       "      <td>0.130</td>\n",
       "      <td>0.737</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>1925</td>\n",
       "      <td>9606.ENSP00000256261</td>\n",
       "      <td>DUSP26</td>\n",
       "      <td>211</td>\n",
       "      <td>dual specificity protein phosphatase  inactiva...</td>\n",
       "      <td>[dual, specificity, protein, phosphatase, inac...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.153</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.847</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>2105</td>\n",
       "      <td>9606.ENSP00000258729</td>\n",
       "      <td>IGF2BP3</td>\n",
       "      <td>579</td>\n",
       "      <td>insulin like growth factor  mrna binding prote...</td>\n",
       "      <td>[insulin, like, growth, factor, mrna, binding,...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.109</td>\n",
       "      <td>0.076</td>\n",
       "      <td>0.815</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>2210</td>\n",
       "      <td>9606.ENSP00000260187</td>\n",
       "      <td>USP2</td>\n",
       "      <td>605</td>\n",
       "      <td>ubiquitin carboxyl terminal hydrolase  hydrola...</td>\n",
       "      <td>[ubiquitin, carboxyl, terminal, hydrolase, hyd...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.157</td>\n",
       "      <td>0.067</td>\n",
       "      <td>0.776</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>2434</td>\n",
       "      <td>9606.ENSP00000261893</td>\n",
       "      <td>LACTB</td>\n",
       "      <td>547</td>\n",
       "      <td>serine beta lactamase like protein lactb mitoc...</td>\n",
       "      <td>[serine, beta, lactamase, like, protein, lactb...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.131</td>\n",
       "      <td>0.054</td>\n",
       "      <td>0.814</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>2461</td>\n",
       "      <td>9606.ENSP00000262102</td>\n",
       "      <td>MTUS1</td>\n",
       "      <td>1270</td>\n",
       "      <td>microtubule associated tumor suppressor  coope...</td>\n",
       "      <td>[microtubule, associated, tumor, suppressor, c...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.215</td>\n",
       "      <td>0.040</td>\n",
       "      <td>0.746</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>2510</td>\n",
       "      <td>9606.ENSP00000262305</td>\n",
       "      <td>RAB11FIP3</td>\n",
       "      <td>756</td>\n",
       "      <td>rab family interacting protein  acts as a regu...</td>\n",
       "      <td>[rab, family, interacting, protein, act, regul...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.049</td>\n",
       "      <td>0.055</td>\n",
       "      <td>0.896</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>2647</td>\n",
       "      <td>9606.ENSP00000263083</td>\n",
       "      <td>DPH1</td>\n",
       "      <td>443</td>\n",
       "      <td>amino  carboxypropyl histidine synthase subu...</td>\n",
       "      <td>[amino, carboxypropyl, histidine, synthase, su...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.092</td>\n",
       "      <td>0.124</td>\n",
       "      <td>0.784</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>2913</td>\n",
       "      <td>9606.ENSP00000264606</td>\n",
       "      <td>HDAC4</td>\n",
       "      <td>1084</td>\n",
       "      <td>histone deacetylase  responsible for the deace...</td>\n",
       "      <td>[histone, deacetylase, responsible, deacetylat...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.077</td>\n",
       "      <td>0.093</td>\n",
       "      <td>0.831</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>2999</td>\n",
       "      <td>9606.ENSP00000265052</td>\n",
       "      <td>MGLL</td>\n",
       "      <td>313</td>\n",
       "      <td>monoglyceride lipase converts monoacylglycerid...</td>\n",
       "      <td>[monoglyceride, lipase, convert, monoacylglyce...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.161</td>\n",
       "      <td>0.133</td>\n",
       "      <td>0.705</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>169</th>\n",
       "      <td>16150</td>\n",
       "      <td>9606.ENSP00000413520</td>\n",
       "      <td>HID1</td>\n",
       "      <td>788</td>\n",
       "      <td>protein hid may play an important role in the ...</td>\n",
       "      <td>[protein, hid, may, play, important, role, dev...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.071</td>\n",
       "      <td>0.214</td>\n",
       "      <td>0.714</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>170</th>\n",
       "      <td>16180</td>\n",
       "      <td>9606.ENSP00000414253</td>\n",
       "      <td>CT47A2</td>\n",
       "      <td>288</td>\n",
       "      <td>cancer testis antigen family  member a</td>\n",
       "      <td>[cancer, testis, antigen, family, member]</td>\n",
       "      <td>1</td>\n",
       "      <td>0.524</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.476</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>171</th>\n",
       "      <td>16193</td>\n",
       "      <td>9606.ENSP00000414509</td>\n",
       "      <td>CT47A12</td>\n",
       "      <td>288</td>\n",
       "      <td>cancer testis antigen family  member a</td>\n",
       "      <td>[cancer, testis, antigen, family, member]</td>\n",
       "      <td>1</td>\n",
       "      <td>0.524</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.476</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>172</th>\n",
       "      <td>16455</td>\n",
       "      <td>9606.ENSP00000418960</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>1884</td>\n",
       "      <td>breast cancer type  susceptibility protein e u...</td>\n",
       "      <td>[breast, cancer, type, susceptibility, protein...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.167</td>\n",
       "      <td>0.022</td>\n",
       "      <td>0.812</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>173</th>\n",
       "      <td>16480</td>\n",
       "      <td>9606.ENSP00000419235</td>\n",
       "      <td>FAM115A</td>\n",
       "      <td>921</td>\n",
       "      <td>trpm channel associated factor  positively reg...</td>\n",
       "      <td>[trpm, channel, associated, factor, positively...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.165</td>\n",
       "      <td>0.139</td>\n",
       "      <td>0.695</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>174</th>\n",
       "      <td>16542</td>\n",
       "      <td>9606.ENSP00000420095</td>\n",
       "      <td>ZMYND8</td>\n",
       "      <td>1234</td>\n",
       "      <td>protein kinase c binding protein  may act as a...</td>\n",
       "      <td>[protein, kinase, c, binding, protein, may, ac...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.095</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.905</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175</th>\n",
       "      <td>16781</td>\n",
       "      <td>9606.ENSP00000425493</td>\n",
       "      <td>CAGE1</td>\n",
       "      <td>839</td>\n",
       "      <td>cancer associated gene  protein cancer antigen</td>\n",
       "      <td>[cancer, associated, gene, protein, cancer, an...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.688</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.312</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>176</th>\n",
       "      <td>16839</td>\n",
       "      <td>9606.ENSP00000427015</td>\n",
       "      <td>RPS6KA2</td>\n",
       "      <td>741</td>\n",
       "      <td>ribosomal protein s kinase alpha  serine threo...</td>\n",
       "      <td>[ribosomal, protein, kinase, alpha, serine, th...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.159</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.841</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>177</th>\n",
       "      <td>16851</td>\n",
       "      <td>9606.ENSP00000427342</td>\n",
       "      <td>CT45A5</td>\n",
       "      <td>189</td>\n",
       "      <td>cancer testis antigen family  member a belongs...</td>\n",
       "      <td>[cancer, testis, antigen, family, member, belo...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.328</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.672</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>178</th>\n",
       "      <td>17080</td>\n",
       "      <td>9606.ENSP00000432353</td>\n",
       "      <td>PBOV1</td>\n",
       "      <td>135</td>\n",
       "      <td>prostate and breast cancer overexpressed</td>\n",
       "      <td>[prostate, breast, cancer, overexpressed]</td>\n",
       "      <td>1</td>\n",
       "      <td>0.524</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.476</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>179</th>\n",
       "      <td>17154</td>\n",
       "      <td>9606.ENSP00000434045</td>\n",
       "      <td>GPR68</td>\n",
       "      <td>365</td>\n",
       "      <td>ovarian cancer g protein coupled receptor  pro...</td>\n",
       "      <td>[ovarian, cancer, g, protein, coupled, recepto...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.111</td>\n",
       "      <td>0.019</td>\n",
       "      <td>0.870</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>180</th>\n",
       "      <td>17433</td>\n",
       "      <td>9606.ENSP00000441140</td>\n",
       "      <td>NOL8</td>\n",
       "      <td>1167</td>\n",
       "      <td>nucleolar protein  plays an essential role in ...</td>\n",
       "      <td>[nucleolar, protein, play, essential, role, su...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.164</td>\n",
       "      <td>0.037</td>\n",
       "      <td>0.799</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>181</th>\n",
       "      <td>17495</td>\n",
       "      <td>9606.ENSP00000442464</td>\n",
       "      <td>NCEH1</td>\n",
       "      <td>448</td>\n",
       "      <td>neutral cholesterol ester hydrolase  hydrolyze...</td>\n",
       "      <td>[neutral, cholesterol, ester, hydrolase, hydro...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.106</td>\n",
       "      <td>0.073</td>\n",
       "      <td>0.821</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>182</th>\n",
       "      <td>17499</td>\n",
       "      <td>9606.ENSP00000442521</td>\n",
       "      <td>BEX2</td>\n",
       "      <td>160</td>\n",
       "      <td>protein bex regulator of mitochondrial apoptos...</td>\n",
       "      <td>[protein, bex, regulator, mitochondrial, apopt...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.207</td>\n",
       "      <td>0.054</td>\n",
       "      <td>0.739</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>183</th>\n",
       "      <td>17574</td>\n",
       "      <td>9606.ENSP00000444412</td>\n",
       "      <td>ZNF268</td>\n",
       "      <td>947</td>\n",
       "      <td>zinc finger protein  isoform  acts as a transc...</td>\n",
       "      <td>[zinc, finger, protein, isoform, act, transcri...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.274</td>\n",
       "      <td>0.048</td>\n",
       "      <td>0.678</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>184</th>\n",
       "      <td>17886</td>\n",
       "      <td>9606.ENSP00000453502</td>\n",
       "      <td>ST20</td>\n",
       "      <td>79</td>\n",
       "      <td>suppressor of tumorigenicity  protein may act ...</td>\n",
       "      <td>[suppressor, tumorigenicity, protein, may, act...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.343</td>\n",
       "      <td>0.118</td>\n",
       "      <td>0.539</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>185</th>\n",
       "      <td>17960</td>\n",
       "      <td>9606.ENSP00000455442</td>\n",
       "      <td>NUPR1L</td>\n",
       "      <td>97</td>\n",
       "      <td>nuclear protein  acts as a transcriptional rep...</td>\n",
       "      <td>[nuclear, protein, act, transcriptional, repre...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.254</td>\n",
       "      <td>0.036</td>\n",
       "      <td>0.710</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>186</th>\n",
       "      <td>18144</td>\n",
       "      <td>9606.ENSP00000463027</td>\n",
       "      <td>NCOA4</td>\n",
       "      <td>650</td>\n",
       "      <td>nuclear receptor coactivator  enhances the and...</td>\n",
       "      <td>[nuclear, receptor, coactivator, enhances, and...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.161</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.839</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>187</th>\n",
       "      <td>18188</td>\n",
       "      <td>9606.ENSP00000464162</td>\n",
       "      <td>GREB1L</td>\n",
       "      <td>1923</td>\n",
       "      <td>greb like protein growth regulation by estroge...</td>\n",
       "      <td>[greb, like, protein, growth, regulation, estr...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.183</td>\n",
       "      <td>0.317</td>\n",
       "      <td>0.500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>188</th>\n",
       "      <td>18371</td>\n",
       "      <td>9606.ENSP00000469441</td>\n",
       "      <td>CTAG1A</td>\n",
       "      <td>180</td>\n",
       "      <td>cancer testis antigen a belongs to the ctag pc...</td>\n",
       "      <td>[cancer, testis, antigen, belongs, ctag, pcc, ...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.355</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.645</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>189</th>\n",
       "      <td>18403</td>\n",
       "      <td>9606.ENSP00000470185</td>\n",
       "      <td>CT45A1</td>\n",
       "      <td>189</td>\n",
       "      <td>cancer testis antigen family  member a belongs...</td>\n",
       "      <td>[cancer, testis, antigen, family, member, belo...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.328</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.672</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>190</th>\n",
       "      <td>18446</td>\n",
       "      <td>9606.ENSP00000471418</td>\n",
       "      <td>ENSG00000269096</td>\n",
       "      <td>189</td>\n",
       "      <td>cancer testis antigen family  member a belongs...</td>\n",
       "      <td>[cancer, testis, antigen, family, member, belo...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.328</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.672</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191</th>\n",
       "      <td>18533</td>\n",
       "      <td>9606.ENSP00000473553</td>\n",
       "      <td>CSNK2A3</td>\n",
       "      <td>391</td>\n",
       "      <td>casein kinase ii subunit alpha  probable catal...</td>\n",
       "      <td>[casein, kinase, ii, subunit, alpha, probable,...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.096</td>\n",
       "      <td>0.121</td>\n",
       "      <td>0.783</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>192</th>\n",
       "      <td>18565</td>\n",
       "      <td>9606.ENSP00000474632</td>\n",
       "      <td>ENSG00000271449</td>\n",
       "      <td>189</td>\n",
       "      <td>cancer testis antigen family  member a</td>\n",
       "      <td>[cancer, testis, antigen, family, member]</td>\n",
       "      <td>1</td>\n",
       "      <td>0.524</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.476</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>193</th>\n",
       "      <td>18861</td>\n",
       "      <td>9606.ENSP00000479608</td>\n",
       "      <td>ENSG00000278289</td>\n",
       "      <td>189</td>\n",
       "      <td>cancer testis antigen family  member a</td>\n",
       "      <td>[cancer, testis, antigen, family, member]</td>\n",
       "      <td>1</td>\n",
       "      <td>0.524</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.476</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>194</th>\n",
       "      <td>19027</td>\n",
       "      <td>9606.ENSP00000481184</td>\n",
       "      <td>CT45A4</td>\n",
       "      <td>189</td>\n",
       "      <td>cancer testis antigen family  member a</td>\n",
       "      <td>[cancer, testis, antigen, family, member]</td>\n",
       "      <td>1</td>\n",
       "      <td>0.524</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.476</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>195</th>\n",
       "      <td>19252</td>\n",
       "      <td>9606.ENSP00000483637</td>\n",
       "      <td>CT45-2</td>\n",
       "      <td>189</td>\n",
       "      <td>cancer testis antigen family  member a</td>\n",
       "      <td>[cancer, testis, antigen, family, member]</td>\n",
       "      <td>1</td>\n",
       "      <td>0.524</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.476</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>196</th>\n",
       "      <td>19289</td>\n",
       "      <td>9606.ENSP00000484135</td>\n",
       "      <td>COLCA2</td>\n",
       "      <td>251</td>\n",
       "      <td>colorectal cancer associated</td>\n",
       "      <td>[colorectal, cancer, associated]</td>\n",
       "      <td>1</td>\n",
       "      <td>0.688</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.312</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>197</th>\n",
       "      <td>19308</td>\n",
       "      <td>9606.ENSP00000484443</td>\n",
       "      <td>NEK3</td>\n",
       "      <td>506</td>\n",
       "      <td>serine threonine protein kinase nek protein ki...</td>\n",
       "      <td>[serine, threonine, protein, kinase, nek, prot...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.062</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.938</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>198</th>\n",
       "      <td>19322</td>\n",
       "      <td>9606.ENSP00000484601</td>\n",
       "      <td>CT45A10</td>\n",
       "      <td>189</td>\n",
       "      <td>cancer testis antigen family  member a</td>\n",
       "      <td>[cancer, testis, antigen, family, member]</td>\n",
       "      <td>1</td>\n",
       "      <td>0.524</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.476</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>199 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "0    index   protein_external_id   preferred_name protein_size  \\\n",
       "0      102  9606.ENSP00000064778          FAM168A          244   \n",
       "1      156  9606.ENSP00000167106            VASH1          365   \n",
       "2      241  9606.ENSP00000201031           TFAP2C          450   \n",
       "3      441  9606.ENSP00000216807           BRMS1L          323   \n",
       "4      514  9606.ENSP00000219022            OLFM4          510   \n",
       "5      589  9606.ENSP00000220812             DKK4          224   \n",
       "6      815  9606.ENSP00000226230           TMEM97          176   \n",
       "7      891  9606.ENSP00000228865           CREBL2          120   \n",
       "8      999  9606.ENSP00000231656             CDX1          265   \n",
       "9     1065  9606.ENSP00000233948             WNT6          365   \n",
       "10    1121  9606.ENSP00000236671             CTSD          412   \n",
       "11    1145  9606.ENSP00000237612            ABCG2          655   \n",
       "12    1178  9606.ENSP00000238823           FAM98A          518   \n",
       "13    1204  9606.ENSP00000239462              TNN         1299   \n",
       "14    1372  9606.ENSP00000244926          SCGB1D2           90   \n",
       "15    1373  9606.ENSP00000244930          SCGB2A1           95   \n",
       "16    1471  9606.ENSP00000247191           DLGAP5          846   \n",
       "17    1478  9606.ENSP00000247306            CTAG2          210   \n",
       "18    1646  9606.ENSP00000251691         KIAA1244         2177   \n",
       "19    1899  9606.ENSP00000255681          MACROD1          325   \n",
       "20    1912  9606.ENSP00000256078             KRAS          189   \n",
       "21    1925  9606.ENSP00000256261           DUSP26          211   \n",
       "22    2105  9606.ENSP00000258729          IGF2BP3          579   \n",
       "23    2210  9606.ENSP00000260187             USP2          605   \n",
       "24    2434  9606.ENSP00000261893            LACTB          547   \n",
       "25    2461  9606.ENSP00000262102            MTUS1         1270   \n",
       "26    2510  9606.ENSP00000262305        RAB11FIP3          756   \n",
       "27    2647  9606.ENSP00000263083             DPH1          443   \n",
       "28    2913  9606.ENSP00000264606            HDAC4         1084   \n",
       "29    2999  9606.ENSP00000265052             MGLL          313   \n",
       "..     ...                   ...              ...          ...   \n",
       "169  16150  9606.ENSP00000413520             HID1          788   \n",
       "170  16180  9606.ENSP00000414253           CT47A2          288   \n",
       "171  16193  9606.ENSP00000414509          CT47A12          288   \n",
       "172  16455  9606.ENSP00000418960            BRCA1         1884   \n",
       "173  16480  9606.ENSP00000419235          FAM115A          921   \n",
       "174  16542  9606.ENSP00000420095           ZMYND8         1234   \n",
       "175  16781  9606.ENSP00000425493            CAGE1          839   \n",
       "176  16839  9606.ENSP00000427015          RPS6KA2          741   \n",
       "177  16851  9606.ENSP00000427342           CT45A5          189   \n",
       "178  17080  9606.ENSP00000432353            PBOV1          135   \n",
       "179  17154  9606.ENSP00000434045            GPR68          365   \n",
       "180  17433  9606.ENSP00000441140             NOL8         1167   \n",
       "181  17495  9606.ENSP00000442464            NCEH1          448   \n",
       "182  17499  9606.ENSP00000442521             BEX2          160   \n",
       "183  17574  9606.ENSP00000444412           ZNF268          947   \n",
       "184  17886  9606.ENSP00000453502             ST20           79   \n",
       "185  17960  9606.ENSP00000455442           NUPR1L           97   \n",
       "186  18144  9606.ENSP00000463027            NCOA4          650   \n",
       "187  18188  9606.ENSP00000464162           GREB1L         1923   \n",
       "188  18371  9606.ENSP00000469441           CTAG1A          180   \n",
       "189  18403  9606.ENSP00000470185           CT45A1          189   \n",
       "190  18446  9606.ENSP00000471418  ENSG00000269096          189   \n",
       "191  18533  9606.ENSP00000473553          CSNK2A3          391   \n",
       "192  18565  9606.ENSP00000474632  ENSG00000271449          189   \n",
       "193  18861  9606.ENSP00000479608  ENSG00000278289          189   \n",
       "194  19027  9606.ENSP00000481184           CT45A4          189   \n",
       "195  19252  9606.ENSP00000483637           CT45-2          189   \n",
       "196  19289  9606.ENSP00000484135           COLCA2          251   \n",
       "197  19308  9606.ENSP00000484443             NEK3          506   \n",
       "198  19322  9606.ENSP00000484601          CT45A10          189   \n",
       "\n",
       "0                                         annotation\\n  \\\n",
       "0    protein fama in cancer context protects cells ...   \n",
       "1    vasohibin  angiogenesis inhibitor inhibits mig...   \n",
       "2    transcription factor ap  gamma sequence specif...   \n",
       "3    breast cancer metastasis suppressor  like prot...   \n",
       "4    olfactomedin  may promote proliferation of pan...   \n",
       "5    dickkopf related protein  antagonizes canonica...   \n",
       "6    sigma intracellular receptor  intracellular or...   \n",
       "7    camp responsive element binding protein like  ...   \n",
       "8    homeobox protein cdx  plays a role in transcri...   \n",
       "9    protein wnt  ligand for members of the frizzle...   \n",
       "10   cathepsin d acid protease active in intracellu...   \n",
       "11   atp binding cassette sub family g member  high...   \n",
       "12   protein fama positively stimulates prmt induce...   \n",
       "13   tenascin n extracellular matrix protein that s...   \n",
       "14   secretoglobin family d member  may bind androg...   \n",
       "15   mammaglobin b may bind androgens and other ste...   \n",
       "16   disks large associated protein  potential cell...   \n",
       "17                             cancer testis antigen     \n",
       "18   brefeldin a inhibited guanine nucleotide excha...   \n",
       "19   o acetyl adp ribose deacetylase macrod removes...   \n",
       "20   gtpase kras ras proteins bind gdp gtp and poss...   \n",
       "21   dual specificity protein phosphatase  inactiva...   \n",
       "22   insulin like growth factor  mrna binding prote...   \n",
       "23   ubiquitin carboxyl terminal hydrolase  hydrola...   \n",
       "24   serine beta lactamase like protein lactb mitoc...   \n",
       "25   microtubule associated tumor suppressor  coope...   \n",
       "26   rab family interacting protein  acts as a regu...   \n",
       "27     amino  carboxypropyl histidine synthase subu...   \n",
       "28   histone deacetylase  responsible for the deace...   \n",
       "29   monoglyceride lipase converts monoacylglycerid...   \n",
       "..                                                 ...   \n",
       "169  protein hid may play an important role in the ...   \n",
       "170            cancer testis antigen family  member a    \n",
       "171            cancer testis antigen family  member a    \n",
       "172  breast cancer type  susceptibility protein e u...   \n",
       "173  trpm channel associated factor  positively reg...   \n",
       "174  protein kinase c binding protein  may act as a...   \n",
       "175   cancer associated gene  protein cancer antigen     \n",
       "176  ribosomal protein s kinase alpha  serine threo...   \n",
       "177  cancer testis antigen family  member a belongs...   \n",
       "178         prostate and breast cancer overexpressed     \n",
       "179  ovarian cancer g protein coupled receptor  pro...   \n",
       "180  nucleolar protein  plays an essential role in ...   \n",
       "181  neutral cholesterol ester hydrolase  hydrolyze...   \n",
       "182  protein bex regulator of mitochondrial apoptos...   \n",
       "183  zinc finger protein  isoform  acts as a transc...   \n",
       "184  suppressor of tumorigenicity  protein may act ...   \n",
       "185  nuclear protein  acts as a transcriptional rep...   \n",
       "186  nuclear receptor coactivator  enhances the and...   \n",
       "187  greb like protein growth regulation by estroge...   \n",
       "188  cancer testis antigen a belongs to the ctag pc...   \n",
       "189  cancer testis antigen family  member a belongs...   \n",
       "190  cancer testis antigen family  member a belongs...   \n",
       "191  casein kinase ii subunit alpha  probable catal...   \n",
       "192            cancer testis antigen family  member a    \n",
       "193            cancer testis antigen family  member a    \n",
       "194            cancer testis antigen family  member a    \n",
       "195            cancer testis antigen family  member a    \n",
       "196                     colorectal cancer associated     \n",
       "197  serine threonine protein kinase nek protein ki...   \n",
       "198            cancer testis antigen family  member a    \n",
       "\n",
       "0                                       nlp_processing  \\\n",
       "0    [protein, fama, cancer, context, protects, cel...   \n",
       "1    [vasohibin, angiogenesis, inhibitor, inhibits,...   \n",
       "2    [transcription, factor, ap, gamma, sequence, s...   \n",
       "3    [breast, cancer, metastasis, suppressor, like,...   \n",
       "4    [olfactomedin, may, promote, proliferation, pa...   \n",
       "5    [dickkopf, related, protein, antagonizes, cano...   \n",
       "6    [sigma, intracellular, receptor, intracellular...   \n",
       "7    [camp, responsive, element, binding, protein, ...   \n",
       "8    [homeobox, protein, cdx, play, role, transcrip...   \n",
       "9    [protein, wnt, ligand, member, frizzled, famil...   \n",
       "10   [cathepsin, acid, protease, active, intracellu...   \n",
       "11   [atp, binding, cassette, sub, family, g, membe...   \n",
       "12   [protein, fama, positively, stimulates, prmt, ...   \n",
       "13   [tenascin, n, extracellular, matrix, protein, ...   \n",
       "14   [secretoglobin, family, member, may, bind, and...   \n",
       "15   [mammaglobin, b, may, bind, androgen, steroid,...   \n",
       "16   [disk, large, associated, protein, potential, ...   \n",
       "17                           [cancer, testis, antigen]   \n",
       "18   [brefeldin, inhibited, guanine, nucleotide, ex...   \n",
       "19   [acetyl, adp, ribose, deacetylase, macrod, rem...   \n",
       "20   [gtpase, kras, ra, protein, bind, gdp, gtp, po...   \n",
       "21   [dual, specificity, protein, phosphatase, inac...   \n",
       "22   [insulin, like, growth, factor, mrna, binding,...   \n",
       "23   [ubiquitin, carboxyl, terminal, hydrolase, hyd...   \n",
       "24   [serine, beta, lactamase, like, protein, lactb...   \n",
       "25   [microtubule, associated, tumor, suppressor, c...   \n",
       "26   [rab, family, interacting, protein, act, regul...   \n",
       "27   [amino, carboxypropyl, histidine, synthase, su...   \n",
       "28   [histone, deacetylase, responsible, deacetylat...   \n",
       "29   [monoglyceride, lipase, convert, monoacylglyce...   \n",
       "..                                                 ...   \n",
       "169  [protein, hid, may, play, important, role, dev...   \n",
       "170          [cancer, testis, antigen, family, member]   \n",
       "171          [cancer, testis, antigen, family, member]   \n",
       "172  [breast, cancer, type, susceptibility, protein...   \n",
       "173  [trpm, channel, associated, factor, positively...   \n",
       "174  [protein, kinase, c, binding, protein, may, ac...   \n",
       "175  [cancer, associated, gene, protein, cancer, an...   \n",
       "176  [ribosomal, protein, kinase, alpha, serine, th...   \n",
       "177  [cancer, testis, antigen, family, member, belo...   \n",
       "178          [prostate, breast, cancer, overexpressed]   \n",
       "179  [ovarian, cancer, g, protein, coupled, recepto...   \n",
       "180  [nucleolar, protein, play, essential, role, su...   \n",
       "181  [neutral, cholesterol, ester, hydrolase, hydro...   \n",
       "182  [protein, bex, regulator, mitochondrial, apopt...   \n",
       "183  [zinc, finger, protein, isoform, act, transcri...   \n",
       "184  [suppressor, tumorigenicity, protein, may, act...   \n",
       "185  [nuclear, protein, act, transcriptional, repre...   \n",
       "186  [nuclear, receptor, coactivator, enhances, and...   \n",
       "187  [greb, like, protein, growth, regulation, estr...   \n",
       "188  [cancer, testis, antigen, belongs, ctag, pcc, ...   \n",
       "189  [cancer, testis, antigen, family, member, belo...   \n",
       "190  [cancer, testis, antigen, family, member, belo...   \n",
       "191  [casein, kinase, ii, subunit, alpha, probable,...   \n",
       "192          [cancer, testis, antigen, family, member]   \n",
       "193          [cancer, testis, antigen, family, member]   \n",
       "194          [cancer, testis, antigen, family, member]   \n",
       "195          [cancer, testis, antigen, family, member]   \n",
       "196                   [colorectal, cancer, associated]   \n",
       "197  [serine, threonine, protein, kinase, nek, prot...   \n",
       "198          [cancer, testis, antigen, family, member]   \n",
       "\n",
       "0    cancer_noun_identification  sent_negative  sent_positive  sent_neutral  \n",
       "0                             1          0.261          0.050         0.690  \n",
       "1                             1          0.221          0.059         0.720  \n",
       "2                             1          0.052          0.037         0.911  \n",
       "3                             1          0.314          0.120         0.566  \n",
       "4                             1          0.135          0.160         0.706  \n",
       "5                             1          0.120          0.080         0.800  \n",
       "6                             1          0.057          0.086         0.857  \n",
       "7                             1          0.045          0.128         0.828  \n",
       "8                             1          0.162          0.081         0.756  \n",
       "9                             1          0.110          0.037         0.853  \n",
       "10                            1          0.161          0.097         0.742  \n",
       "11                            1          0.066          0.086         0.848  \n",
       "12                            1          0.092          0.235         0.672  \n",
       "13                            1          0.074          0.058         0.869  \n",
       "14                            1          0.140          0.000         0.860  \n",
       "15                            1          0.124          0.000         0.876  \n",
       "16                            1          0.084          0.095         0.821  \n",
       "17                            1          0.688          0.000         0.312  \n",
       "18                            1          0.119          0.000         0.881  \n",
       "19                            1          0.000          0.072         0.928  \n",
       "20                            1          0.133          0.130         0.737  \n",
       "21                            1          0.153          0.000         0.847  \n",
       "22                            1          0.109          0.076         0.815  \n",
       "23                            1          0.157          0.067         0.776  \n",
       "24                            1          0.131          0.054         0.814  \n",
       "25                            1          0.215          0.040         0.746  \n",
       "26                            1          0.049          0.055         0.896  \n",
       "27                            1          0.092          0.124         0.784  \n",
       "28                            1          0.077          0.093         0.831  \n",
       "29                            1          0.161          0.133         0.705  \n",
       "..                          ...            ...            ...           ...  \n",
       "169                           1          0.071          0.214         0.714  \n",
       "170                           1          0.524          0.000         0.476  \n",
       "171                           1          0.524          0.000         0.476  \n",
       "172                           1          0.167          0.022         0.812  \n",
       "173                           1          0.165          0.139         0.695  \n",
       "174                           1          0.095          0.000         0.905  \n",
       "175                           1          0.688          0.000         0.312  \n",
       "176                           1          0.159          0.000         0.841  \n",
       "177                           1          0.328          0.000         0.672  \n",
       "178                           1          0.524          0.000         0.476  \n",
       "179                           1          0.111          0.019         0.870  \n",
       "180                           1          0.164          0.037         0.799  \n",
       "181                           1          0.106          0.073         0.821  \n",
       "182                           1          0.207          0.054         0.739  \n",
       "183                           1          0.274          0.048         0.678  \n",
       "184                           1          0.343          0.118         0.539  \n",
       "185                           1          0.254          0.036         0.710  \n",
       "186                           1          0.161          0.000         0.839  \n",
       "187                           1          0.183          0.317         0.500  \n",
       "188                           1          0.355          0.000         0.645  \n",
       "189                           1          0.328          0.000         0.672  \n",
       "190                           1          0.328          0.000         0.672  \n",
       "191                           1          0.096          0.121         0.783  \n",
       "192                           1          0.524          0.000         0.476  \n",
       "193                           1          0.524          0.000         0.476  \n",
       "194                           1          0.524          0.000         0.476  \n",
       "195                           1          0.524          0.000         0.476  \n",
       "196                           1          0.688          0.000         0.312  \n",
       "197                           1          0.062          0.000         0.938  \n",
       "198                           1          0.524          0.000         0.476  \n",
       "\n",
       "[199 rows x 10 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cancer_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "cancer_df['cancer'] = np.where((cancer_df['sent_negative'] > cancer_df['sent_positive']) & (cancer_df['sent_negative'] > cancer_df['sent_neutral']),[1],[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>protein_external_id</th>\n",
       "      <th>preferred_name</th>\n",
       "      <th>protein_size</th>\n",
       "      <th>annotation</th>\n",
       "      <th>nlp_processing</th>\n",
       "      <th>cancer_noun_identification</th>\n",
       "      <th>sent_negative</th>\n",
       "      <th>sent_positive</th>\n",
       "      <th>sent_neutral</th>\n",
       "      <th>cancer</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>102</td>\n",
       "      <td>9606.ENSP00000064778</td>\n",
       "      <td>FAM168A</td>\n",
       "      <td>244</td>\n",
       "      <td>protein fama in cancer context protects cells ...</td>\n",
       "      <td>[protein, fama, cancer, context, protects, cel...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.261</td>\n",
       "      <td>0.050</td>\n",
       "      <td>0.690</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>156</td>\n",
       "      <td>9606.ENSP00000167106</td>\n",
       "      <td>VASH1</td>\n",
       "      <td>365</td>\n",
       "      <td>vasohibin  angiogenesis inhibitor inhibits mig...</td>\n",
       "      <td>[vasohibin, angiogenesis, inhibitor, inhibits,...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.221</td>\n",
       "      <td>0.059</td>\n",
       "      <td>0.720</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>241</td>\n",
       "      <td>9606.ENSP00000201031</td>\n",
       "      <td>TFAP2C</td>\n",
       "      <td>450</td>\n",
       "      <td>transcription factor ap  gamma sequence specif...</td>\n",
       "      <td>[transcription, factor, ap, gamma, sequence, s...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.052</td>\n",
       "      <td>0.037</td>\n",
       "      <td>0.911</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>441</td>\n",
       "      <td>9606.ENSP00000216807</td>\n",
       "      <td>BRMS1L</td>\n",
       "      <td>323</td>\n",
       "      <td>breast cancer metastasis suppressor  like prot...</td>\n",
       "      <td>[breast, cancer, metastasis, suppressor, like,...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.314</td>\n",
       "      <td>0.120</td>\n",
       "      <td>0.566</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>514</td>\n",
       "      <td>9606.ENSP00000219022</td>\n",
       "      <td>OLFM4</td>\n",
       "      <td>510</td>\n",
       "      <td>olfactomedin  may promote proliferation of pan...</td>\n",
       "      <td>[olfactomedin, may, promote, proliferation, pa...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.135</td>\n",
       "      <td>0.160</td>\n",
       "      <td>0.706</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>589</td>\n",
       "      <td>9606.ENSP00000220812</td>\n",
       "      <td>DKK4</td>\n",
       "      <td>224</td>\n",
       "      <td>dickkopf related protein  antagonizes canonica...</td>\n",
       "      <td>[dickkopf, related, protein, antagonizes, cano...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.120</td>\n",
       "      <td>0.080</td>\n",
       "      <td>0.800</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>815</td>\n",
       "      <td>9606.ENSP00000226230</td>\n",
       "      <td>TMEM97</td>\n",
       "      <td>176</td>\n",
       "      <td>sigma intracellular receptor  intracellular or...</td>\n",
       "      <td>[sigma, intracellular, receptor, intracellular...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.057</td>\n",
       "      <td>0.086</td>\n",
       "      <td>0.857</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>891</td>\n",
       "      <td>9606.ENSP00000228865</td>\n",
       "      <td>CREBL2</td>\n",
       "      <td>120</td>\n",
       "      <td>camp responsive element binding protein like  ...</td>\n",
       "      <td>[camp, responsive, element, binding, protein, ...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.045</td>\n",
       "      <td>0.128</td>\n",
       "      <td>0.828</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>999</td>\n",
       "      <td>9606.ENSP00000231656</td>\n",
       "      <td>CDX1</td>\n",
       "      <td>265</td>\n",
       "      <td>homeobox protein cdx  plays a role in transcri...</td>\n",
       "      <td>[homeobox, protein, cdx, play, role, transcrip...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.162</td>\n",
       "      <td>0.081</td>\n",
       "      <td>0.756</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1065</td>\n",
       "      <td>9606.ENSP00000233948</td>\n",
       "      <td>WNT6</td>\n",
       "      <td>365</td>\n",
       "      <td>protein wnt  ligand for members of the frizzle...</td>\n",
       "      <td>[protein, wnt, ligand, member, frizzled, famil...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.110</td>\n",
       "      <td>0.037</td>\n",
       "      <td>0.853</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>1121</td>\n",
       "      <td>9606.ENSP00000236671</td>\n",
       "      <td>CTSD</td>\n",
       "      <td>412</td>\n",
       "      <td>cathepsin d acid protease active in intracellu...</td>\n",
       "      <td>[cathepsin, acid, protease, active, intracellu...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.161</td>\n",
       "      <td>0.097</td>\n",
       "      <td>0.742</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>1145</td>\n",
       "      <td>9606.ENSP00000237612</td>\n",
       "      <td>ABCG2</td>\n",
       "      <td>655</td>\n",
       "      <td>atp binding cassette sub family g member  high...</td>\n",
       "      <td>[atp, binding, cassette, sub, family, g, membe...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.066</td>\n",
       "      <td>0.086</td>\n",
       "      <td>0.848</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>1178</td>\n",
       "      <td>9606.ENSP00000238823</td>\n",
       "      <td>FAM98A</td>\n",
       "      <td>518</td>\n",
       "      <td>protein fama positively stimulates prmt induce...</td>\n",
       "      <td>[protein, fama, positively, stimulates, prmt, ...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.092</td>\n",
       "      <td>0.235</td>\n",
       "      <td>0.672</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>1204</td>\n",
       "      <td>9606.ENSP00000239462</td>\n",
       "      <td>TNN</td>\n",
       "      <td>1299</td>\n",
       "      <td>tenascin n extracellular matrix protein that s...</td>\n",
       "      <td>[tenascin, n, extracellular, matrix, protein, ...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.074</td>\n",
       "      <td>0.058</td>\n",
       "      <td>0.869</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>1372</td>\n",
       "      <td>9606.ENSP00000244926</td>\n",
       "      <td>SCGB1D2</td>\n",
       "      <td>90</td>\n",
       "      <td>secretoglobin family d member  may bind androg...</td>\n",
       "      <td>[secretoglobin, family, member, may, bind, and...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.140</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.860</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>1373</td>\n",
       "      <td>9606.ENSP00000244930</td>\n",
       "      <td>SCGB2A1</td>\n",
       "      <td>95</td>\n",
       "      <td>mammaglobin b may bind androgens and other ste...</td>\n",
       "      <td>[mammaglobin, b, may, bind, androgen, steroid,...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.124</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.876</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>1471</td>\n",
       "      <td>9606.ENSP00000247191</td>\n",
       "      <td>DLGAP5</td>\n",
       "      <td>846</td>\n",
       "      <td>disks large associated protein  potential cell...</td>\n",
       "      <td>[disk, large, associated, protein, potential, ...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.084</td>\n",
       "      <td>0.095</td>\n",
       "      <td>0.821</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>1478</td>\n",
       "      <td>9606.ENSP00000247306</td>\n",
       "      <td>CTAG2</td>\n",
       "      <td>210</td>\n",
       "      <td>cancer testis antigen</td>\n",
       "      <td>[cancer, testis, antigen]</td>\n",
       "      <td>1</td>\n",
       "      <td>0.688</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.312</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>1646</td>\n",
       "      <td>9606.ENSP00000251691</td>\n",
       "      <td>KIAA1244</td>\n",
       "      <td>2177</td>\n",
       "      <td>brefeldin a inhibited guanine nucleotide excha...</td>\n",
       "      <td>[brefeldin, inhibited, guanine, nucleotide, ex...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.119</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.881</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>1899</td>\n",
       "      <td>9606.ENSP00000255681</td>\n",
       "      <td>MACROD1</td>\n",
       "      <td>325</td>\n",
       "      <td>o acetyl adp ribose deacetylase macrod removes...</td>\n",
       "      <td>[acetyl, adp, ribose, deacetylase, macrod, rem...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.072</td>\n",
       "      <td>0.928</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>1912</td>\n",
       "      <td>9606.ENSP00000256078</td>\n",
       "      <td>KRAS</td>\n",
       "      <td>189</td>\n",
       "      <td>gtpase kras ras proteins bind gdp gtp and poss...</td>\n",
       "      <td>[gtpase, kras, ra, protein, bind, gdp, gtp, po...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.133</td>\n",
       "      <td>0.130</td>\n",
       "      <td>0.737</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>1925</td>\n",
       "      <td>9606.ENSP00000256261</td>\n",
       "      <td>DUSP26</td>\n",
       "      <td>211</td>\n",
       "      <td>dual specificity protein phosphatase  inactiva...</td>\n",
       "      <td>[dual, specificity, protein, phosphatase, inac...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.153</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.847</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>2105</td>\n",
       "      <td>9606.ENSP00000258729</td>\n",
       "      <td>IGF2BP3</td>\n",
       "      <td>579</td>\n",
       "      <td>insulin like growth factor  mrna binding prote...</td>\n",
       "      <td>[insulin, like, growth, factor, mrna, binding,...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.109</td>\n",
       "      <td>0.076</td>\n",
       "      <td>0.815</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>2210</td>\n",
       "      <td>9606.ENSP00000260187</td>\n",
       "      <td>USP2</td>\n",
       "      <td>605</td>\n",
       "      <td>ubiquitin carboxyl terminal hydrolase  hydrola...</td>\n",
       "      <td>[ubiquitin, carboxyl, terminal, hydrolase, hyd...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.157</td>\n",
       "      <td>0.067</td>\n",
       "      <td>0.776</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>2434</td>\n",
       "      <td>9606.ENSP00000261893</td>\n",
       "      <td>LACTB</td>\n",
       "      <td>547</td>\n",
       "      <td>serine beta lactamase like protein lactb mitoc...</td>\n",
       "      <td>[serine, beta, lactamase, like, protein, lactb...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.131</td>\n",
       "      <td>0.054</td>\n",
       "      <td>0.814</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>2461</td>\n",
       "      <td>9606.ENSP00000262102</td>\n",
       "      <td>MTUS1</td>\n",
       "      <td>1270</td>\n",
       "      <td>microtubule associated tumor suppressor  coope...</td>\n",
       "      <td>[microtubule, associated, tumor, suppressor, c...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.215</td>\n",
       "      <td>0.040</td>\n",
       "      <td>0.746</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>2510</td>\n",
       "      <td>9606.ENSP00000262305</td>\n",
       "      <td>RAB11FIP3</td>\n",
       "      <td>756</td>\n",
       "      <td>rab family interacting protein  acts as a regu...</td>\n",
       "      <td>[rab, family, interacting, protein, act, regul...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.049</td>\n",
       "      <td>0.055</td>\n",
       "      <td>0.896</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>2647</td>\n",
       "      <td>9606.ENSP00000263083</td>\n",
       "      <td>DPH1</td>\n",
       "      <td>443</td>\n",
       "      <td>amino  carboxypropyl histidine synthase subu...</td>\n",
       "      <td>[amino, carboxypropyl, histidine, synthase, su...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.092</td>\n",
       "      <td>0.124</td>\n",
       "      <td>0.784</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>2913</td>\n",
       "      <td>9606.ENSP00000264606</td>\n",
       "      <td>HDAC4</td>\n",
       "      <td>1084</td>\n",
       "      <td>histone deacetylase  responsible for the deace...</td>\n",
       "      <td>[histone, deacetylase, responsible, deacetylat...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.077</td>\n",
       "      <td>0.093</td>\n",
       "      <td>0.831</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>2999</td>\n",
       "      <td>9606.ENSP00000265052</td>\n",
       "      <td>MGLL</td>\n",
       "      <td>313</td>\n",
       "      <td>monoglyceride lipase converts monoacylglycerid...</td>\n",
       "      <td>[monoglyceride, lipase, convert, monoacylglyce...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.161</td>\n",
       "      <td>0.133</td>\n",
       "      <td>0.705</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>169</th>\n",
       "      <td>16150</td>\n",
       "      <td>9606.ENSP00000413520</td>\n",
       "      <td>HID1</td>\n",
       "      <td>788</td>\n",
       "      <td>protein hid may play an important role in the ...</td>\n",
       "      <td>[protein, hid, may, play, important, role, dev...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.071</td>\n",
       "      <td>0.214</td>\n",
       "      <td>0.714</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>170</th>\n",
       "      <td>16180</td>\n",
       "      <td>9606.ENSP00000414253</td>\n",
       "      <td>CT47A2</td>\n",
       "      <td>288</td>\n",
       "      <td>cancer testis antigen family  member a</td>\n",
       "      <td>[cancer, testis, antigen, family, member]</td>\n",
       "      <td>1</td>\n",
       "      <td>0.524</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.476</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>171</th>\n",
       "      <td>16193</td>\n",
       "      <td>9606.ENSP00000414509</td>\n",
       "      <td>CT47A12</td>\n",
       "      <td>288</td>\n",
       "      <td>cancer testis antigen family  member a</td>\n",
       "      <td>[cancer, testis, antigen, family, member]</td>\n",
       "      <td>1</td>\n",
       "      <td>0.524</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.476</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>172</th>\n",
       "      <td>16455</td>\n",
       "      <td>9606.ENSP00000418960</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>1884</td>\n",
       "      <td>breast cancer type  susceptibility protein e u...</td>\n",
       "      <td>[breast, cancer, type, susceptibility, protein...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.167</td>\n",
       "      <td>0.022</td>\n",
       "      <td>0.812</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>173</th>\n",
       "      <td>16480</td>\n",
       "      <td>9606.ENSP00000419235</td>\n",
       "      <td>FAM115A</td>\n",
       "      <td>921</td>\n",
       "      <td>trpm channel associated factor  positively reg...</td>\n",
       "      <td>[trpm, channel, associated, factor, positively...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.165</td>\n",
       "      <td>0.139</td>\n",
       "      <td>0.695</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>174</th>\n",
       "      <td>16542</td>\n",
       "      <td>9606.ENSP00000420095</td>\n",
       "      <td>ZMYND8</td>\n",
       "      <td>1234</td>\n",
       "      <td>protein kinase c binding protein  may act as a...</td>\n",
       "      <td>[protein, kinase, c, binding, protein, may, ac...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.095</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.905</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175</th>\n",
       "      <td>16781</td>\n",
       "      <td>9606.ENSP00000425493</td>\n",
       "      <td>CAGE1</td>\n",
       "      <td>839</td>\n",
       "      <td>cancer associated gene  protein cancer antigen</td>\n",
       "      <td>[cancer, associated, gene, protein, cancer, an...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.688</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.312</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>176</th>\n",
       "      <td>16839</td>\n",
       "      <td>9606.ENSP00000427015</td>\n",
       "      <td>RPS6KA2</td>\n",
       "      <td>741</td>\n",
       "      <td>ribosomal protein s kinase alpha  serine threo...</td>\n",
       "      <td>[ribosomal, protein, kinase, alpha, serine, th...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.159</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.841</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>177</th>\n",
       "      <td>16851</td>\n",
       "      <td>9606.ENSP00000427342</td>\n",
       "      <td>CT45A5</td>\n",
       "      <td>189</td>\n",
       "      <td>cancer testis antigen family  member a belongs...</td>\n",
       "      <td>[cancer, testis, antigen, family, member, belo...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.328</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.672</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>178</th>\n",
       "      <td>17080</td>\n",
       "      <td>9606.ENSP00000432353</td>\n",
       "      <td>PBOV1</td>\n",
       "      <td>135</td>\n",
       "      <td>prostate and breast cancer overexpressed</td>\n",
       "      <td>[prostate, breast, cancer, overexpressed]</td>\n",
       "      <td>1</td>\n",
       "      <td>0.524</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.476</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>179</th>\n",
       "      <td>17154</td>\n",
       "      <td>9606.ENSP00000434045</td>\n",
       "      <td>GPR68</td>\n",
       "      <td>365</td>\n",
       "      <td>ovarian cancer g protein coupled receptor  pro...</td>\n",
       "      <td>[ovarian, cancer, g, protein, coupled, recepto...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.111</td>\n",
       "      <td>0.019</td>\n",
       "      <td>0.870</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>180</th>\n",
       "      <td>17433</td>\n",
       "      <td>9606.ENSP00000441140</td>\n",
       "      <td>NOL8</td>\n",
       "      <td>1167</td>\n",
       "      <td>nucleolar protein  plays an essential role in ...</td>\n",
       "      <td>[nucleolar, protein, play, essential, role, su...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.164</td>\n",
       "      <td>0.037</td>\n",
       "      <td>0.799</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>181</th>\n",
       "      <td>17495</td>\n",
       "      <td>9606.ENSP00000442464</td>\n",
       "      <td>NCEH1</td>\n",
       "      <td>448</td>\n",
       "      <td>neutral cholesterol ester hydrolase  hydrolyze...</td>\n",
       "      <td>[neutral, cholesterol, ester, hydrolase, hydro...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.106</td>\n",
       "      <td>0.073</td>\n",
       "      <td>0.821</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>182</th>\n",
       "      <td>17499</td>\n",
       "      <td>9606.ENSP00000442521</td>\n",
       "      <td>BEX2</td>\n",
       "      <td>160</td>\n",
       "      <td>protein bex regulator of mitochondrial apoptos...</td>\n",
       "      <td>[protein, bex, regulator, mitochondrial, apopt...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.207</td>\n",
       "      <td>0.054</td>\n",
       "      <td>0.739</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>183</th>\n",
       "      <td>17574</td>\n",
       "      <td>9606.ENSP00000444412</td>\n",
       "      <td>ZNF268</td>\n",
       "      <td>947</td>\n",
       "      <td>zinc finger protein  isoform  acts as a transc...</td>\n",
       "      <td>[zinc, finger, protein, isoform, act, transcri...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.274</td>\n",
       "      <td>0.048</td>\n",
       "      <td>0.678</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>184</th>\n",
       "      <td>17886</td>\n",
       "      <td>9606.ENSP00000453502</td>\n",
       "      <td>ST20</td>\n",
       "      <td>79</td>\n",
       "      <td>suppressor of tumorigenicity  protein may act ...</td>\n",
       "      <td>[suppressor, tumorigenicity, protein, may, act...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.343</td>\n",
       "      <td>0.118</td>\n",
       "      <td>0.539</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>185</th>\n",
       "      <td>17960</td>\n",
       "      <td>9606.ENSP00000455442</td>\n",
       "      <td>NUPR1L</td>\n",
       "      <td>97</td>\n",
       "      <td>nuclear protein  acts as a transcriptional rep...</td>\n",
       "      <td>[nuclear, protein, act, transcriptional, repre...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.254</td>\n",
       "      <td>0.036</td>\n",
       "      <td>0.710</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>186</th>\n",
       "      <td>18144</td>\n",
       "      <td>9606.ENSP00000463027</td>\n",
       "      <td>NCOA4</td>\n",
       "      <td>650</td>\n",
       "      <td>nuclear receptor coactivator  enhances the and...</td>\n",
       "      <td>[nuclear, receptor, coactivator, enhances, and...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.161</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.839</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>187</th>\n",
       "      <td>18188</td>\n",
       "      <td>9606.ENSP00000464162</td>\n",
       "      <td>GREB1L</td>\n",
       "      <td>1923</td>\n",
       "      <td>greb like protein growth regulation by estroge...</td>\n",
       "      <td>[greb, like, protein, growth, regulation, estr...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.183</td>\n",
       "      <td>0.317</td>\n",
       "      <td>0.500</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>188</th>\n",
       "      <td>18371</td>\n",
       "      <td>9606.ENSP00000469441</td>\n",
       "      <td>CTAG1A</td>\n",
       "      <td>180</td>\n",
       "      <td>cancer testis antigen a belongs to the ctag pc...</td>\n",
       "      <td>[cancer, testis, antigen, belongs, ctag, pcc, ...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.355</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.645</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>189</th>\n",
       "      <td>18403</td>\n",
       "      <td>9606.ENSP00000470185</td>\n",
       "      <td>CT45A1</td>\n",
       "      <td>189</td>\n",
       "      <td>cancer testis antigen family  member a belongs...</td>\n",
       "      <td>[cancer, testis, antigen, family, member, belo...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.328</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.672</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>190</th>\n",
       "      <td>18446</td>\n",
       "      <td>9606.ENSP00000471418</td>\n",
       "      <td>ENSG00000269096</td>\n",
       "      <td>189</td>\n",
       "      <td>cancer testis antigen family  member a belongs...</td>\n",
       "      <td>[cancer, testis, antigen, family, member, belo...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.328</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.672</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191</th>\n",
       "      <td>18533</td>\n",
       "      <td>9606.ENSP00000473553</td>\n",
       "      <td>CSNK2A3</td>\n",
       "      <td>391</td>\n",
       "      <td>casein kinase ii subunit alpha  probable catal...</td>\n",
       "      <td>[casein, kinase, ii, subunit, alpha, probable,...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.096</td>\n",
       "      <td>0.121</td>\n",
       "      <td>0.783</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>192</th>\n",
       "      <td>18565</td>\n",
       "      <td>9606.ENSP00000474632</td>\n",
       "      <td>ENSG00000271449</td>\n",
       "      <td>189</td>\n",
       "      <td>cancer testis antigen family  member a</td>\n",
       "      <td>[cancer, testis, antigen, family, member]</td>\n",
       "      <td>1</td>\n",
       "      <td>0.524</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.476</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>193</th>\n",
       "      <td>18861</td>\n",
       "      <td>9606.ENSP00000479608</td>\n",
       "      <td>ENSG00000278289</td>\n",
       "      <td>189</td>\n",
       "      <td>cancer testis antigen family  member a</td>\n",
       "      <td>[cancer, testis, antigen, family, member]</td>\n",
       "      <td>1</td>\n",
       "      <td>0.524</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.476</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>194</th>\n",
       "      <td>19027</td>\n",
       "      <td>9606.ENSP00000481184</td>\n",
       "      <td>CT45A4</td>\n",
       "      <td>189</td>\n",
       "      <td>cancer testis antigen family  member a</td>\n",
       "      <td>[cancer, testis, antigen, family, member]</td>\n",
       "      <td>1</td>\n",
       "      <td>0.524</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.476</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>195</th>\n",
       "      <td>19252</td>\n",
       "      <td>9606.ENSP00000483637</td>\n",
       "      <td>CT45-2</td>\n",
       "      <td>189</td>\n",
       "      <td>cancer testis antigen family  member a</td>\n",
       "      <td>[cancer, testis, antigen, family, member]</td>\n",
       "      <td>1</td>\n",
       "      <td>0.524</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.476</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>196</th>\n",
       "      <td>19289</td>\n",
       "      <td>9606.ENSP00000484135</td>\n",
       "      <td>COLCA2</td>\n",
       "      <td>251</td>\n",
       "      <td>colorectal cancer associated</td>\n",
       "      <td>[colorectal, cancer, associated]</td>\n",
       "      <td>1</td>\n",
       "      <td>0.688</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.312</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>197</th>\n",
       "      <td>19308</td>\n",
       "      <td>9606.ENSP00000484443</td>\n",
       "      <td>NEK3</td>\n",
       "      <td>506</td>\n",
       "      <td>serine threonine protein kinase nek protein ki...</td>\n",
       "      <td>[serine, threonine, protein, kinase, nek, prot...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.062</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.938</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>198</th>\n",
       "      <td>19322</td>\n",
       "      <td>9606.ENSP00000484601</td>\n",
       "      <td>CT45A10</td>\n",
       "      <td>189</td>\n",
       "      <td>cancer testis antigen family  member a</td>\n",
       "      <td>[cancer, testis, antigen, family, member]</td>\n",
       "      <td>1</td>\n",
       "      <td>0.524</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.476</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>199 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "0    index   protein_external_id   preferred_name protein_size  \\\n",
       "0      102  9606.ENSP00000064778          FAM168A          244   \n",
       "1      156  9606.ENSP00000167106            VASH1          365   \n",
       "2      241  9606.ENSP00000201031           TFAP2C          450   \n",
       "3      441  9606.ENSP00000216807           BRMS1L          323   \n",
       "4      514  9606.ENSP00000219022            OLFM4          510   \n",
       "5      589  9606.ENSP00000220812             DKK4          224   \n",
       "6      815  9606.ENSP00000226230           TMEM97          176   \n",
       "7      891  9606.ENSP00000228865           CREBL2          120   \n",
       "8      999  9606.ENSP00000231656             CDX1          265   \n",
       "9     1065  9606.ENSP00000233948             WNT6          365   \n",
       "10    1121  9606.ENSP00000236671             CTSD          412   \n",
       "11    1145  9606.ENSP00000237612            ABCG2          655   \n",
       "12    1178  9606.ENSP00000238823           FAM98A          518   \n",
       "13    1204  9606.ENSP00000239462              TNN         1299   \n",
       "14    1372  9606.ENSP00000244926          SCGB1D2           90   \n",
       "15    1373  9606.ENSP00000244930          SCGB2A1           95   \n",
       "16    1471  9606.ENSP00000247191           DLGAP5          846   \n",
       "17    1478  9606.ENSP00000247306            CTAG2          210   \n",
       "18    1646  9606.ENSP00000251691         KIAA1244         2177   \n",
       "19    1899  9606.ENSP00000255681          MACROD1          325   \n",
       "20    1912  9606.ENSP00000256078             KRAS          189   \n",
       "21    1925  9606.ENSP00000256261           DUSP26          211   \n",
       "22    2105  9606.ENSP00000258729          IGF2BP3          579   \n",
       "23    2210  9606.ENSP00000260187             USP2          605   \n",
       "24    2434  9606.ENSP00000261893            LACTB          547   \n",
       "25    2461  9606.ENSP00000262102            MTUS1         1270   \n",
       "26    2510  9606.ENSP00000262305        RAB11FIP3          756   \n",
       "27    2647  9606.ENSP00000263083             DPH1          443   \n",
       "28    2913  9606.ENSP00000264606            HDAC4         1084   \n",
       "29    2999  9606.ENSP00000265052             MGLL          313   \n",
       "..     ...                   ...              ...          ...   \n",
       "169  16150  9606.ENSP00000413520             HID1          788   \n",
       "170  16180  9606.ENSP00000414253           CT47A2          288   \n",
       "171  16193  9606.ENSP00000414509          CT47A12          288   \n",
       "172  16455  9606.ENSP00000418960            BRCA1         1884   \n",
       "173  16480  9606.ENSP00000419235          FAM115A          921   \n",
       "174  16542  9606.ENSP00000420095           ZMYND8         1234   \n",
       "175  16781  9606.ENSP00000425493            CAGE1          839   \n",
       "176  16839  9606.ENSP00000427015          RPS6KA2          741   \n",
       "177  16851  9606.ENSP00000427342           CT45A5          189   \n",
       "178  17080  9606.ENSP00000432353            PBOV1          135   \n",
       "179  17154  9606.ENSP00000434045            GPR68          365   \n",
       "180  17433  9606.ENSP00000441140             NOL8         1167   \n",
       "181  17495  9606.ENSP00000442464            NCEH1          448   \n",
       "182  17499  9606.ENSP00000442521             BEX2          160   \n",
       "183  17574  9606.ENSP00000444412           ZNF268          947   \n",
       "184  17886  9606.ENSP00000453502             ST20           79   \n",
       "185  17960  9606.ENSP00000455442           NUPR1L           97   \n",
       "186  18144  9606.ENSP00000463027            NCOA4          650   \n",
       "187  18188  9606.ENSP00000464162           GREB1L         1923   \n",
       "188  18371  9606.ENSP00000469441           CTAG1A          180   \n",
       "189  18403  9606.ENSP00000470185           CT45A1          189   \n",
       "190  18446  9606.ENSP00000471418  ENSG00000269096          189   \n",
       "191  18533  9606.ENSP00000473553          CSNK2A3          391   \n",
       "192  18565  9606.ENSP00000474632  ENSG00000271449          189   \n",
       "193  18861  9606.ENSP00000479608  ENSG00000278289          189   \n",
       "194  19027  9606.ENSP00000481184           CT45A4          189   \n",
       "195  19252  9606.ENSP00000483637           CT45-2          189   \n",
       "196  19289  9606.ENSP00000484135           COLCA2          251   \n",
       "197  19308  9606.ENSP00000484443             NEK3          506   \n",
       "198  19322  9606.ENSP00000484601          CT45A10          189   \n",
       "\n",
       "0                                         annotation\\n  \\\n",
       "0    protein fama in cancer context protects cells ...   \n",
       "1    vasohibin  angiogenesis inhibitor inhibits mig...   \n",
       "2    transcription factor ap  gamma sequence specif...   \n",
       "3    breast cancer metastasis suppressor  like prot...   \n",
       "4    olfactomedin  may promote proliferation of pan...   \n",
       "5    dickkopf related protein  antagonizes canonica...   \n",
       "6    sigma intracellular receptor  intracellular or...   \n",
       "7    camp responsive element binding protein like  ...   \n",
       "8    homeobox protein cdx  plays a role in transcri...   \n",
       "9    protein wnt  ligand for members of the frizzle...   \n",
       "10   cathepsin d acid protease active in intracellu...   \n",
       "11   atp binding cassette sub family g member  high...   \n",
       "12   protein fama positively stimulates prmt induce...   \n",
       "13   tenascin n extracellular matrix protein that s...   \n",
       "14   secretoglobin family d member  may bind androg...   \n",
       "15   mammaglobin b may bind androgens and other ste...   \n",
       "16   disks large associated protein  potential cell...   \n",
       "17                             cancer testis antigen     \n",
       "18   brefeldin a inhibited guanine nucleotide excha...   \n",
       "19   o acetyl adp ribose deacetylase macrod removes...   \n",
       "20   gtpase kras ras proteins bind gdp gtp and poss...   \n",
       "21   dual specificity protein phosphatase  inactiva...   \n",
       "22   insulin like growth factor  mrna binding prote...   \n",
       "23   ubiquitin carboxyl terminal hydrolase  hydrola...   \n",
       "24   serine beta lactamase like protein lactb mitoc...   \n",
       "25   microtubule associated tumor suppressor  coope...   \n",
       "26   rab family interacting protein  acts as a regu...   \n",
       "27     amino  carboxypropyl histidine synthase subu...   \n",
       "28   histone deacetylase  responsible for the deace...   \n",
       "29   monoglyceride lipase converts monoacylglycerid...   \n",
       "..                                                 ...   \n",
       "169  protein hid may play an important role in the ...   \n",
       "170            cancer testis antigen family  member a    \n",
       "171            cancer testis antigen family  member a    \n",
       "172  breast cancer type  susceptibility protein e u...   \n",
       "173  trpm channel associated factor  positively reg...   \n",
       "174  protein kinase c binding protein  may act as a...   \n",
       "175   cancer associated gene  protein cancer antigen     \n",
       "176  ribosomal protein s kinase alpha  serine threo...   \n",
       "177  cancer testis antigen family  member a belongs...   \n",
       "178         prostate and breast cancer overexpressed     \n",
       "179  ovarian cancer g protein coupled receptor  pro...   \n",
       "180  nucleolar protein  plays an essential role in ...   \n",
       "181  neutral cholesterol ester hydrolase  hydrolyze...   \n",
       "182  protein bex regulator of mitochondrial apoptos...   \n",
       "183  zinc finger protein  isoform  acts as a transc...   \n",
       "184  suppressor of tumorigenicity  protein may act ...   \n",
       "185  nuclear protein  acts as a transcriptional rep...   \n",
       "186  nuclear receptor coactivator  enhances the and...   \n",
       "187  greb like protein growth regulation by estroge...   \n",
       "188  cancer testis antigen a belongs to the ctag pc...   \n",
       "189  cancer testis antigen family  member a belongs...   \n",
       "190  cancer testis antigen family  member a belongs...   \n",
       "191  casein kinase ii subunit alpha  probable catal...   \n",
       "192            cancer testis antigen family  member a    \n",
       "193            cancer testis antigen family  member a    \n",
       "194            cancer testis antigen family  member a    \n",
       "195            cancer testis antigen family  member a    \n",
       "196                     colorectal cancer associated     \n",
       "197  serine threonine protein kinase nek protein ki...   \n",
       "198            cancer testis antigen family  member a    \n",
       "\n",
       "0                                       nlp_processing  \\\n",
       "0    [protein, fama, cancer, context, protects, cel...   \n",
       "1    [vasohibin, angiogenesis, inhibitor, inhibits,...   \n",
       "2    [transcription, factor, ap, gamma, sequence, s...   \n",
       "3    [breast, cancer, metastasis, suppressor, like,...   \n",
       "4    [olfactomedin, may, promote, proliferation, pa...   \n",
       "5    [dickkopf, related, protein, antagonizes, cano...   \n",
       "6    [sigma, intracellular, receptor, intracellular...   \n",
       "7    [camp, responsive, element, binding, protein, ...   \n",
       "8    [homeobox, protein, cdx, play, role, transcrip...   \n",
       "9    [protein, wnt, ligand, member, frizzled, famil...   \n",
       "10   [cathepsin, acid, protease, active, intracellu...   \n",
       "11   [atp, binding, cassette, sub, family, g, membe...   \n",
       "12   [protein, fama, positively, stimulates, prmt, ...   \n",
       "13   [tenascin, n, extracellular, matrix, protein, ...   \n",
       "14   [secretoglobin, family, member, may, bind, and...   \n",
       "15   [mammaglobin, b, may, bind, androgen, steroid,...   \n",
       "16   [disk, large, associated, protein, potential, ...   \n",
       "17                           [cancer, testis, antigen]   \n",
       "18   [brefeldin, inhibited, guanine, nucleotide, ex...   \n",
       "19   [acetyl, adp, ribose, deacetylase, macrod, rem...   \n",
       "20   [gtpase, kras, ra, protein, bind, gdp, gtp, po...   \n",
       "21   [dual, specificity, protein, phosphatase, inac...   \n",
       "22   [insulin, like, growth, factor, mrna, binding,...   \n",
       "23   [ubiquitin, carboxyl, terminal, hydrolase, hyd...   \n",
       "24   [serine, beta, lactamase, like, protein, lactb...   \n",
       "25   [microtubule, associated, tumor, suppressor, c...   \n",
       "26   [rab, family, interacting, protein, act, regul...   \n",
       "27   [amino, carboxypropyl, histidine, synthase, su...   \n",
       "28   [histone, deacetylase, responsible, deacetylat...   \n",
       "29   [monoglyceride, lipase, convert, monoacylglyce...   \n",
       "..                                                 ...   \n",
       "169  [protein, hid, may, play, important, role, dev...   \n",
       "170          [cancer, testis, antigen, family, member]   \n",
       "171          [cancer, testis, antigen, family, member]   \n",
       "172  [breast, cancer, type, susceptibility, protein...   \n",
       "173  [trpm, channel, associated, factor, positively...   \n",
       "174  [protein, kinase, c, binding, protein, may, ac...   \n",
       "175  [cancer, associated, gene, protein, cancer, an...   \n",
       "176  [ribosomal, protein, kinase, alpha, serine, th...   \n",
       "177  [cancer, testis, antigen, family, member, belo...   \n",
       "178          [prostate, breast, cancer, overexpressed]   \n",
       "179  [ovarian, cancer, g, protein, coupled, recepto...   \n",
       "180  [nucleolar, protein, play, essential, role, su...   \n",
       "181  [neutral, cholesterol, ester, hydrolase, hydro...   \n",
       "182  [protein, bex, regulator, mitochondrial, apopt...   \n",
       "183  [zinc, finger, protein, isoform, act, transcri...   \n",
       "184  [suppressor, tumorigenicity, protein, may, act...   \n",
       "185  [nuclear, protein, act, transcriptional, repre...   \n",
       "186  [nuclear, receptor, coactivator, enhances, and...   \n",
       "187  [greb, like, protein, growth, regulation, estr...   \n",
       "188  [cancer, testis, antigen, belongs, ctag, pcc, ...   \n",
       "189  [cancer, testis, antigen, family, member, belo...   \n",
       "190  [cancer, testis, antigen, family, member, belo...   \n",
       "191  [casein, kinase, ii, subunit, alpha, probable,...   \n",
       "192          [cancer, testis, antigen, family, member]   \n",
       "193          [cancer, testis, antigen, family, member]   \n",
       "194          [cancer, testis, antigen, family, member]   \n",
       "195          [cancer, testis, antigen, family, member]   \n",
       "196                   [colorectal, cancer, associated]   \n",
       "197  [serine, threonine, protein, kinase, nek, prot...   \n",
       "198          [cancer, testis, antigen, family, member]   \n",
       "\n",
       "0    cancer_noun_identification  sent_negative  sent_positive  sent_neutral  \\\n",
       "0                             1          0.261          0.050         0.690   \n",
       "1                             1          0.221          0.059         0.720   \n",
       "2                             1          0.052          0.037         0.911   \n",
       "3                             1          0.314          0.120         0.566   \n",
       "4                             1          0.135          0.160         0.706   \n",
       "5                             1          0.120          0.080         0.800   \n",
       "6                             1          0.057          0.086         0.857   \n",
       "7                             1          0.045          0.128         0.828   \n",
       "8                             1          0.162          0.081         0.756   \n",
       "9                             1          0.110          0.037         0.853   \n",
       "10                            1          0.161          0.097         0.742   \n",
       "11                            1          0.066          0.086         0.848   \n",
       "12                            1          0.092          0.235         0.672   \n",
       "13                            1          0.074          0.058         0.869   \n",
       "14                            1          0.140          0.000         0.860   \n",
       "15                            1          0.124          0.000         0.876   \n",
       "16                            1          0.084          0.095         0.821   \n",
       "17                            1          0.688          0.000         0.312   \n",
       "18                            1          0.119          0.000         0.881   \n",
       "19                            1          0.000          0.072         0.928   \n",
       "20                            1          0.133          0.130         0.737   \n",
       "21                            1          0.153          0.000         0.847   \n",
       "22                            1          0.109          0.076         0.815   \n",
       "23                            1          0.157          0.067         0.776   \n",
       "24                            1          0.131          0.054         0.814   \n",
       "25                            1          0.215          0.040         0.746   \n",
       "26                            1          0.049          0.055         0.896   \n",
       "27                            1          0.092          0.124         0.784   \n",
       "28                            1          0.077          0.093         0.831   \n",
       "29                            1          0.161          0.133         0.705   \n",
       "..                          ...            ...            ...           ...   \n",
       "169                           1          0.071          0.214         0.714   \n",
       "170                           1          0.524          0.000         0.476   \n",
       "171                           1          0.524          0.000         0.476   \n",
       "172                           1          0.167          0.022         0.812   \n",
       "173                           1          0.165          0.139         0.695   \n",
       "174                           1          0.095          0.000         0.905   \n",
       "175                           1          0.688          0.000         0.312   \n",
       "176                           1          0.159          0.000         0.841   \n",
       "177                           1          0.328          0.000         0.672   \n",
       "178                           1          0.524          0.000         0.476   \n",
       "179                           1          0.111          0.019         0.870   \n",
       "180                           1          0.164          0.037         0.799   \n",
       "181                           1          0.106          0.073         0.821   \n",
       "182                           1          0.207          0.054         0.739   \n",
       "183                           1          0.274          0.048         0.678   \n",
       "184                           1          0.343          0.118         0.539   \n",
       "185                           1          0.254          0.036         0.710   \n",
       "186                           1          0.161          0.000         0.839   \n",
       "187                           1          0.183          0.317         0.500   \n",
       "188                           1          0.355          0.000         0.645   \n",
       "189                           1          0.328          0.000         0.672   \n",
       "190                           1          0.328          0.000         0.672   \n",
       "191                           1          0.096          0.121         0.783   \n",
       "192                           1          0.524          0.000         0.476   \n",
       "193                           1          0.524          0.000         0.476   \n",
       "194                           1          0.524          0.000         0.476   \n",
       "195                           1          0.524          0.000         0.476   \n",
       "196                           1          0.688          0.000         0.312   \n",
       "197                           1          0.062          0.000         0.938   \n",
       "198                           1          0.524          0.000         0.476   \n",
       "\n",
       "0    cancer  \n",
       "0         0  \n",
       "1         0  \n",
       "2         0  \n",
       "3         0  \n",
       "4         0  \n",
       "5         0  \n",
       "6         0  \n",
       "7         0  \n",
       "8         0  \n",
       "9         0  \n",
       "10        0  \n",
       "11        0  \n",
       "12        0  \n",
       "13        0  \n",
       "14        0  \n",
       "15        0  \n",
       "16        0  \n",
       "17        1  \n",
       "18        0  \n",
       "19        0  \n",
       "20        0  \n",
       "21        0  \n",
       "22        0  \n",
       "23        0  \n",
       "24        0  \n",
       "25        0  \n",
       "26        0  \n",
       "27        0  \n",
       "28        0  \n",
       "29        0  \n",
       "..      ...  \n",
       "169       0  \n",
       "170       1  \n",
       "171       1  \n",
       "172       0  \n",
       "173       0  \n",
       "174       0  \n",
       "175       1  \n",
       "176       0  \n",
       "177       0  \n",
       "178       1  \n",
       "179       0  \n",
       "180       0  \n",
       "181       0  \n",
       "182       0  \n",
       "183       0  \n",
       "184       0  \n",
       "185       0  \n",
       "186       0  \n",
       "187       0  \n",
       "188       0  \n",
       "189       0  \n",
       "190       0  \n",
       "191       0  \n",
       "192       1  \n",
       "193       1  \n",
       "194       1  \n",
       "195       1  \n",
       "196       1  \n",
       "197       0  \n",
       "198       1  \n",
       "\n",
       "[199 rows x 11 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cancer_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Evaluation of the method. I am going to annotated which ones are true and which ones not and compare against the reality. \n",
    "# Given that I do have only 200 samples.\n",
    "cancer_df.to_csv('cancer_pred_sent.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "cancer_ground_truth = pd.read_csv('cancer_ground_annotator.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "# F1 score of predicting with sentiment analysis the actual cancer label\n",
    "f1 = float(cancer_ground_truth['TP'].sum()) / (float(cancer_ground_truth['TP'].sum()) + 0.5 * (cancer_ground_truth['FP'].sum() + cancer_ground_truth['FN'].sum()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.25333333333333335"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Prediting the cancer label with natural language processing was not that useful. I leave as labels the ones assigned by me.\n",
    "f1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Data Preprocessing: Working with the graph"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this part of the process, I am going to work with the data available, to build a network and extract properties out of it. There are three different ways to perform the task. In this assignment, I will address all of them to produce a final dataset, where I will have a protein with information about the amount of neighbors, how much information it sends to other proteins, how much information it receives from other proteins, special properties from the nodes (closeness, betweenness and centrality degree) and properties from the transmission action itself (neighborhood, neighborhood_transferred, fusion, cooccurence, homology, coexpression, coexpression_transferred, experiments, experiments_transferred, database, database_transferred, textmining, textmining_transferred, combined_score), which were obtained from the network description dataset. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The first thing that I am going to do is to, define data frames which contain the nodes and the edges."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "nodes = list(np.array(proteinCancerFinal['protein_external_id']))\n",
    "edges = list(tuple(np.array(pd.concat([proteinNetwork['protein1'], proteinNetwork['protein2']], axis=1, sort = False))))\n",
    "print(type(proteinCancerFinal['protein_external_id']))\n",
    "pd.DataFrame(proteinCancerFinal['protein_external_id']).to_csv('nodes.csv')\n",
    "pd.concat([proteinNetwork['protein1'], proteinNetwork['protein2']], axis=1, sort = False).to_csv(\"edges.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Working with the Graph (Part 1): NetworkX Library and GraphFrames"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I used first NetworkX library to extract properties from the network. I created a graph with the nodes and edges generated above, and I wanted to extract specific information from the network such as:\n",
    "- Number of neighbors per node.\n",
    "- Number of outDegrees per node (number of times that information is sent to another node).\n",
    "- Number of inDegrees (number of times that information is received from another node).\n",
    "- Centrality degree (location of the node within the network).\n",
    "- Special properties from a node for a given network ([betweenness](https://en.m.wikipedia.org/wiki/Betweenness_cetrality) and [closeness](https://neo4j.com/docs/graph-algorithms/current/algorithms/closeness-centrality/))\n",
    "\n",
    "From the previous information, I inferred that the nodes in the network were around 20,000 and the edges (amount of information transmitted) were 11,000,000."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### My experience with NetworkX"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I used my laptop to extract information from the network at first instant. However, due to the graph extension and the software architecture of the NetworkX library, turned out to be too complex to extract the desire features (centrality degree, closeness and betweenness).\n",
    "\n",
    "In my studies at TU Delft I learned to use AWS. Thus, I decided to run the code in a AWS Jupyter Notebook with an EMR cluster. For the AWS configuration I used the S3 Bucket, an r3.xlarge instance (good price vs computational power) and a Jupyter Notebook (AWS). My findings were quite negative. Unfortunately, even in the cluster the properties were not extracted. I concluded that, even if the python library is called from a cluster, the computational cost of finding out the properties of each node, in a graph of around 20,000 nodes and 11 million edges turns out to be too complex. The involved library architecture does not support to work with too complex datasets, I will describe NetworkX as a good library for small datasets. The option of using GraphX from spark (well known for big data) does not support the graph property extraction such as closeness, betweenness etc. Thus, I decided to proceed with using pyspark and aggregating the edge data per node with map and reduce operations, using the network dataset (see [Working with the Graph (part2)]). From my personal point of view, finding the properties of the network and the data aggregation from the edge information will be perfect to strongly predict cancer. However, due to the current technology. I ended up using NetworkX only to extract the amount of neighbors per node."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# #TO DO IDENTIFY CLIQUES TO MAKE EFFICIENT THE COMPUTATION OF METRICS AND CHECK WHETHER A CLIQUE MIGHT BE A STRONG INDICATOR OF CANCER. GIVEN THAT I AM IN A NEIGHBORHOOD WITH CANCER, IT IS MORE PROBABLE THAT I DO HAVE CANCER. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The next cell represents the code on the aws Jupyter Notebook previously configured to work with the EMR cluster. I am going to get the list of proteins which are related directly in a negative way to cancer. I want to map those relationshipt that might be \"toxic\" in such way that in my protein network those proteins in contact with \"cancer proteins\" already have cancer in such way that we will be able to identify a cancer relationship. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [],
   "source": [
    "cancer = pd.read_csv('cancer.csv',header='infer')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [],
   "source": [
    "cancer = cancer[['protein_external_id','negative']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [],
   "source": [
    "cancer = cancer[cancer['negative'] == -1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [],
   "source": [
    "cancer['cancer'] = 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [],
   "source": [
    "cancer = cancer[['protein_external_id','cancer']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [],
   "source": [
    "cancer.columns = ['protein1','cancer']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['protein1', 'protein2', 'neighborhood', 'fusion', 'cooccurence',\n",
       "       'coexpression', 'experimental', 'database', 'textmining',\n",
       "       'combined_score'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 96,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "proteinNetwork.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [],
   "source": [
    "proteinNetworkCancer = pd.merge(proteinNetwork, cancer, on='protein1', how='left')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [],
   "source": [
    "cancer2 = cancer[['protein1','cancer']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [],
   "source": [
    "cancer2['protein2'] = cancer2['protein1']\n",
    "cancer2['cancer2'] = cancer2['cancer']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [],
   "source": [
    "#I have to do in this dirty way because the column renaming was crashing the dataframe. Sorry!\n",
    "cancer2 = cancer2[['protein2','cancer2']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein1</th>\n",
       "      <th>protein2</th>\n",
       "      <th>neighborhood</th>\n",
       "      <th>fusion</th>\n",
       "      <th>cooccurence</th>\n",
       "      <th>coexpression</th>\n",
       "      <th>experimental</th>\n",
       "      <th>database</th>\n",
       "      <th>textmining</th>\n",
       "      <th>combined_score</th>\n",
       "      <th>cancer</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9606.ENSP00000000233</td>\n",
       "      <td>9606.ENSP00000272298</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>332</td>\n",
       "      <td>62</td>\n",
       "      <td>181</td>\n",
       "      <td>0</td>\n",
       "      <td>125</td>\n",
       "      <td>490</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>9606.ENSP00000000233</td>\n",
       "      <td>9606.ENSP00000253401</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>186</td>\n",
       "      <td>0</td>\n",
       "      <td>56</td>\n",
       "      <td>198</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>9606.ENSP00000000233</td>\n",
       "      <td>9606.ENSP00000401445</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>159</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>159</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>9606.ENSP00000000233</td>\n",
       "      <td>9606.ENSP00000418915</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>61</td>\n",
       "      <td>158</td>\n",
       "      <td>0</td>\n",
       "      <td>542</td>\n",
       "      <td>606</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>9606.ENSP00000000233</td>\n",
       "      <td>9606.ENSP00000327801</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>88</td>\n",
       "      <td>78</td>\n",
       "      <td>0</td>\n",
       "      <td>89</td>\n",
       "      <td>167</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11759449</th>\n",
       "      <td>9606.ENSP00000485678</td>\n",
       "      <td>9606.ENSP00000310488</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>167</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>167</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11759450</th>\n",
       "      <td>9606.ENSP00000485678</td>\n",
       "      <td>9606.ENSP00000342448</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>175</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>175</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11759451</th>\n",
       "      <td>9606.ENSP00000485678</td>\n",
       "      <td>9606.ENSP00000350222</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>195</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>195</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11759452</th>\n",
       "      <td>9606.ENSP00000485678</td>\n",
       "      <td>9606.ENSP00000367590</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>900</td>\n",
       "      <td>0</td>\n",
       "      <td>900</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11759453</th>\n",
       "      <td>9606.ENSP00000485678</td>\n",
       "      <td>9606.ENSP00000349930</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>213</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>213</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>11759454 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      protein1              protein2  neighborhood  fusion  \\\n",
       "0         9606.ENSP00000000233  9606.ENSP00000272298             0       0   \n",
       "1         9606.ENSP00000000233  9606.ENSP00000253401             0       0   \n",
       "2         9606.ENSP00000000233  9606.ENSP00000401445             0       0   \n",
       "3         9606.ENSP00000000233  9606.ENSP00000418915             0       0   \n",
       "4         9606.ENSP00000000233  9606.ENSP00000327801             0       0   \n",
       "...                        ...                   ...           ...     ...   \n",
       "11759449  9606.ENSP00000485678  9606.ENSP00000310488             0       0   \n",
       "11759450  9606.ENSP00000485678  9606.ENSP00000342448             0       0   \n",
       "11759451  9606.ENSP00000485678  9606.ENSP00000350222             0       0   \n",
       "11759452  9606.ENSP00000485678  9606.ENSP00000367590             0       0   \n",
       "11759453  9606.ENSP00000485678  9606.ENSP00000349930             0       0   \n",
       "\n",
       "          cooccurence  coexpression  experimental  database  textmining  \\\n",
       "0                 332            62           181         0         125   \n",
       "1                   0             0           186         0          56   \n",
       "2                   0             0           159         0           0   \n",
       "3                   0            61           158         0         542   \n",
       "4                   0            88            78         0          89   \n",
       "...               ...           ...           ...       ...         ...   \n",
       "11759449            0           167             0         0           0   \n",
       "11759450            0           175             0         0           0   \n",
       "11759451            0           195             0         0           0   \n",
       "11759452            0             0             0       900           0   \n",
       "11759453            0           213             0         0           0   \n",
       "\n",
       "          combined_score  cancer  \n",
       "0                    490     NaN  \n",
       "1                    198     NaN  \n",
       "2                    159     NaN  \n",
       "3                    606     NaN  \n",
       "4                    167     NaN  \n",
       "...                  ...     ...  \n",
       "11759449             167     NaN  \n",
       "11759450             175     NaN  \n",
       "11759451             195     NaN  \n",
       "11759452             900     NaN  \n",
       "11759453             213     NaN  \n",
       "\n",
       "[11759454 rows x 11 columns]"
      ]
     },
     "execution_count": 101,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "proteinNetworkCancer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [],
   "source": [
    "proteinNetworkCancer = pd.merge(proteinNetworkCancer,cancer2, on='protein2', how='left')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [],
   "source": [
    "proteinNetworkCancer['cancer_final'] = (proteinNetworkCancer['cancer'] + proteinNetworkCancer['cancer2']).fillna(0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [],
   "source": [
    "def input_cancer_val(val):\n",
    "    check = 0\n",
    "    if val >= 1:\n",
    "        check = 1\n",
    "    else:\n",
    "        check = 0\n",
    "    return check\n",
    "proteinNetworkCancer['cancer_final'] = list(map(input_cancer_val,proteinNetworkCancer['cancer_final']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein1</th>\n",
       "      <th>protein2</th>\n",
       "      <th>neighborhood</th>\n",
       "      <th>fusion</th>\n",
       "      <th>cooccurence</th>\n",
       "      <th>coexpression</th>\n",
       "      <th>experimental</th>\n",
       "      <th>database</th>\n",
       "      <th>textmining</th>\n",
       "      <th>combined_score</th>\n",
       "      <th>cancer_final</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9606.ENSP00000000233</td>\n",
       "      <td>9606.ENSP00000272298</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>332</td>\n",
       "      <td>62</td>\n",
       "      <td>181</td>\n",
       "      <td>0</td>\n",
       "      <td>125</td>\n",
       "      <td>490</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>9606.ENSP00000000233</td>\n",
       "      <td>9606.ENSP00000253401</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>186</td>\n",
       "      <td>0</td>\n",
       "      <td>56</td>\n",
       "      <td>198</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>9606.ENSP00000000233</td>\n",
       "      <td>9606.ENSP00000401445</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>159</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>159</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>9606.ENSP00000000233</td>\n",
       "      <td>9606.ENSP00000418915</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>61</td>\n",
       "      <td>158</td>\n",
       "      <td>0</td>\n",
       "      <td>542</td>\n",
       "      <td>606</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>9606.ENSP00000000233</td>\n",
       "      <td>9606.ENSP00000327801</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>88</td>\n",
       "      <td>78</td>\n",
       "      <td>0</td>\n",
       "      <td>89</td>\n",
       "      <td>167</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11759449</th>\n",
       "      <td>9606.ENSP00000485678</td>\n",
       "      <td>9606.ENSP00000310488</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>167</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>167</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11759450</th>\n",
       "      <td>9606.ENSP00000485678</td>\n",
       "      <td>9606.ENSP00000342448</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>175</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>175</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11759451</th>\n",
       "      <td>9606.ENSP00000485678</td>\n",
       "      <td>9606.ENSP00000350222</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>195</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>195</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11759452</th>\n",
       "      <td>9606.ENSP00000485678</td>\n",
       "      <td>9606.ENSP00000367590</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>900</td>\n",
       "      <td>0</td>\n",
       "      <td>900</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11759453</th>\n",
       "      <td>9606.ENSP00000485678</td>\n",
       "      <td>9606.ENSP00000349930</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>213</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>213</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>11759454 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      protein1              protein2  neighborhood  fusion  \\\n",
       "0         9606.ENSP00000000233  9606.ENSP00000272298             0       0   \n",
       "1         9606.ENSP00000000233  9606.ENSP00000253401             0       0   \n",
       "2         9606.ENSP00000000233  9606.ENSP00000401445             0       0   \n",
       "3         9606.ENSP00000000233  9606.ENSP00000418915             0       0   \n",
       "4         9606.ENSP00000000233  9606.ENSP00000327801             0       0   \n",
       "...                        ...                   ...           ...     ...   \n",
       "11759449  9606.ENSP00000485678  9606.ENSP00000310488             0       0   \n",
       "11759450  9606.ENSP00000485678  9606.ENSP00000342448             0       0   \n",
       "11759451  9606.ENSP00000485678  9606.ENSP00000350222             0       0   \n",
       "11759452  9606.ENSP00000485678  9606.ENSP00000367590             0       0   \n",
       "11759453  9606.ENSP00000485678  9606.ENSP00000349930             0       0   \n",
       "\n",
       "          cooccurence  coexpression  experimental  database  textmining  \\\n",
       "0                 332            62           181         0         125   \n",
       "1                   0             0           186         0          56   \n",
       "2                   0             0           159         0           0   \n",
       "3                   0            61           158         0         542   \n",
       "4                   0            88            78         0          89   \n",
       "...               ...           ...           ...       ...         ...   \n",
       "11759449            0           167             0         0           0   \n",
       "11759450            0           175             0         0           0   \n",
       "11759451            0           195             0         0           0   \n",
       "11759452            0             0             0       900           0   \n",
       "11759453            0           213             0         0           0   \n",
       "\n",
       "          combined_score  cancer_final  \n",
       "0                    490             0  \n",
       "1                    198             0  \n",
       "2                    159             0  \n",
       "3                    606             0  \n",
       "4                    167             0  \n",
       "...                  ...           ...  \n",
       "11759449             167             0  \n",
       "11759450             175             0  \n",
       "11759451             195             0  \n",
       "11759452             900             0  \n",
       "11759453             213             0  \n",
       "\n",
       "[11759454 rows x 11 columns]"
      ]
     },
     "execution_count": 105,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "proteinNetworkCancer.drop(['cancer','cancer2'], axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein1</th>\n",
       "      <th>protein2</th>\n",
       "      <th>neighborhood</th>\n",
       "      <th>fusion</th>\n",
       "      <th>cooccurence</th>\n",
       "      <th>coexpression</th>\n",
       "      <th>experimental</th>\n",
       "      <th>database</th>\n",
       "      <th>textmining</th>\n",
       "      <th>combined_score</th>\n",
       "      <th>cancer</th>\n",
       "      <th>cancer2</th>\n",
       "      <th>cancer_final</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>110962</th>\n",
       "      <td>9606.ENSP00000167106</td>\n",
       "      <td>9606.ENSP00000383185</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>232</td>\n",
       "      <td>232</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110971</th>\n",
       "      <td>9606.ENSP00000167106</td>\n",
       "      <td>9606.ENSP00000282397</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>289</td>\n",
       "      <td>289</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>367033</th>\n",
       "      <td>9606.ENSP00000219022</td>\n",
       "      <td>9606.ENSP00000309572</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>187</td>\n",
       "      <td>187</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>367408</th>\n",
       "      <td>9606.ENSP00000219022</td>\n",
       "      <td>9606.ENSP00000406861</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>212</td>\n",
       "      <td>212</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>426058</th>\n",
       "      <td>9606.ENSP00000220812</td>\n",
       "      <td>9606.ENSP00000233948</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>93</td>\n",
       "      <td>0</td>\n",
       "      <td>641</td>\n",
       "      <td>660</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10750844</th>\n",
       "      <td>9606.ENSP00000442521</td>\n",
       "      <td>9606.ENSP00000379133</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>259</td>\n",
       "      <td>259</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10750907</th>\n",
       "      <td>9606.ENSP00000442521</td>\n",
       "      <td>9606.ENSP00000379762</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>61</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>233</td>\n",
       "      <td>249</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11297763</th>\n",
       "      <td>9606.ENSP00000473553</td>\n",
       "      <td>9606.ENSP00000287394</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>64</td>\n",
       "      <td>123</td>\n",
       "      <td>0</td>\n",
       "      <td>50</td>\n",
       "      <td>152</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11297765</th>\n",
       "      <td>9606.ENSP00000473553</td>\n",
       "      <td>9606.ENSP00000334061</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>63</td>\n",
       "      <td>157</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>176</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11298005</th>\n",
       "      <td>9606.ENSP00000473553</td>\n",
       "      <td>9606.ENSP00000264606</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>63</td>\n",
       "      <td>157</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>176</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>192 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      protein1              protein2  neighborhood  fusion  \\\n",
       "110962    9606.ENSP00000167106  9606.ENSP00000383185             0       0   \n",
       "110971    9606.ENSP00000167106  9606.ENSP00000282397             0       0   \n",
       "367033    9606.ENSP00000219022  9606.ENSP00000309572             0       0   \n",
       "367408    9606.ENSP00000219022  9606.ENSP00000406861             0       0   \n",
       "426058    9606.ENSP00000220812  9606.ENSP00000233948             0       0   \n",
       "...                        ...                   ...           ...     ...   \n",
       "10750844  9606.ENSP00000442521  9606.ENSP00000379133             0       0   \n",
       "10750907  9606.ENSP00000442521  9606.ENSP00000379762             0       0   \n",
       "11297763  9606.ENSP00000473553  9606.ENSP00000287394             0       0   \n",
       "11297765  9606.ENSP00000473553  9606.ENSP00000334061             0       0   \n",
       "11298005  9606.ENSP00000473553  9606.ENSP00000264606             0       0   \n",
       "\n",
       "          cooccurence  coexpression  experimental  database  textmining  \\\n",
       "110962              0             0             0         0         232   \n",
       "110971              0             0             0         0         289   \n",
       "367033              0             0             0         0         187   \n",
       "367408              0             0             0         0         212   \n",
       "426058              0             0            93         0         641   \n",
       "...               ...           ...           ...       ...         ...   \n",
       "10750844            0             0             0         0         259   \n",
       "10750907            0            61             0         0         233   \n",
       "11297763            0            64           123         0          50   \n",
       "11297765            0            63           157         0           0   \n",
       "11298005            0            63           157         0           0   \n",
       "\n",
       "          combined_score  cancer  cancer2  cancer_final  \n",
       "110962               232     1.0      1.0             1  \n",
       "110971               289     1.0      1.0             1  \n",
       "367033               187     1.0      1.0             1  \n",
       "367408               212     1.0      1.0             1  \n",
       "426058               660     1.0      1.0             1  \n",
       "...                  ...     ...      ...           ...  \n",
       "10750844             259     1.0      1.0             1  \n",
       "10750907             249     1.0      1.0             1  \n",
       "11297763             152     1.0      1.0             1  \n",
       "11297765             176     1.0      1.0             1  \n",
       "11298005             176     1.0      1.0             1  \n",
       "\n",
       "[192 rows x 13 columns]"
      ]
     },
     "execution_count": 106,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "proteinNetworkCancer[proteinNetworkCancer['cancer_final'] == 1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein1</th>\n",
       "      <th>protein2</th>\n",
       "      <th>neighborhood</th>\n",
       "      <th>fusion</th>\n",
       "      <th>cooccurence</th>\n",
       "      <th>coexpression</th>\n",
       "      <th>experimental</th>\n",
       "      <th>database</th>\n",
       "      <th>textmining</th>\n",
       "      <th>combined_score</th>\n",
       "      <th>cancer_final</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9606.ENSP00000000233</td>\n",
       "      <td>9606.ENSP00000272298</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>332</td>\n",
       "      <td>62</td>\n",
       "      <td>181</td>\n",
       "      <td>0</td>\n",
       "      <td>125</td>\n",
       "      <td>490</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>9606.ENSP00000000233</td>\n",
       "      <td>9606.ENSP00000253401</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>186</td>\n",
       "      <td>0</td>\n",
       "      <td>56</td>\n",
       "      <td>198</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>9606.ENSP00000000233</td>\n",
       "      <td>9606.ENSP00000401445</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>159</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>159</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>9606.ENSP00000000233</td>\n",
       "      <td>9606.ENSP00000418915</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>61</td>\n",
       "      <td>158</td>\n",
       "      <td>0</td>\n",
       "      <td>542</td>\n",
       "      <td>606</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>9606.ENSP00000000233</td>\n",
       "      <td>9606.ENSP00000327801</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>88</td>\n",
       "      <td>78</td>\n",
       "      <td>0</td>\n",
       "      <td>89</td>\n",
       "      <td>167</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11759449</th>\n",
       "      <td>9606.ENSP00000485678</td>\n",
       "      <td>9606.ENSP00000310488</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>167</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>167</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11759450</th>\n",
       "      <td>9606.ENSP00000485678</td>\n",
       "      <td>9606.ENSP00000342448</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>175</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>175</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11759451</th>\n",
       "      <td>9606.ENSP00000485678</td>\n",
       "      <td>9606.ENSP00000350222</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>195</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>195</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11759452</th>\n",
       "      <td>9606.ENSP00000485678</td>\n",
       "      <td>9606.ENSP00000367590</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>900</td>\n",
       "      <td>0</td>\n",
       "      <td>900</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11759453</th>\n",
       "      <td>9606.ENSP00000485678</td>\n",
       "      <td>9606.ENSP00000349930</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>213</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>213</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>11759454 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      protein1              protein2  neighborhood  fusion  \\\n",
       "0         9606.ENSP00000000233  9606.ENSP00000272298             0       0   \n",
       "1         9606.ENSP00000000233  9606.ENSP00000253401             0       0   \n",
       "2         9606.ENSP00000000233  9606.ENSP00000401445             0       0   \n",
       "3         9606.ENSP00000000233  9606.ENSP00000418915             0       0   \n",
       "4         9606.ENSP00000000233  9606.ENSP00000327801             0       0   \n",
       "...                        ...                   ...           ...     ...   \n",
       "11759449  9606.ENSP00000485678  9606.ENSP00000310488             0       0   \n",
       "11759450  9606.ENSP00000485678  9606.ENSP00000342448             0       0   \n",
       "11759451  9606.ENSP00000485678  9606.ENSP00000350222             0       0   \n",
       "11759452  9606.ENSP00000485678  9606.ENSP00000367590             0       0   \n",
       "11759453  9606.ENSP00000485678  9606.ENSP00000349930             0       0   \n",
       "\n",
       "          cooccurence  coexpression  experimental  database  textmining  \\\n",
       "0                 332            62           181         0         125   \n",
       "1                   0             0           186         0          56   \n",
       "2                   0             0           159         0           0   \n",
       "3                   0            61           158         0         542   \n",
       "4                   0            88            78         0          89   \n",
       "...               ...           ...           ...       ...         ...   \n",
       "11759449            0           167             0         0           0   \n",
       "11759450            0           175             0         0           0   \n",
       "11759451            0           195             0         0           0   \n",
       "11759452            0             0             0       900           0   \n",
       "11759453            0           213             0         0           0   \n",
       "\n",
       "          combined_score  cancer_final  \n",
       "0                    490             0  \n",
       "1                    198             0  \n",
       "2                    159             0  \n",
       "3                    606             0  \n",
       "4                    167             0  \n",
       "...                  ...           ...  \n",
       "11759449             167             0  \n",
       "11759450             175             0  \n",
       "11759451             195             0  \n",
       "11759452             900             0  \n",
       "11759453             213             0  \n",
       "\n",
       "[11759454 rows x 11 columns]"
      ]
     },
     "execution_count": 107,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "proteinNetworkCancer.drop(['cancer','cancer2'], axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "unexpected EOF while parsing (<ipython-input-108-0c2e71e44391>, line 7)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  File \u001b[0;32m\"<ipython-input-108-0c2e71e44391>\"\u001b[0;36m, line \u001b[0;32m7\u001b[0m\n\u001b[0;31m    nodes = list(pd.read_csv(nodes_data).iloc[:,1]\u001b[0m\n\u001b[0m                                                  ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m unexpected EOF while parsing\n"
     ]
    }
   ],
   "source": [
    "bucket='networkanalysis'\n",
    "nodes_data = 'nodes.csv'\n",
    "edges_data = 'edges.csv'\n",
    "nodes_data = 's3://{}/{}'.format(bucket, nodes_data)\n",
    "edges_data = 's3://{}/{}'.format(bucket, edges_data)\n",
    "edges = list(tuple(np.array(pd.read_csv(edges_data).iloc[:,1:])))\n",
    "nodes = list(pd.read_csv(nodes_data).iloc[:,1]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Build the graph with networks as nx."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Create graph object. \n",
    "G = nx.MultiDiGraph()\n",
    "#getting list of nodes\n",
    "nodes = list(set(proteinNetworkCancer['protein1'].append(proteinNetworkCancer['protein2'])))\n",
    "#getting list of edgdesmosomees\n",
    "edges = list(zip(proteinNetworkCancer.protein1, proteinNetworkCancer.protein2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [],
   "source": [
    "G.add_nodes_from(nodes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " ...]"
      ]
     },
     "execution_count": 111,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "G.add_edges_from(edges)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of Nodes 19354\n",
      "Number of Directed Edges: 11759454\n"
     ]
    }
   ],
   "source": [
    "#print graph properties\n",
    "print(\"Number of Nodes \" + str(G.number_of_nodes()))\n",
    "print(\"Number of Directed Edges: \" + str(G.number_of_edges()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [],
   "source": [
    "#the amount of in_degree, out_degree and neighbors is the same, because this graph is an undirected graph.\n",
    "#I want to extract the amount of neighbors in the column protein1 to add the neighbors per relationship on the graph.\n",
    "#I did not proceed on getting the amount of neighbors on the set of nodes and then doing a join because\n",
    "#it was too expensive to compute and the computer could not perform the left join. \n",
    "#I divided my data in such ugly way because the np.split() method works for equally divided arrays. \n",
    "#It was necessary to work by chunks of information otherwise I could not compute. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_indegree_node(node):\n",
    "    return G.in_degree(node)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {},
   "outputs": [],
   "source": [
    "#I need to use the ugly loop because otherwise is too heavy.\n",
    "neighbors = []\n",
    "for item in proteinNetworkCancer['protein1']:\n",
    "    neighbors.append(get_indegree_node(item))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_indegree_node(node):\n",
    "    return G.in_degree(node)\n",
    "def get_outdegree_nodes(node):\n",
    "    return G.out_degree(node)\n",
    "def get_node_neighbors(node):\n",
    "    return len(list(G.neighbors(node)))\n",
    "\n",
    "nodes_indegree = list(map(get_indegree_node,nodes))\n",
    "nodes_outdegree = list(map(get_outdegree_nodes,nodes))\n",
    "nodes_neighbors = list(map(get_node_neighbors,nodes))\n",
    "# Run the below commands in aws and my personal computer for several hours. It never finished.\n",
    "# Network size 20000 nodes and 11 million edges. \n",
    "    \n",
    "#closenessDict = nx.closeness_centrality(G)\n",
    "#betweennessDict = nx.betweenness_centrality(G)\n",
    "#centralityDegreeDict = nx.degree_centrality(G)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {},
   "outputs": [],
   "source": [
    "#this is only to proof what I wrote above\n",
    "nodes_information = pd.DataFrame({'node_id':nodes,'indegree':nodes_indegree,'outdegree':nodes_outdegree,'nodes_neighbors':nodes_neighbors})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>node_id</th>\n",
       "      <th>indegree</th>\n",
       "      <th>outdegree</th>\n",
       "      <th>nodes_neighbors</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9606.ENSP00000386226</td>\n",
       "      <td>1441</td>\n",
       "      <td>1441</td>\n",
       "      <td>1441</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>9606.ENSP00000217652</td>\n",
       "      <td>1389</td>\n",
       "      <td>1389</td>\n",
       "      <td>1389</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>9606.ENSP00000317895</td>\n",
       "      <td>237</td>\n",
       "      <td>237</td>\n",
       "      <td>237</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>9606.ENSP00000360149</td>\n",
       "      <td>326</td>\n",
       "      <td>326</td>\n",
       "      <td>326</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>9606.ENSP00000345968</td>\n",
       "      <td>348</td>\n",
       "      <td>348</td>\n",
       "      <td>348</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19349</th>\n",
       "      <td>9606.ENSP00000354583</td>\n",
       "      <td>156</td>\n",
       "      <td>156</td>\n",
       "      <td>156</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19350</th>\n",
       "      <td>9606.ENSP00000344468</td>\n",
       "      <td>526</td>\n",
       "      <td>526</td>\n",
       "      <td>526</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19351</th>\n",
       "      <td>9606.ENSP00000346733</td>\n",
       "      <td>738</td>\n",
       "      <td>738</td>\n",
       "      <td>738</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19352</th>\n",
       "      <td>9606.ENSP00000222224</td>\n",
       "      <td>127</td>\n",
       "      <td>127</td>\n",
       "      <td>127</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19353</th>\n",
       "      <td>9606.ENSP00000379773</td>\n",
       "      <td>206</td>\n",
       "      <td>206</td>\n",
       "      <td>206</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19354 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    node_id  indegree  outdegree  nodes_neighbors\n",
       "0      9606.ENSP00000386226      1441       1441             1441\n",
       "1      9606.ENSP00000217652      1389       1389             1389\n",
       "2      9606.ENSP00000317895       237        237              237\n",
       "3      9606.ENSP00000360149       326        326              326\n",
       "4      9606.ENSP00000345968       348        348              348\n",
       "...                     ...       ...        ...              ...\n",
       "19349  9606.ENSP00000354583       156        156              156\n",
       "19350  9606.ENSP00000344468       526        526              526\n",
       "19351  9606.ENSP00000346733       738        738              738\n",
       "19352  9606.ENSP00000222224       127        127              127\n",
       "19353  9606.ENSP00000379773       206        206              206\n",
       "\n",
       "[19354 rows x 4 columns]"
      ]
     },
     "execution_count": 118,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nodes_information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "11759454"
      ]
     },
     "execution_count": 119,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(neighbors)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "BY the numbers obtained above we can infer easly that proteins in a network behave as an undirect graph or the dataset is too synthetic. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Given the computational cost I only extract the amount of neighbors."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### My Experience with GraphFrames (Pyspark)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I still wanted to have more network properties to train the machine learning algorithm. Thus, I found out the [ GraphFrames](https://graphframes.github.io/graphframes/docs/_site/user-guide.html#triangle-count) library, which works well with SQL-Pyspark. I used the library to extract the outDegrees and inDegrees and the number of triangles passing to each node. I can describe the library to be quite handy to process bigger files, but still they do not offer the previous mentioned properties (Betweenness, Closenness and Centrality Degree). The reason for that is the complexity involved in computing them. Overall, I extracted a dataset where I had a protein id with outDegrees, inDegrees and the triangle count(pending). "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from graphframes import * #follow the documentation to open pyspark-shell with graphframes\n",
    "edges = sqlContext.read.format(\"csv\").option(\"header\",\"true\").option(\"inferschema\",\"true\").load(\"/home/abi/Documents/Network Analysis/edges.csv\")\n",
    "nodes = sqlContext.read.format(\"csv\").option(\"header\",\"true\").option(\"inferschema\",\"true\").load(\"/home/abi/Documents/Network Analysis/nodes.csv\")\n",
    "#data cleaning\n",
    "newedges = edges[[\"protein1\",\"protein2\"]]\n",
    "newnodes = nodes[[\"protein_external_id\"]]\n",
    "#change newnodes header to \"id\" matter of the library graphframe specification. Read documentation.\n",
    "oldcolumnnodes = newnodes.schema.names\n",
    "newcolumnnodes = ['id']\n",
    "newnodes2 = reduce(lambda newnodes, idx: newnodes.withColumnRenamed(oldcolumnnodes[idx], newcolumnnodes[idx]), xrange(len(oldcolumnnodes)), newnodes)\n",
    "#change edges header to src and dst.\n",
    "oldcolumnedges = newedges.schema.names\n",
    "newcolumnedges = ['src', 'dst']\n",
    "newedges2 = reduce(lambda newedges, idx: newedges.withColumnRenamed(oldcolumnedges[idx], newcolumnedges[idx]), xrange(len(oldcolumnedges)), newedges)\n",
    "g = GraphFrame(newnodes2,newedges2)\n",
    "vertexInDegrees = g.inDegrees\n",
    "vertexOutDegrees = g.outDegrees\n",
    "vertexOutDegrees.toPandas().to_csv(\"vertexOutDegrees.csv\")\n",
    "vertexInDegrees.toPandas().to_csv(\"vertexInDegrees.csv\")\n",
    "triangles = g.triangleCount()\n",
    "triangles.toPandas().to_csv(\"vertexTriangles.csv\")\n",
    "#join Data by key, which is equal to the protein tag.\n",
    "#A_B = A.join(B, A.id == B.id).select(A.*, B.b1, B.b2)\n",
    "joinedDegrees = vertexInDegrees.join(vertexOutDegrees,vertexOutDegrees.id == vertexInDegrees.id)\n",
    "#change schema due to id repetition.\n",
    "oldcolumndegrees = joinedDegrees.schema.names\n",
    "newcolumndegrees = ['id','inDegree','id1','outDegree']\n",
    "joinedDegrees2 = reduce(lambda joinedDegrees, idx: joinedDegrees.withColumnRenamed(oldcolumndegrees[idx], newcolumndegrees[idx]), xrange(len(oldcolumndegrees)), joinedDegrees)\n",
    "#drop id1\n",
    "pandasclean = joinedDegrees2.toPandas()\n",
    "pandas = pandasclean[['id','inDegree','outDegree']]\n",
    "pandas.to_csv('degrees.csv')\n",
    "#THE INDEGREES AND OUT DEGREES TURN OUT TO BE THE SAME. THIS FACT IS A PROOF OF HOW WONDERFUL\n",
    "#AND PERFECT THE ENGINEERING NATURE IS. HOWEVER, WHEN WORKING WITH OTHER TYPES OF NETWORKS THE OUTDEGREES AND\n",
    "#INDEGREES ARE RARELY THE SAME\n",
    "joinTrianglesVertexIn = triangles.join(vertexInDegrees, vertexInDegrees.id == triangles.id)\n",
    "joinTrianglesVertexIn.toPandas().to_csv(\"firstjoin.csv\")\n",
    "secondJoin = rddJoin.join(vertexOutDegrees, vertexOutDegrees.id == rddJoin.id)\n",
    "secondJoin.toPandas().to_csv(\"secondJoin.csv\")\n",
    "#*This join operations were done witth rdds, I am aware that it can be done as well by using join alike SQL operations."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Information Summary\n",
    "\n",
    "Thanks to NetworkX, I extracted the neighbors per node.\n",
    "Thanks to GraphFrames I got the inDegree, outDegree and triangle count for each node.\n",
    "The resulted format of the dataset is: \n",
    "\n",
    "node, number_of_neighbors, outDegree, inDegree, triangle_count."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Working with the Graph (Part 2): Spark Streaming"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The network file was not really big, it had a size of only 1 gb. The problem is that it was too long. Taking into account the hardware side and how registers need a space in a RAM memory to perform operations 11000000 of registers are a problem and cannot be proccessed at once by any small device pay-able by a regular human (Of course until we manage to find out the secrets of the quantum computing, but I will leave that for the future :) ). Thus, I decided to use Spark Streaming. It basically consists in reading your data by chunks of information. In my case, nicely break my long data into small subsets that I can process with spark (each chunk at a time). Below, you can appreciate the code that I created."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "My intention with the code below was to produce an aggregation of the \"9606.protein.links.full.v11.0.txt\"(edge description data) in such way that I will end up producing a dataset including a protein name, plus the sum of each of the mentioned properties per node. In my case, I took into account the information transmitted from protein 1 to protein 2. Thus, I deleted protein 2 column."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from operator import add, sub\n",
    "from time import sleep\n",
    "from pyspark import SparkContext\n",
    "from pyspark.streaming import StreamingContext\n",
    "from os import walk\n",
    "\n",
    "sc = SparkContext(appName=\"streamingProteins\")\n",
    "ssc = StreamingContext(sc,5)\n",
    "#Reshape Data\n",
    "file = open(\"/home/abi/Documents/Network Analysis/9606.protein.links.full.v11.0.txt\", \"r\").readlines()\n",
    "del file[0]\n",
    "def cleanStr(str):\n",
    "    newStr = str.replace(\"\\n\",\"\")\n",
    "    newStr2 = newStr.replace(newStr[21:42],\"\")\n",
    "    return newStr2\n",
    "newFile = map(cleanStr, file)\n",
    "#write Data to new txt file\n",
    "with open('proteinTransmission.txt', 'w') as filehandle:\n",
    "    for item in newFile:\n",
    "        filehandle.write('%s\\n' % item)\n",
    "\n",
    "#create definition to save outputdata\n",
    "def saveResults(rdd):\n",
    "    if not rdd.isEmpty():\n",
    "        rdd.toDF([\"protein1\",\"1\",\"2\",\"3\",\"4\",\"5\",\"6\",\"7\",\"8\",\"9\",\"10\",\"11\",\"12\",\"13\",\"14\"]).write.save(\"rddsprocess\", format=\"csv\", mode=\"append\")\n",
    "#Load proteinTransmission file\n",
    "proteinTransmission = open(\"proteinTransmission.csv\", \"r\").readlines()\n",
    "#create file to store the rdds manually file in the spark folder with name rddsprocess.txt\n",
    "#create rdd array like = queue \n",
    "rddQueue = []\n",
    "# store in queue the information batches\n",
    "for batch in proteinTransmission:\n",
    "    rddQueue += [ssc.sparkContext.parallelize(batch)]\n",
    "\n",
    "inputStream = ssc.queueStream(rddQueue)\n",
    "inputStreamSplit = inputStream.map(lambda x: x.split(\",\"))\n",
    "inputStreamTuple = inputStreamSplit.map(lambda x: (x[0],(x[1],x[2],x[3],x[4],x[5],x[6],x[7],x[8],x[9],x[10],x[11],x[12],x[13],x[14])))\n",
    "inputStreamInt = inputStreamTuple.map(lambda x: (x[0], [int(i) for i in x[1]]))\n",
    "totalStream = inputStreamInt.reduceByKey(lambda a,b: (a[0]+b[0],a[1]+b[1], a[2]+b[2],a[3]+b[3],a[4]+b[4],a[5]+b[5],a[6]+b[6],a[7]+b[7],a[8]+b[8],a[9]+b[9],a[10]+b[10],a[12]+b[12],a[13]+b[13],a[14]+b[14]))\n",
    "totalStream.foreachRDD(saveResults)\n",
    "\n",
    "scc.start()\n",
    "scc.awaitTermination()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Working with the Graph (Part 3): The Rudymentary Way with Pyspark and Map and Reduce Operations"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Working with small computer, I decided to break the file (using os), into several files and perform data aggregation over each file to produce a reduced file. Later, I used all the reduced file to produced a second reduction and finally get the aggregated data of the network."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Define the amount of lines that my chunk files are going to have \n",
    "linesPerFile = 5000\n",
    "smallfile = None\n",
    "with open(\"proteinTransmission.txt\") as file:\n",
    "    for lineno, line in enumerate(file):\n",
    "        if lineno % linesPerFile == 0:\n",
    "            if smallfile:\n",
    "                smallfile.close()\n",
    "            smallName = \"filereduction/smallfile_{}.txt\".format(lineno + linesPerFile)\n",
    "            smallfile = open(smallName, \"w\")\n",
    "        smallfile.write(line)\n",
    "    if smallfile:\n",
    "        smallfile.close()\n",
    "def changestr(str):\n",
    "    newstr = str.replace(\"\\n\",\"\")\n",
    "    return str\n",
    "#iteration over the documents to perform map and reduce operation over each of them and print the reduced rdd to \n",
    "#a newreducefile.\n",
    "breakedData = os.listdir(\"/home/abi/spark/spark-2.4.3-bin-hadoop2.7/filereduction\")\n",
    "for file in breakedData:\n",
    "    workfile = \"filereduction/\" + file\n",
    "    workfilelist = open(workfile, \"r\").readlines()\n",
    "    newworkfilelist = map(changestr, workfilelist)\n",
    "    firstRdd = sc.parallelize(newworkfilelist)\n",
    "    tupleRDD = firstRdd.map(lambda x: x.split(\" \"))\n",
    "    tupleRdd = tupleRDD.map(lambda x: (x[0],(x[1],x[2],x[3],x[4],x[5],x[6],x[7],x[8],x[9],x[10],x[11],x[12],x[13],x[14])))\n",
    "    intrdd = tupleRdd.map(lambda x: (x[0], [int(i) for i in x[1]]))\n",
    "    #Personal comment: I still do not understand why it works with up to 13 instead of 14. But it works that way.\n",
    "    total = intrdd.reduceByKey(lambda a,b: (a[0]+b[0],a[1]+b[1],a[2]+b[2],a[3]+b[3],a[4]+b[4],a[5]+b[5],a[6]+b[6],a[7]+b[7],a[8]+b[8],a[9]+b[9],a[10]+b[10],b[11]+b[11],a[12]+b[12],a[13]+b[13]))\n",
    "    newName = \"filereduction/reduced/\" + \"red-\" + file.replace(\"txt\",\"csv\")\n",
    "    df = spark.createDataFrame(total)\n",
    "    df.toPandas().to_csv(newName)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The output of the previous cell represented the aggregate data of each batch of information. In order to fully aggregate the data, it is necessary to append all the information together and to perform a map and reduce operation again. Each file resulted in a ~20 row key value aggregation."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### More Data Processing to aggregate the chunks of information."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The cell below is to combine all the reduced files into one reduced file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#get the paths from reduced folder\n",
    "reducedData = os.listdir(\"/home/abi/spark/spark-2.4.3-bin-hadoop2.7/filereduction/reduced\")\n",
    "#create manually a file called result.txt contained in the reduced folder\n",
    "#save full paths of reduced documents to a list\n",
    "def updateListPath(pathDoc):\n",
    "    str = \"/home/abi/spark/spark-2.4.3-bin-hadoop2.7/filereduction/reduced/\" + pathDoc\n",
    "    return str\n",
    "#Update path list with the full path, initially it only contains the file names.\n",
    "updated = map(updateListPath,reducedData)\n",
    "\n",
    "#Append all the txt files together into one txt file.\n",
    "with open((\"/home/abi/spark/spark-2.4.3-bin-hadoop2.7/filereduction/reduced/reduced.txt\"),\"w\") as result:\n",
    "    for file in updated:\n",
    "        for line in open(file, \"r\"):\n",
    "            result.write(line)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The next code is to reduce all the information in the file. = More map and reduce."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Data preprocessing again. Now I had to clean the data again because the spark data frame returned the file with another format.\n",
    "#load the resulted file and delete rows with unnecessary strings.\n",
    "reducedDataDf = spark.read.csv(\"/home/abi/spark/spark-2.4.3-bin-hadoop2.7/filereduction/reduced/reduced.txt\")\n",
    "#delete rows with null values and delete column _c0 index from the previous datasets.\n",
    "reducedDfDeleteRow = reducedDataDf.filter(reducedDataDf._c0 != 'null').drop(reducedDataDf._c0)\n",
    "#print the table as csv file to perform text processing over the values.\n",
    "reducedPandas = reducedDfDeleteRow.toPandas()\n",
    "noisyTuples = list(reducedPandas[\"_c2\"])\n",
    "#clean column\n",
    "def cleanTup(str):\n",
    "    del1 = str.replace(\"Row(\",\"\")\n",
    "    del1 = del1.replace(\"_1=\",\"\")\n",
    "    del2 = del1.replace(\"_2=\",\"\")\n",
    "    del3 = del2.replace(\"_3=\",\"\")\n",
    "    del4 = del3.replace(\"_4=\",\"\")\n",
    "    del5 = del4.replace(\"_5=\",\"\")\n",
    "    del6 = del5.replace(\"_6=\",\"\")\n",
    "    del7 = del6.replace(\"_7=\",\"\")\n",
    "    del8 = del7.replace(\"_8=\",\"\")\n",
    "    del9 = del8.replace(\"_9=\",\"\")\n",
    "    del10 = del9.replace(\"_10=\",\"\")\n",
    "    del11 = del10.replace(\"_11=\",\"\")\n",
    "    del12 = del11.replace(\"_12=\",\"\")\n",
    "    del13 = del12.replace(\"_13=\",\"\")\n",
    "    del14 = del13.replace(\"_14=\",\"\")\n",
    "    del15 = del14.replace(\")\",\"\")\n",
    "    del15 = del15.replace(\"u'\",\"\")\n",
    "    return del15\n",
    "\n",
    "clean = map(cleanTup,noisyTuples)\n",
    "protein = list(reducedPandas['_c1'])\n",
    "df = pandas.DataFrame(clean,protein)\n",
    "#print csv ready to perform last step of map and reduce.\n",
    "df.to_csv(\"last_step_mp.csv\", sep=\" \")\n",
    "#load csv to perform last map and reduce operation.\n",
    "#clean data again.\n",
    "file = open(\"last_step_mp.csv\", \"r\").readlines()\n",
    "\n",
    "def cleanstr(str):\n",
    "    str = str.replace(\",\",\"\")\n",
    "    str2 = str.replace('\"',\"\")\n",
    "    str3 = str2.replace(\"\\n\",\"\")\n",
    "    return str3\n",
    "\n",
    "newFile = map(cleanstr,file)\n",
    "del newFile[0]\n",
    "\n",
    "#work with the map and reduce logic.\n",
    "firstRdd = sc.parallelize(newFile)\n",
    "tupleRDD = firstRdd.map(lambda x: x.split(\" \"))\n",
    "tupleRdd = tupleRDD.map(lambda x: (x[0],(x[1],x[2],x[3],x[4],x[5],x[6],x[7],x[8],x[9],x[10],x[11],x[12],x[13],x[14])))\n",
    "intrdd = tupleRdd.map(lambda x: (x[0], [int(i) for i in x[1]]))\n",
    "#Personal comment: I still do not understand why it works with up to 13 instead of 14. But it works that way.\n",
    "total = intrdd.reduceByKey(lambda a,b: (a[0]+b[0],a[1]+b[1],a[2]+b[2],a[3]+b[3],a[4]+b[4],a[5]+b[5],a[6]+b[6],a[7]+b[7],a[8]+b[8],a[9]+b[9],a[10]+b[10],b[11]+b[11],a[12]+b[12],a[13]+b[13]))\n",
    "df = spark.createDataFrame(total).toPandas().to_csv(\"nodesnetaggregation.csv\")\n",
    "\n",
    "#*Data cleaning might be less complicated by using sql alike tables or Tableau itself."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "After performing the map and reduce operations over the values, it was noted that only around 7000 nodes were involved in the network dataset. The result makes sense since all the possible proteins registered with the description are present in the first dataset. Now, I will aggregate the data to link the corresponding network properties with each protein. Overall, the four datasets that I am going to combine are the \"neighbors.csv\", \"degrees.csv\",\"finalData.csv\"(cancer label) and the \"nodesnetaggregation.csv\"."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Aggregating the datasets for the machine learning algorithm (more map and reduce)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#I cleaned manually the files. Erasing unnecessary columns and renaming columns.\n",
    "# Creating a string list of the files\n",
    "degrees = open(\"degrees.csv\", \"r\").readlines()\n",
    "neighbors = open(\"neighbors.csv\",\"r\").readlines()\n",
    "nodesnet = open(\"nodesnetaggregation.csv\").readlines()\n",
    "cancer = open(\"finalData.csv\").readlines()\n",
    "\n",
    "# Erasing \"\\n\" from the list of strings\n",
    "def clean(str):\n",
    "    newstr2 = str.replace(\"\\n\",\"\")\n",
    "    return newstr2\n",
    "\n",
    "degreesClean = map(clean,degrees)\n",
    "neighborsClean = map(clean,neighbors)\n",
    "nodesnetClean = map(clean,nodesnet)\n",
    "cancerClean = map (clean,cancer)\n",
    "del degreesClean[0]\n",
    "del neighborsClean[0]\n",
    "del nodesnetClean[0]\n",
    "del cancerClean[0]\n",
    "# Creating first rdd of strings\n",
    "degreesRdd = sc.parallelize(degreesClean)\n",
    "neighborsRdd = sc.parallelize(neighborsClean)\n",
    "nodesnetRdd = sc.parallelize(nodesnetClean)\n",
    "cancerRdd = sc.parallelize(cancerClean)\n",
    "\n",
    "# Spliting first rdd by \",\"\n",
    "degreesRddSch =  degreesRdd.map(lambda x: x.split(\",\"))\n",
    "neighborsRddSch = neighborsRdd.map(lambda x: x.split(\",\"))\n",
    "nodesnetRddSch = nodesnetRdd.map(lambda x: x.split(\",\"))\n",
    "cancerRddSch = cancerRdd.map(lambda x: x.split(\",\"))\n",
    "\n",
    "# Turn tuple RDD to data frame.\n",
    "degreesTuple = degreesRddSch.map(lambda x: (x[0],(x[1],x[2])))\n",
    "neighborsTuple = neighborsRddSch.map(lambda x: (x[0],x[1]))\n",
    "nodesnetTuple = nodesnetRddSch.map(lambda x: (x[0],(x[1],x[2],x[3],x[4],x[5],x[6],x[7],x[8],x[9],x[10],x[11],x[12],x[13],x[14])))\n",
    "cancerTuple = cancerRddSch.map(lambda x: (x[0],x[1]))\n",
    "# Performing join operations over the RDDs.\n",
    "nodesdeg = nodesnetTuple.join(degreesTuple)\n",
    "#Transform to data frames the rdds.\n",
    "nodesnetDf = nodesnetTuple.toDF()\n",
    "degreesDf = degreesTuple.toDF()\n",
    "neighborsDf = neighborsTuple.toDF()\n",
    "cancerDf = cancerRddSch.toDF()\n",
    "#write union read file back clean data and create tuple rdd\n",
    "nodesDeg = nodesnetDf.join(degreesDf, nodesnetDf._1 == degreesDf._1).select(nodesnetDf[\"*\"],degreesDf[\"_2\"])\n",
    "nodesDegNeig = nodesDeg.join(neighborsDf, nodesDeg._1 == neighborsDf._1).select(nodesDeg[\"*\"], neighborsDf[\"_2\"])\n",
    "nodesDegNeigCan = nodesDegNeig.join(cancerDf, nodesDegNeig._1 == cancerDf._1).select(nodesDegNeig[\"*\"], cancerDf[\"_2\"])\n",
    "#Now we had a dataset with the form of protein Id, netproperties, indegrees, out degrees, amount of neighbors and cancer\n",
    "#I will print it and clean it again.\n",
    "#9606.ENSP00000257013\tRow(_1=u'0', _2=u'0', _3=u'0', _4=u'0', _5=u'3628', _6=u'7274', _7=u'239', _8=u'1874', _9=u'0', _10=u'0', _11=u'0', _12=u'1132', _13=u'7225', _14=u'34034')\tRow(_1=u'135', _2=u'135')\t135\t0\n",
    "nodesDegNeigCan.toPandas().to_csv(\"mlpreviousfinal.csv\")\n",
    "#manually delete first row"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Welcome to the soup: Data Cleaning to get our final dataset to perform the SVM algorithm."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this part, what I am going to do is preprocessing the output file again from the previous step to give it a shape of a proper matrix, for applying the machine learning algorithm. Afterwards, I am going to normalize the data and perform a PCA analysis, to find out the components/features that are valuable enough to be used in the machine learning algorithm. The principal component analysis algorithm is well known to be used as feature reduction. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Load the dataset as a list of strings\n",
    "datasemiFinal = open(\"/home/abi/spark/spark-2.4.3-bin-hadoop2.7/mlpreviousfinal.csv\",\"r\").readlines()\n",
    "#Clean the ugly string again T_T\n",
    "'9606.ENSP00000257013,\"Row(_1=u\\'0\\', _2=u\\'0\\', _3=u\\'0\\', _4=u\\'0\\', _5=u\\'3628\\', _6=u\\'7274\\', _7=u\\'239\\', _8=u\\'1874\\', _9=u\\'0\\', _10=u\\'0\\', _11=u\\'0\\', _12=u\\'1132\\', _13=u\\'7225\\', _14=u\\'34034\\')\",\"Row(_1=u\\'135\\', _2=u\\'135\\')\",135,0\\n'\n",
    "def cleanStr(str):\n",
    "    newstr = str.replace('\"',\"\")\n",
    "    str1 = newstr.replace('\\'',\"\")\n",
    "    str2 = str1.replace(\"Row(\",\"\")\n",
    "    str3 = str2.replace(\"_2=u\",\"\")\n",
    "    str4 = str3.replace(\"_3=u\",\"\")\n",
    "    str5 = str4.replace(\"_4=u\",\"\")\n",
    "    str6 = str5.replace(\"_5=u\",\"\")\n",
    "    str7 = str6.replace(\"_6=u\",\"\")\n",
    "    str8 = str7.replace(\"_7=u\",\"\")\n",
    "    str9 = str8.replace(\"_8=u\",\"\")\n",
    "    str10 = str9.replace(\"_9=u\",\"\")\n",
    "    str11 = str10.replace(\"_10=u\",\"\")\n",
    "    str12 = str11.replace(\"_11=u\",\"\")\n",
    "    str13 = str12.replace(\"_12=u\",\"\")\n",
    "    str14 = str13.replace(\"_13=u\",\"\")\n",
    "    str15 = str14.replace(\"_14=u\",\"\")\n",
    "    str16 = str15.replace(\"Row(_1=u\",\"\")\n",
    "    str17 = str16.replace(\"_2=u\",\"\")\n",
    "    str18 = str17.replace('\\')\"',\"\")\n",
    "    str19 = str18.replace(\"\\n\",\"\")\n",
    "    str20 = str19.replace(\")\",\"\")\n",
    "    str21 = str20.replace(\"_1=u\",\"\")\n",
    "    return str21\n",
    "datasemiFinalClean = map(cleanStr,datasemiFinal)\n",
    "#Turn into rdd to gave it a shape\n",
    "datasemiRdd = sc.parallelize(datasemiFinalClean).map(lambda x: x.split(\",\"))\n",
    "#Turn into pandas csv and print.\n",
    "datasemiRdd.toDF().toPandas().to_csv(\"machineLearningReady.csv\")\n",
    "#manually configure column names according to previous steps."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "After writing manually columns the machineLearningReady.csv. Has the columns in the format below:\n",
    "- id\n",
    "- neighborhood\n",
    "- neighborhoodTransferred\n",
    "- fusion \n",
    "- coocurence\n",
    "- homology\n",
    "- coexpression\n",
    "- coexpressionTransferred\n",
    "- experiments\n",
    "- experimentsTransferred\n",
    "- database\n",
    "- databaseTransferred\n",
    "- combinedScores\n",
    "- neighbors\n",
    "- inDegree\n",
    "- outDegree\n",
    "- cancer\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Machine Learning Part"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Descriptive Analytics: Dataset Exploration\n",
    "\n",
    "First of all, my data is too much for my computer. Thus, I am going to explore the dataset with some descritive analytics and then I am going to take a sample from my data. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {},
   "outputs": [],
   "source": [
    "#I am going to create my extra attribute with a for loop because any other method is too heavy for computation\n",
    "proteinNetworkCancer['neighbors'] = neighbors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "metadata": {},
   "outputs": [],
   "source": [
    "dataMl = proteinNetworkCancer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein1</th>\n",
       "      <th>protein2</th>\n",
       "      <th>neighborhood</th>\n",
       "      <th>fusion</th>\n",
       "      <th>cooccurence</th>\n",
       "      <th>coexpression</th>\n",
       "      <th>experimental</th>\n",
       "      <th>database</th>\n",
       "      <th>textmining</th>\n",
       "      <th>combined_score</th>\n",
       "      <th>cancer</th>\n",
       "      <th>cancer2</th>\n",
       "      <th>cancer_final</th>\n",
       "      <th>neighbors</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9606.ENSP00000000233</td>\n",
       "      <td>9606.ENSP00000272298</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>332</td>\n",
       "      <td>62</td>\n",
       "      <td>181</td>\n",
       "      <td>0</td>\n",
       "      <td>125</td>\n",
       "      <td>490</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>1662</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>9606.ENSP00000000233</td>\n",
       "      <td>9606.ENSP00000253401</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>186</td>\n",
       "      <td>0</td>\n",
       "      <td>56</td>\n",
       "      <td>198</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>1662</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>9606.ENSP00000000233</td>\n",
       "      <td>9606.ENSP00000401445</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>159</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>159</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>1662</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>9606.ENSP00000000233</td>\n",
       "      <td>9606.ENSP00000418915</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>61</td>\n",
       "      <td>158</td>\n",
       "      <td>0</td>\n",
       "      <td>542</td>\n",
       "      <td>606</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>1662</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>9606.ENSP00000000233</td>\n",
       "      <td>9606.ENSP00000327801</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>88</td>\n",
       "      <td>78</td>\n",
       "      <td>0</td>\n",
       "      <td>89</td>\n",
       "      <td>167</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>1662</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11759449</th>\n",
       "      <td>9606.ENSP00000485678</td>\n",
       "      <td>9606.ENSP00000310488</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>167</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>167</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11759450</th>\n",
       "      <td>9606.ENSP00000485678</td>\n",
       "      <td>9606.ENSP00000342448</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>175</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>175</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11759451</th>\n",
       "      <td>9606.ENSP00000485678</td>\n",
       "      <td>9606.ENSP00000350222</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>195</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>195</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11759452</th>\n",
       "      <td>9606.ENSP00000485678</td>\n",
       "      <td>9606.ENSP00000367590</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>900</td>\n",
       "      <td>0</td>\n",
       "      <td>900</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11759453</th>\n",
       "      <td>9606.ENSP00000485678</td>\n",
       "      <td>9606.ENSP00000349930</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>213</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>213</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>11759454 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      protein1              protein2  neighborhood  fusion  \\\n",
       "0         9606.ENSP00000000233  9606.ENSP00000272298             0       0   \n",
       "1         9606.ENSP00000000233  9606.ENSP00000253401             0       0   \n",
       "2         9606.ENSP00000000233  9606.ENSP00000401445             0       0   \n",
       "3         9606.ENSP00000000233  9606.ENSP00000418915             0       0   \n",
       "4         9606.ENSP00000000233  9606.ENSP00000327801             0       0   \n",
       "...                        ...                   ...           ...     ...   \n",
       "11759449  9606.ENSP00000485678  9606.ENSP00000310488             0       0   \n",
       "11759450  9606.ENSP00000485678  9606.ENSP00000342448             0       0   \n",
       "11759451  9606.ENSP00000485678  9606.ENSP00000350222             0       0   \n",
       "11759452  9606.ENSP00000485678  9606.ENSP00000367590             0       0   \n",
       "11759453  9606.ENSP00000485678  9606.ENSP00000349930             0       0   \n",
       "\n",
       "          cooccurence  coexpression  experimental  database  textmining  \\\n",
       "0                 332            62           181         0         125   \n",
       "1                   0             0           186         0          56   \n",
       "2                   0             0           159         0           0   \n",
       "3                   0            61           158         0         542   \n",
       "4                   0            88            78         0          89   \n",
       "...               ...           ...           ...       ...         ...   \n",
       "11759449            0           167             0         0           0   \n",
       "11759450            0           175             0         0           0   \n",
       "11759451            0           195             0         0           0   \n",
       "11759452            0             0             0       900           0   \n",
       "11759453            0           213             0         0           0   \n",
       "\n",
       "          combined_score  cancer  cancer2  cancer_final  neighbors  \n",
       "0                    490     NaN      NaN             0       1662  \n",
       "1                    198     NaN      NaN             0       1662  \n",
       "2                    159     NaN      NaN             0       1662  \n",
       "3                    606     NaN      NaN             0       1662  \n",
       "4                    167     NaN      NaN             0       1662  \n",
       "...                  ...     ...      ...           ...        ...  \n",
       "11759449             167     NaN      NaN             0        100  \n",
       "11759450             175     NaN      NaN             0        100  \n",
       "11759451             195     NaN      NaN             0        100  \n",
       "11759452             900     NaN      NaN             0        100  \n",
       "11759453             213     NaN      NaN             0        100  \n",
       "\n",
       "[11759454 rows x 14 columns]"
      ]
     },
     "execution_count": 122,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataMl"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Proportions\n",
    "\n",
    "I will check how is the proportion of 1 and 0 on my depending variable to check if I am in a problem with imbalanced data. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "#save the data to csv, because the kernel is constantly dying. \n",
    "#dataMl.to_csv('cancer_data.csv')\n",
    "dataMl = pd.read_csv('cancer_data.csv', sep=',', header='infer')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein1</th>\n",
       "      <th>protein2</th>\n",
       "      <th>neighborhood</th>\n",
       "      <th>fusion</th>\n",
       "      <th>cooccurence</th>\n",
       "      <th>coexpression</th>\n",
       "      <th>experimental</th>\n",
       "      <th>database</th>\n",
       "      <th>textmining</th>\n",
       "      <th>combined_score</th>\n",
       "      <th>neighbors</th>\n",
       "      <th>cancer</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9606.ENSP00000000233</td>\n",
       "      <td>9606.ENSP00000272298</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>332</td>\n",
       "      <td>62</td>\n",
       "      <td>181</td>\n",
       "      <td>0</td>\n",
       "      <td>125</td>\n",
       "      <td>490</td>\n",
       "      <td>1662</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>9606.ENSP00000000233</td>\n",
       "      <td>9606.ENSP00000253401</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>186</td>\n",
       "      <td>0</td>\n",
       "      <td>56</td>\n",
       "      <td>198</td>\n",
       "      <td>1662</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>9606.ENSP00000000233</td>\n",
       "      <td>9606.ENSP00000401445</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>159</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>159</td>\n",
       "      <td>1662</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>9606.ENSP00000000233</td>\n",
       "      <td>9606.ENSP00000418915</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>61</td>\n",
       "      <td>158</td>\n",
       "      <td>0</td>\n",
       "      <td>542</td>\n",
       "      <td>606</td>\n",
       "      <td>1662</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>9606.ENSP00000000233</td>\n",
       "      <td>9606.ENSP00000327801</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>88</td>\n",
       "      <td>78</td>\n",
       "      <td>0</td>\n",
       "      <td>89</td>\n",
       "      <td>167</td>\n",
       "      <td>1662</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11759449</th>\n",
       "      <td>9606.ENSP00000485678</td>\n",
       "      <td>9606.ENSP00000310488</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>167</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>167</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11759450</th>\n",
       "      <td>9606.ENSP00000485678</td>\n",
       "      <td>9606.ENSP00000342448</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>175</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>175</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11759451</th>\n",
       "      <td>9606.ENSP00000485678</td>\n",
       "      <td>9606.ENSP00000350222</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>195</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>195</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11759452</th>\n",
       "      <td>9606.ENSP00000485678</td>\n",
       "      <td>9606.ENSP00000367590</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>900</td>\n",
       "      <td>0</td>\n",
       "      <td>900</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11759453</th>\n",
       "      <td>9606.ENSP00000485678</td>\n",
       "      <td>9606.ENSP00000349930</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>213</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>213</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>11759454 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      protein1              protein2  neighborhood  fusion  \\\n",
       "0         9606.ENSP00000000233  9606.ENSP00000272298             0       0   \n",
       "1         9606.ENSP00000000233  9606.ENSP00000253401             0       0   \n",
       "2         9606.ENSP00000000233  9606.ENSP00000401445             0       0   \n",
       "3         9606.ENSP00000000233  9606.ENSP00000418915             0       0   \n",
       "4         9606.ENSP00000000233  9606.ENSP00000327801             0       0   \n",
       "...                        ...                   ...           ...     ...   \n",
       "11759449  9606.ENSP00000485678  9606.ENSP00000310488             0       0   \n",
       "11759450  9606.ENSP00000485678  9606.ENSP00000342448             0       0   \n",
       "11759451  9606.ENSP00000485678  9606.ENSP00000350222             0       0   \n",
       "11759452  9606.ENSP00000485678  9606.ENSP00000367590             0       0   \n",
       "11759453  9606.ENSP00000485678  9606.ENSP00000349930             0       0   \n",
       "\n",
       "          cooccurence  coexpression  experimental  database  textmining  \\\n",
       "0                 332            62           181         0         125   \n",
       "1                   0             0           186         0          56   \n",
       "2                   0             0           159         0           0   \n",
       "3                   0            61           158         0         542   \n",
       "4                   0            88            78         0          89   \n",
       "...               ...           ...           ...       ...         ...   \n",
       "11759449            0           167             0         0           0   \n",
       "11759450            0           175             0         0           0   \n",
       "11759451            0           195             0         0           0   \n",
       "11759452            0             0             0       900           0   \n",
       "11759453            0           213             0         0           0   \n",
       "\n",
       "          combined_score  neighbors  cancer  \n",
       "0                    490       1662       0  \n",
       "1                    198       1662       0  \n",
       "2                    159       1662       0  \n",
       "3                    606       1662       0  \n",
       "4                    167       1662       0  \n",
       "...                  ...        ...     ...  \n",
       "11759449             167        100       0  \n",
       "11759450             175        100       0  \n",
       "11759451             195        100       0  \n",
       "11759452             900        100       0  \n",
       "11759453             213        100       0  \n",
       "\n",
       "[11759454 rows x 12 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "del dataMl['Unnamed: 0']\n",
    "del dataMl['Unnamed: 0.1']\n",
    "dataMl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "dataMl_pos = dataMl[dataMl['cancer'] == 1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "dataMl_neg = dataMl[dataMl['cancer'] == 0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "cancer_pos = float(len(dataMl_pos))/float(len(dataMl))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "cancer_neg = float(len(dataMl_neg))/float(len(dataMl))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.999983672711335"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cancer_neg"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1.6327288664932914e-05"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cancer_pos"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As we can appreciate it is a clearly anomaly detection case where our negative label is less than 1% of our data. I am going to proceed to check some descriptive analtics over the variables and later on perform an oversampling method only on our negative class. I choose the oversampling over the negative class, because I am interested in been able to detect the negative class. Furthermore, I am not doing any undersampling method on the postive class because I do not want to break the patterns created by the network naturally. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* Neighborhood attributes: I am extracting the maximum and minimum values to get an idea what it will be a good size for the bins in my histogram to analyze distribution."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "max_val_neigh = dataMl['neighborhood'].max()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "min_val_neigh = dataMl['neighborhood'].min()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "std_val_negih = dataMl['neighborhood'].std()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "mean_val_neigh = dataMl['neighborhood'].mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "385"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "max_val_neigh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "min_val_neigh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "20.5037852888835"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "std_val_negih"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3.187574355067846"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mean_val_neigh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXoAAAEICAYAAABRSj9aAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/Il7ecAAAACXBIWXMAAAsTAAALEwEAmpwYAAAWJUlEQVR4nO3df5xldX3f8ddbELUsQZR0iyy6aEnSrViECRhJ46zRZlHLto9Qu5So5IFuk4pNq1ahpv6gMdX0gbYiiW5SihpkRBvDVmkwKlubGhG2Ij9Ft7Apv2QjCxtXiYp++sc9q9dxdmfm7p25Z7++no/HPPaec773nPc97Lz3zPfeOaSqkCS161GTDiBJWloWvSQ1zqKXpMZZ9JLUOItekhpn0UtS4yx6HTCS3JJkeoFjtyd53l62TSe5e4y5Lk3yW+Pa3z6OM9bc+vFx8KQDSAtVVX930hmkA5FX9NJ+SOLFknrPotey66ZVXpvkxiS7knwoyWO7bS9KckOSh5J8NskzZj3ved3jxyV5X5IHk9yW5HVzTGucMNcxhvb3b5N8rdvvWUPrD0/y/iR/meQvkvxmkkd1285O8r+TvDPJA8Cbu6cdkeTjSb6e5NokTxva37OTXNfluC7Js4e2PSnJ5iQ7k2xL8oqhbY/rpoUeTHIr8LP7eer1Y8qi16S8GFgHHAs8Azg7yTOBS4B/DjwReC+wOclj5nj+m4DVwFOB5wO/spBjDG37W8CRwNHAy4BNSX6623YRcHi37+cALwV+dei5pwB3ACuBt3brNgBvAY4Atu1Zn+QJwMeBd3Wv6R3Ax5M8sXveDHA38CTgDOC3kzx36DU+rfv6pS6ntGgTLfoklyTZkeTmBYx9Z3eld0OSLyd5aBkiaum8q6ruraqdwH8HTgA2Au+tqmur6rtV9T7gW8Cz5nj+i4HfrqoHq+puBkW6kGMM+3dV9a2q+p8MyvjFSQ5iUNrnV9XXq2o7cCHwkqHn3VtVF1XVI1X1cLfuo1X1+ap6BLhs6FgvBL5SVR/oxl8OfAn4h0mOAU4FXl9Vf11VNwB/wOAflj2v8a1VtbOq7trLa5TmNekr+ksZXHHNq6r+dVWdUFUnMLji+qMlzKWl99Whx98EVgBPAV7TTds81P1jfgyDq93ZngTcNbR81xxj5jrGHg9W1TeGlv+i2+eRwKO75eFtR494rCfN2tfw/p4E7Kyqr+/lWLNf4+z9SAsy0aKvqs8AO4fXJXlakj9JsjXJ/0ryM3M89Uzg8mUJqeV0F4Mr2McPff2N7ip4tvuAVUPLxyzyWEckOXRo+cnAvcDXgO8w+EdneNs9Q8uLueXrvbP2Nby/e4EnJDlsL8e6jx9+XU9exHGl75v0Ff1cNgGvqqqTgNcCvzu8MclTGMy5fnoC2bS0fh/4tSSnZODQJC+cVYR7XAGcn+SIJEcD545wvLckOSTJ3wdeBHy4qr7b7futSQ7r/r69GvjDEV/TVcBPJflnSQ5O8k+BNcDHuumYzwL/Icljuzeezxk61vBrXAW8asQM+jHXq6JPsgJ4NvDhJDcweDPuqFnDNgAf6b4h1ZCquh54BfBu4EEGb2qevZfhFzB4E/NO4JPARxjM5y/UV7tj3MtgTv3XqupL3bZXAd9g8IbrnwEfZPAm8aJV1QMM/hF5DfAA8DrgRVX1tW7ImQzeVL4X+Cjwpqr6ZLftLQyma+4EPgF8YJQMUib9Px5JsprB1c3Tk/wEcHtVzS734fFfAF5ZVZ9drozqvyS/DmyoqudMOovUN726oq+qvwLuTPJPALof3//enu3dfP0RwJ9PKKJ6IslRSU5N8qjuY5GvYXBFLGmWSX+88nIGpf3TSe5Ocg5wFnBOki8CtwDrh56yAZipSf8Yoj44hMHU3tcZvF9zJbPez5E0MPGpG0nS0urV1I0kafwmdkOmI488slavXj3Sc7/xjW9w6KGHzj9wQvqcr8/ZoN/5+pwN+p2vz9ngwMq3devWr1XVTy5qB1U1ka+TTjqpRnXNNdeM/Nzl0Od8fc5W1e98fc5W1e98fc5WdWDlA66vRfatUzeS1DiLXpIaZ9FLUuMseklqnEUvSY2z6CWpcRa9JDXOopekxln0ktS4id0CYX/cdM8uzj7v4xM59va3vXAix5WkUXlFL0mNs+glqXEWvSQ1zqKXpMZZ9JLUOItekhpn0UtS4yx6SWqcRS9JjbPoJalxFr0kNc6il6TGWfSS1DiLXpIaZ9FLUuMseklqnEUvSY2z6CWpcRa9JDXOopekxs1b9EkuSbIjyc172Z4k70qyLcmNSU4cf0xJ0qgWckV/KbBuH9tPA47rvjYCv7f/sSRJ4zJv0VfVZ4Cd+xiyHnh/DXwOeHySo8YVUJK0f8YxR380cNfQ8t3dOklSD6Sq5h+UrAY+VlVPn2Pbx4C3VdWfdcufAl5fVdfPMXYjg+kdVq5cedLMzMxIoXfs3MX9D4/01P12/NGHzztm9+7drFixYhnSLF6fs0G/8/U5G/Q7X5+zwYGVb+3atVuramoxzz94DBnuAY4ZWl7VrfsRVbUJ2AQwNTVV09PTIx3wosuu5MKbxhF98bafNT3vmC1btjDqa1tqfc4G/c7X52zQ73x9zgbt5xvH1M1m4KXdp2+eBeyqqvvGsF9J0hjMe1mc5HJgGjgyyd3Am4BHA1TVe4CrgBcA24BvAr+6VGElSYs3b9FX1ZnzbC/glWNLJEkaK38zVpIaZ9FLUuMseklqnEUvSY2z6CWpcRa9JDXOopekxln0ktQ4i16SGmfRS1LjLHpJapxFL0mNs+glqXEWvSQ1zqKXpMZZ9JLUOItekhpn0UtS4yx6SWqcRS9JjbPoJalxFr0kNc6il6TGWfSS1DiLXpIaZ9FLUuMseklqnEUvSY1bUNEnWZfk9iTbkpw3x/YnJ7kmyReS3JjkBeOPKkkaxbxFn+Qg4GLgNGANcGaSNbOG/SZwRVU9E9gA/O64g0qSRrOQK/qTgW1VdUdVfRuYAdbPGlPAT3SPDwfuHV9ESdL+SFXte0ByBrCuql7eLb8EOKWqzh0acxTwCeAI4FDgeVW1dY59bQQ2AqxcufKkmZmZkULv2LmL+x8e6an77fijD593zO7du1mxYsUypFm8PmeDfufrczbod74+Z4MDK9/atWu3VtXUYp5/8JhynAlcWlUXJvk54ANJnl5V3xseVFWbgE0AU1NTNT09PdLBLrrsSi68aVzRF2f7WdPzjtmyZQujvral1uds0O98fc4G/c7X52zQfr6FTN3cAxwztLyqWzfsHOAKgKr6c+CxwJEjp5Ikjc1Civ464LgkxyY5hMGbrZtnjfl/wC8CJPk7DIr+L8cZVJI0mnmLvqoeAc4FrgZuY/DpmluSXJDk9G7Ya4BXJPkicDlwds03+S9JWhYLmuiuqquAq2ate+PQ41uBU8cbTZI0Dv5mrCQ1zqKXpMZZ9JLUOItekhpn0UtS4yx6SWqcRS9JjbPoJalxFr0kNc6il6TGWfSS1DiLXpIaZ9FLUuMseklqnEUvSY2z6CWpcRa9JDXOopekxln0ktQ4i16SGmfRS1LjLHpJapxFL0mNs+glqXEWvSQ1zqKXpMZZ9JLUuAUVfZJ1SW5Psi3JeXsZ8+Iktya5JckHxxtTkjSqg+cbkOQg4GLg+cDdwHVJNlfVrUNjjgPOB06tqgeT/M2lCixJWpyFXNGfDGyrqjuq6tvADLB+1phXABdX1YMAVbVjvDElSaNKVe17QHIGsK6qXt4tvwQ4parOHRrzx8CXgVOBg4A3V9WfzLGvjcBGgJUrV540MzMzUugdO3dx/8MjPXW/HX/04fOO2b17NytWrFiGNIvX52zQ73x9zgb9ztfnbHBg5Vu7du3WqppazPPnnbpZxH6OA6aBVcBnkhxfVQ8ND6qqTcAmgKmpqZqenh7pYBdddiUX3jSu6Iuz/azpecds2bKFUV/bUutzNuh3vj5ng37n63M2aD/fQqZu7gGOGVpe1a0bdjewuaq+U1V3Mri6P27kVJKksVlI0V8HHJfk2CSHABuAzbPG/DGDq3mSHAn8FHDH+GJKkkY1b9FX1SPAucDVwG3AFVV1S5ILkpzeDbsaeCDJrcA1wL+pqgeWKrQkaeEWNNFdVVcBV81a98ahxwW8uvuSJPWIvxkrSY2z6CWpcRa9JDXOopekxln0ktQ4i16SGmfRS1LjLHpJapxFL0mNs+glqXEWvSQ1zqKXpMZZ9JLUOItekhpn0UtS4yx6SWqcRS9JjbPoJalxFr0kNc6il6TGWfSS1DiLXpIaZ9FLUuMseklqnEUvSY2z6CWpcRa9JDXOopekxi2o6JOsS3J7km1JztvHuF9OUkmmxhdRkrQ/5i36JAcBFwOnAWuAM5OsmWPcYcBvANeOO6QkaXQLuaI/GdhWVXdU1beBGWD9HOP+PfB24K/HmE+StJ9SVfsekJwBrKuql3fLLwFOqapzh8acCLyhqn45yRbgtVV1/Rz72ghsBFi5cuVJMzMzI4XesXMX9z880lP32/FHHz7vmN27d7NixYplSLN4fc4G/c7X52zQ73x9zgYHVr61a9durapFTY8fvL8BkjwKeAdw9nxjq2oTsAlgamqqpqenRzrmRZddyYU37Xf0kWw/a3reMVu2bGHU17bU+pwN+p2vz9mg3/n6nA3az7eQqZt7gGOGlld16/Y4DHg6sCXJduBZwGbfkJWkflhI0V8HHJfk2CSHABuAzXs2VtWuqjqyqlZX1Wrgc8Dpc03dSJKW37xFX1WPAOcCVwO3AVdU1S1JLkhy+lIHlCTtnwVNdFfVVcBVs9a9cS9jp/c/liRpXPzNWElqnEUvSY2z6CWpcRa9JDXOopekxln0ktQ4i16SGmfRS1LjLHpJapxFL0mNs+glqXEWvSQ1zqKXpMZZ9JLUOItekhpn0UtS4yx6SWqcRS9JjbPoJalxFr0kNc6il6TGWfSS1DiLXpIaZ9FLUuMseklqnEUvSY2z6CWpcQsq+iTrktyeZFuS8+bY/uoktya5Mcmnkjxl/FElSaOYt+iTHARcDJwGrAHOTLJm1rAvAFNV9QzgI8DvjDuoJGk0C7miPxnYVlV3VNW3gRlg/fCAqrqmqr7ZLX4OWDXemJKkUaWq9j0gOQNYV1Uv75ZfApxSVefuZfy7ga9W1W/NsW0jsBFg5cqVJ83MzIwUesfOXdz/8EhP3W/HH334vGN2797NihUrliHN4vU5G/Q7X5+zQb/z9TkbHFj51q5du7Wqphbz/IPHGSbJrwBTwHPm2l5Vm4BNAFNTUzU9PT3ScS667EouvGms0Rds+1nT847ZsmULo762pdbnbNDvfH3OBv3O1+ds0H6+hbTlPcAxQ8urunU/JMnzgDcAz6mqb42cSJI0VguZo78OOC7JsUkOATYAm4cHJHkm8F7g9KraMf6YkqRRzVv0VfUIcC5wNXAbcEVV3ZLkgiSnd8P+I7AC+HCSG5Js3svuJEnLbEET3VV1FXDVrHVvHHr8vDHnkiSNib8ZK0mNs+glqXEWvSQ1zqKXpMZZ9JLUOItekhpn0UtS4yx6SWqcRS9JjbPoJalxFr0kNc6il6TGWfSS1DiLXpIaZ9FLUuMseklqnEUvSY2z6CWpcRa9JDXOopekxln0ktQ4i16SGmfRS1LjLHpJapxFL0mNs+glqXEWvSQ17uBJBzjQrD7v4/OOec3xj3D2AsYtxva3vXCs+5P042NBV/RJ1iW5Pcm2JOfNsf0xST7Ubb82yeqxJ5UkjWTeok9yEHAxcBqwBjgzyZpZw84BHqyqvw28E3j7uINKkkazkKmbk4FtVXUHQJIZYD1w69CY9cCbu8cfAd6dJFVVY8z6Y20hU0YLsRTTSuPU53zLkc0pOi2FhRT90cBdQ8t3A6fsbUxVPZJkF/BE4GvDg5JsBDZ2i7uT3D5KaODI2fvuk3/Z43x9zgb9zrcc2bJ/Pwv39tzR72xwYOV7ymKfvKxvxlbVJmDT/u4nyfVVNTWGSEuiz/n6nA36na/P2aDf+fqcDdrPt5A3Y+8BjhlaXtWtm3NMkoOBw4EHRg0lSRqfhRT9dcBxSY5NcgiwAdg8a8xm4GXd4zOATzs/L0n9MO/UTTfnfi5wNXAQcElV3ZLkAuD6qtoM/BfgA0m2ATsZ/GOwlPZ7+meJ9Tlfn7NBv/P1ORv0O1+fs0Hj+eKFtyS1zVsgSFLjLHpJatwBV/Tz3Y5hAnm2J7kpyQ1Jru/WPSHJnyb5SvfnEcuY55IkO5LcPLRuzjwZeFd3Lm9McuKE8r05yT3dObwhyQuGtp3f5bs9yS8tcbZjklyT5NYktyT5jW79xM/fPrL15dw9Nsnnk3yxy/eWbv2x3W1RtnW3STmkW79st03ZR7ZLk9w5dO5O6NYv+/dFd9yDknwhyce65fGdu6o6YL4YvBn8f4GnAocAXwTWTDjTduDIWet+Bzive3we8PZlzPMLwInAzfPlAV4A/A8gwLOAayeU783Aa+cYu6b7b/wY4Njuv/1BS5jtKODE7vFhwJe7DBM/f/vI1pdzF2BF9/jRwLXdObkC2NCtfw/w693jfwG8p3u8AfjQBLJdCpwxx/hl/77ojvtq4IPAx7rlsZ27A+2K/vu3Y6iqbwN7bsfQN+uB93WP3wf8o+U6cFV9hsEnnxaSZz3w/hr4HPD4JEdNIN/erAdmqupbVXUnsI3B34GlynZfVf2f7vHXgdsY/Nb3xM/fPrLtzXKfu6qq3d3io7uvAp7L4LYo8KPnbs85/Qjwi0myzNn2Ztm/L5KsAl4I/EG3HMZ47g60op/rdgz7+su+HAr4RJKtGdziAWBlVd3XPf4qsHIy0b5vb3n6dD7P7X5MvmRoqmti+bofh5/J4OqvV+dvVjboybnrph5uAHYAf8rgp4iHquqROTL80G1TgD23TVmWbFW159y9tTt370zymNnZ5si9VP4T8Drge93yExnjuTvQir6Pfr6qTmRwd89XJvmF4Y01+PmqN59h7Vuezu8BTwNOAO4DLpxkmCQrgP8G/Kuq+qvhbZM+f3Nk6825q6rvVtUJDH57/mTgZyaVZbbZ2ZI8HTifQcafBZ4AvH4S2ZK8CNhRVVuX6hgHWtEv5HYMy6qq7un+3AF8lMFf8Pv3/KjX/bljcglhH3l6cT6r6v7uG/F7wO/zgymGZc+X5NEMivSyqvqjbnUvzt9c2fp07vaoqoeAa4CfYzDtsecXM4czTOS2KUPZ1nXTYVVV3wL+K5M7d6cCpyfZzmA6+rnAf2aM5+5AK/qF3I5h2SQ5NMlhex4D/wC4mR++JcTLgCsnk/D79pZnM/DS7lMGzwJ2DU1RLJtZ85//mME53JNvQ/cpg2OB44DPL2GOMPgt79uq6h1DmyZ+/vaWrUfn7ieTPL57/Djg+QzeR7iGwW1R4EfP3bLcNmUv2b409I93GMx/D5+7Zfu+qKrzq2pVVa1m0GmfrqqzGOe5W+p3ksf9xeAd8S8zmP97w4SzPJXBJxu+CNyyJw+D+bJPAV8BPgk8YRkzXc7gR/jvMJjXO2dveRh8quDi7lzeBExNKN8HuuPf2P0lPmpo/Bu6fLcDpy1xtp9nMC1zI3BD9/WCPpy/fWTry7l7BvCFLsfNwBuHvkc+z+DN4A8Dj+nWP7Zb3tZtf+oEsn26O3c3A3/IDz6Zs+zfF0NZp/nBp27Gdu68BYIkNe5Am7qRJC2SRS9JjbPoJalxFr0kNc6il6TGWfSS1DiLXpIa9/8Bi4u63/d8ktUAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "dataMl.hist(column='neighborhood', bins=10);"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* Fusion attributes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "max_val_fus = dataMl['fusion'].max()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "min_val_fus = dataMl['fusion'].min()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "std_val_fus = dataMl['fusion'].std()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "mean_val_fus = dataMl['fusion'].mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "900"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "max_val_fus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "min_val_fus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "7.620421168353864"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "std_val_fus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.16457702883144065"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mean_val_fus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXQAAAEICAYAAABPgw/pAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/Il7ecAAAACXBIWXMAAAsTAAALEwEAmpwYAAAS40lEQVR4nO3df5Bd9Xnf8fcnyGAb2Rhb7RZLxMKJYlcDDrY3/Kgz9WKTRjgZNNMmGTQ0iTPYms6YxnWIOzBpSUs6bdyU0JpgO5rYJfFQVEw8QUMU0wbYcVrHFFQTMBDZCihG8g8cY3AXN8Gqn/5xj7x3l9Xeq9XdvdJ336+ZHd1zznfvfe6jo4/Ofu85Z1NVSJJOfN837gIkSaNhoEtSIwx0SWqEgS5JjTDQJakRBrokNcJAV3OSvC7Jg0n+T5JfXOJzfCTJvxx1bdJyiuehqzVJPgp8q6reN+5apJXkEbpa9BrgkXEXIa00A11NSXIPcBHwW0lmknw5ybv6tr8zyf/oHifJDUmeSvKtJA8nObvbdnOSf9P3fe9Osi/J00l2JXl137ZK8k+SfDHJM0luSpKVe9dSj4GuplTV24A/Aa6sqrXAFxYZ/g+Avw/8EHAa8DPAN+YPSvI24N91288A/hLYOW/YTwI/AryhG/fjx/RGpCUYa6An+Vh3dPT5Icbe0H3Q9WCSLyR5ZgVKVNu+A7wMeD29z5Meq6qvLDDucuBjVfW/q+pvgGuAC5Ns7Bvz61X1TFV9CbgXOHd5S5deaNxH6DcDW4YZWFXvq6pzq+pc4Ebgk8tYl1aBqroH+C3gJuCpJDuSvHyBoa+md1R++Ptm6B3Jr+8b89W+x98G1o6+YmlxYw30qvo08HT/uiQ/kORTSfYk+ZMkr1/gW7cBt65IkTrRPQe8tG/57/RvrKoPVtWbgc30pl7ev8BzfJneB60AJDkVeBVwcOTVSsdg3EfoC9kB/NPuH9kvAx/q35jkNcBZwD1jqE0nngeBf5jkpUl+ELji8IYkP5Lk/CQvohf8fw18d4HnuBX4hSTnJjkF+LfAfVW1f9mrl47CmnEX0C/JWuDvAZ/oO0nglHnDLgNur6r/t5K16YR1A70PK78GPATcAlzcbXt5t/219ML8LuA35j9BVf1xd5HR7wOnA5+htx9Kx5WxX1jUfbB0Z1Wd3c1f7q2qMxYZ/zngPVX1mZWqUZJOBMfVlEtVfQt4IslPw/fOE/7hw9u7+fTTgT8dU4mSdNwa92mLt9IL59clOZDkCnqniF2R5M/oXe23te9bLgN21rh/rJCk49DYp1wkSaNxXE25SJKWbmxnuaxbt642bty4pO997rnnOPXUU0db0AnMfsxlP2bZi7la6MeePXv+qqr+1kLbxhboGzdu5IEHHljS905PTzM1NTXagk5g9mMu+zHLXszVQj+S/OWRtjnlIkmNGBjog26gleTyJA91tx79TP9phpKklTPMEfrNLH4DrSeAt1bVOcCv0bt0X5K0wgbOoVfVp+fdJnT+9v4rNj8LbBhBXZKkozTUeej9l+cPGPfLwOur6l1H2L4d2A4wMTHx5p075/+OgOHMzMywdq13Jz3MfsxlP2bZi7la6MdFF120p6omF9o2srNcklxE7052P3qkMVW1g25KZnJyspb6aXMLn1SPkv2Yy37Mshdztd6PkQR6kjcAvwNcUlUv+BVekqTld8ynLSb5fnq/Pehnq2qx398oSVpGA4/QuxtoTQHrkhwAfhV4EUBVfQS4lt5vb/lQdw/zQ0ea35EkLZ9hznLZNmD7u4AFPwRdLg8ffJZ3Xv2HK/mS37P/139iLK8rSYN4pagkNcJAl6RGGOiS1AgDXZIaYaBLUiMMdElqhIEuSY0w0CWpEQa6JDXCQJekRhjoktQIA12SGmGgS1IjDHRJaoSBLkmNMNAlqREGuiQ1wkCXpEYY6JLUCANdkhphoEtSIwx0SWqEgS5JjTDQJakRBrokNcJAl6RGDAz0JB9L8lSSzx9he5J8MMm+JA8ledPoy5QkDTLMEfrNwJZFtl8CbOq+tgMfPvayJElHa2CgV9WngacXGbIV+L3q+SzwiiRnjKpASdJwRjGHvh54sm/5QLdOkrSC1qzkiyXZTm9ahomJCaanp5f0PBMvgavOOTTCyoa31JqX08zMzHFZ17jYj1n2Yq7W+zGKQD8InNm3vKFb9wJVtQPYATA5OVlTU1NLesEbb7mD6x9e0f+Lvmf/5VNjed3FTE9Ps9Retsh+zLIXc7Xej1FMuewCfq472+UC4Nmq+soInleSdBQGHuYmuRWYAtYlOQD8KvAigKr6CLAbeAewD/g28AvLVawk6cgGBnpVbRuwvYD3jKwiSdKSeKWoJDXCQJekRhjoktQIA12SGmGgS1IjDHRJaoSBLkmNMNAlqREGuiQ1wkCXpEYY6JLUCANdkhphoEtSIwx0SWqEgS5JjTDQJakRBrokNcJAl6RGGOiS1AgDXZIaYaBLUiMMdElqhIEuSY0w0CWpEQa6JDXCQJekRhjoktSIoQI9yZYke5PsS3L1Atu/P8m9ST6X5KEk7xh9qZKkxQwM9CQnATcBlwCbgW1JNs8b9i+A26rqjcBlwIdGXagkaXHDHKGfB+yrqser6nlgJ7B13pgCXt49Pg348uhKlCQNI1W1+IDkp4AtVfWubvlngfOr6sq+MWcA/w04HTgVuLiq9izwXNuB7QATExNv3rlz55KKfurpZ/na/13Stx6zc9afNp4XXsTMzAxr164ddxnHDfsxy17M1UI/Lrrooj1VNbnQtjUjeo1twM1VdX2SC4GPJzm7qr7bP6iqdgA7ACYnJ2tqampJL3bjLXdw/cOjKv3o7L98aiyvu5jp6WmW2ssW2Y9Z9mKu1vsxzJTLQeDMvuUN3bp+VwC3AVTVnwIvBtaNokBJ0nCGCfT7gU1JzkpyMr0PPXfNG/Ml4O0ASf4uvUD/+igLlSQtbmCgV9Uh4ErgLuAxemezPJLkuiSXdsOuAt6d5M+AW4F31qDJeUnSSA01EV1Vu4Hd89Zd2/f4UeAtoy1NknQ0vFJUkhphoEtSIwx0SWqEgS5JjTDQJakRBrokNcJAl6RGGOiS1AgDXZIaYaBLUiMMdElqhIEuSY0w0CWpEQa6JDXCQJekRhjoktQIA12SGmGgS1IjDHRJaoSBLkmNMNAlqREGuiQ1wkCXpEYY6JLUCANdkhphoEtSI4YK9CRbkuxNsi/J1UcY8zNJHk3ySJL/MtoyJUmDrBk0IMlJwE3AjwEHgPuT7KqqR/vGbAKuAd5SVd9M8reXq2BJ0sKGOUI/D9hXVY9X1fPATmDrvDHvBm6qqm8CVNVToy1TkjTIwCN0YD3wZN/yAeD8eWN+CCDJ/wROAv5VVX1q/hMl2Q5sB5iYmGB6enoJJcPES+Cqcw4t6XuP1VJrXk4zMzPHZV3jYj9m2Yu5Wu/HMIE+7PNsAqaADcCnk5xTVc/0D6qqHcAOgMnJyZqamlrSi914yx1c//CoSj86+y+fGsvrLmZ6epql9rJF9mOWvZir9X4MM+VyEDizb3lDt67fAWBXVX2nqp4AvkAv4CVJK2SYQL8f2JTkrCQnA5cBu+aN+QN6R+ckWUdvCubx0ZUpSRpkYKBX1SHgSuAu4DHgtqp6JMl1SS7tht0FfCPJo8C9wPur6hvLVbQk6YWGmoiuqt3A7nnrru17XMAvdV+SpDHwSlFJaoSBLkmNMNAlqREGuiQ1wkCXpEYY6JLUCANdkhphoEtSIwx0SWqEgS5JjTDQJakRBrokNcJAl6RGGOiS1AgDXZIaYaBLUiMMdElqhIEuSY0w0CWpEQa6JDXCQJekRhjoktQIA12SGmGgS1IjDHRJaoSBLkmNGCrQk2xJsjfJviRXLzLuHyWpJJOjK1GSNIyBgZ7kJOAm4BJgM7AtyeYFxr0MeC9w36iLlCQNNswR+nnAvqp6vKqeB3YCWxcY92vAB4C/HmF9kqQhrRlizHrgyb7lA8D5/QOSvAk4s6r+MMn7j/RESbYD2wEmJiaYnp4+6oIBJl4CV51zaEnfe6yWWvNympmZOS7rGhf7MctezNV6P4YJ9EUl+T7gN4F3DhpbVTuAHQCTk5M1NTW1pNe88ZY7uP7hYy59SfZfPjWW113M9PQ0S+1li+zHLHsxV+v9GGbK5SBwZt/yhm7dYS8Dzgamk+wHLgB2+cGoJK2sYQL9fmBTkrOSnAxcBuw6vLGqnq2qdVW1sao2Ap8FLq2qB5alYknSggYGelUdAq4E7gIeA26rqkeSXJfk0uUuUJI0nKEmoqtqN7B73rprjzB26tjLkiQdLa8UlaRGGOiS1AgDXZIaYaBLUiMMdElqhIEuSY0w0CWpEQa6JDXCQJekRhjoktQIA12SGmGgS1IjDHRJaoSBLkmNMNAlqREGuiQ1wkCXpEYY6JLUCANdkhphoEtSIwx0SWqEgS5JjTDQJakRBrokNcJAl6RGGOiS1IihAj3JliR7k+xLcvUC238pyaNJHkpyd5LXjL5USdJiBgZ6kpOAm4BLgM3AtiSb5w37HDBZVW8Abgf+/agLlSQtbpgj9POAfVX1eFU9D+wEtvYPqKp7q+rb3eJngQ2jLVOSNMgwgb4eeLJv+UC37kiuAP7oWIqSJB29NaN8siT/GJgE3nqE7duB7QATExNMT08v6XUmXgJXnXNoiVUem6XWvJxmZmaOy7rGxX7Mshdztd6PYQL9IHBm3/KGbt0cSS4GfgV4a1X9zUJPVFU7gB0Ak5OTNTU1dbT1AnDjLXdw/cMj/b9oaPsvnxrL6y5menqapfayRfZjlr2Yq/V+DDPlcj+wKclZSU4GLgN29Q9I8kbgt4FLq+qp0ZcpSRpkYKBX1SHgSuAu4DHgtqp6JMl1SS7thv0GsBb4RJIHk+w6wtNJkpbJUPMWVbUb2D1v3bV9jy8ecV2SpKPklaKS1AgDXZIaYaBLUiMMdElqhIEuSY0w0CWpEQa6JDXCQJekRhjoktQIA12SGmGgS1IjDHRJaoSBLkmNMNAlqREGuiQ1wkCXpEYY6JLUCANdkhphoEtSIwx0SWqEgS5JjTDQJakRBrokNcJAl6RGGOiS1AgDXZIaYaBLUiOGCvQkW5LsTbIvydULbD8lyX/ttt+XZOPIK5UkLWpgoCc5CbgJuATYDGxLsnnesCuAb1bVDwI3AB8YdaGSpMUNc4R+HrCvqh6vqueBncDWeWO2Ar/bPb4deHuSjK5MSdIga4YYsx54sm/5AHD+kcZU1aEkzwKvAv6qf1CS7cD2bnEmyd6lFA2sm//cKyXH588eY+vHccp+zLIXc7XQj9ccacMwgT4yVbUD2HGsz5PkgaqaHEFJTbAfc9mPWfZirtb7McyUy0HgzL7lDd26BcckWQOcBnxjFAVKkoYzTKDfD2xKclaSk4HLgF3zxuwCfr57/FPAPVVVoytTkjTIwCmXbk78SuAu4CTgY1X1SJLrgAeqahfwUeDjSfYBT9ML/eV0zNM2jbEfc9mPWfZirqb7EQ+kJakNXikqSY0w0CWpESdcoA+6DUFrkpyZ5N4kjyZ5JMl7u/WvTPLfk3yx+/P0bn2SfLDrz0NJ3jTed7A8kpyU5HNJ7uyWz+puO7Gvuw3Fyd365m9LkeQVSW5P8udJHkty4WrdP5K8r/t38vkktyZ58WraN06oQB/yNgStOQRcVVWbgQuA93Tv+Wrg7qraBNzdLUOvN5u6r+3Ah1e+5BXxXuCxvuUPADd0t5/4Jr3bUcDquC3FfwI+VVWvB36YXl9W3f6RZD3wi8BkVZ1N7ySOy1hN+0ZVnTBfwIXAXX3L1wDXjLuuFe7BHcCPAXuBM7p1ZwB7u8e/DWzrG/+9ca180bsW4m7gbcCdQOhd/bdm/n5C7+ysC7vHa7pxGfd7GGEvTgOemP+eVuP+wewV66/s/q7vBH58Ne0bJ9QROgvfhmD9mGpZcd2PhG8E7gMmquor3aavAhPd49XQo/8I/HPgu93yq4BnqupQt9z/nufclgI4fFuKVpwFfB34z90U1O8kOZVVuH9U1UHgPwBfAr5C7+96D6to3zjRAn3VSrIW+H3gn1XVt/q3Ve8QY1Wcf5rkJ4GnqmrPuGs5TqwB3gR8uKreCDzH7PQKsHr2j+5zgq30/pN7NXAqsGWsRa2wEy3Qh7kNQXOSvIhemN9SVZ/sVn8tyRnd9jOAp7r1rffoLcClSfbTu/Pn2+jNIb+iu+0EzH3Prd+W4gBwoKru65Zvpxfwq3H/uBh4oqq+XlXfAT5Jb39ZNfvGiRbow9yGoCndbYg/CjxWVb/Zt6n/dgs/T29u/fD6n+vOZrgAeLbvR+8TXlVdU1Ubqmojvb//e6rqcuBeeredgBf2o9nbUlTVV4Enk7yuW/V24FFW5/7xJeCCJC/t/t0c7sXq2TfGPYm/hA8+3gF8AfgL4FfGXc8KvN8fpffj8kPAg93XO+jN9d0NfBH4Y+CV3fjQOxPoL4CH6X3iP/b3sUy9mQLu7B6/FvhfwD7gE8Ap3foXd8v7uu2vHXfdy9CHc4EHun3kD4DTV+v+Afxr4M+BzwMfB05ZTfuGl/5LUiNOtCkXSdIRGOiS1AgDXZIaYaBLUiMMdElqhIEuSY0w0CWpEf8flYXhXL4BMwcAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "dataMl.hist(column='fusion', bins=10);"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* Cooccurence attributes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "max_val_coo = dataMl['cooccurence'].max()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "min_val_coo = dataMl['cooccurence'].min()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "std_val_coo = dataMl['cooccurence'].std()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "mean_val_fus = dataMl['cooccurence'].mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "448"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "max_val_coo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "min_val_coo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "13.466937620097767"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "std_val_coo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.8819394165749532"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mean_val_fus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[<AxesSubplot:title={'center':'cooccurence'}>]], dtype=object)"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXQAAAEICAYAAABPgw/pAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/Il7ecAAAACXBIWXMAAAsTAAALEwEAmpwYAAAUFklEQVR4nO3df7DldX3f8edLFoSw8SfmhrAri2HV7IiJegtYM+Mlms6CDXSmNmVL/NGiO9OKQyKmxZpiiklH20GbIEZ3EsroEDZobdjitmiRWzI1KFCQn7NmRZBdf6zID1mp4uq7f5zvZs65uT/O3j13z97PfT5mzuz3x+d8v+/73ruv+72fc853U1VIkpa/Z4y7AEnSaBjoktQIA12SGmGgS1IjDHRJaoSBLkmNMNAlqREGuiQ1wkCXFiHJqnHXIM1koOuwkGRtks8k+W6S7yX5SJJnJPm9JA8l2ZPkE0me3fecs5Pcm+TxJNNJfmm+4/Xte3uS+5M8meS+JK/stleSk/vGXZXkD7rlqSS7kvybJN8G/ktX38VJvtad49okz+vGr+uO95Yk30jySJL39h37iCT/tnvuk0luT7K22/fSJJ9P8miSHUl+cwlbr4YY6Bq7JEcA1wMPAeuAE4CtwFu7xxnAi4DVwEe657wYuAb4beAFwHbgvyc5ap7jkeSfAL8PvBl4FnA28L0hS/154HnAicBm4J3APwJeC/wC8BhwxYzn/CrwEuB1wCV9P3TeBWwCzurq+BfAU0mOBT4P/Dnwc8C5wEeTbBiyRq1kVTW2B3AlsAe4Z4ixHwbu7B5fBR4fZ+0+Rvp98Grgu8CqGdtvBP5V3/pLgB8Dq4B/B1zbt+8ZwG5gaq7jdeNuAC6co44CTu5bvwr4g255CngaOLpv//3A6/rWj++rb113vDV9+78MnNst7wDOmaWGfwr81YxtHwfeN+6/Jx+H/2Pc84BX0bvi+sRCA6vqd/YvJ3kn8IqlK0uH2FrgoaraN2P7L9C7yt7vIXphOTFzX1X9NMnD9K7GfzzH8faf62uLrPO7VfXDvvUTgf+W5Kd9237S1bfft/uWn6L3W8Z8dZwInJbk8b5tq4BPLrJmrSBjnXKpqpuBR/u3JfnFJP+zm1P8qyQvneWpm+j9uq02PAy8cJYXGr9JL+D2eyGwD/jOzH1JQi8kd89zvP3n+sU56ngK+Jm+9Z+fsX/mrUkfBs6squf0PY6uqt1zHH+YOh4G/veMY66uqn85xDG1wh2Oc+hbgHdW1auAdwMf7d+Z5ETgJOALY6hNS+PLwLeADyQ5NsnRSV5D74f27yQ5Kclq4D8Af9FdeV8LvCHJ65IcCVwE/Aj44jzHA/hT4N1JXpWek7vvKehN5/2z7gXLjfTmxufzMeAP9z8/yQuSnDPk1/ynwPuTrO/qeHmS59Ob+39xkjclObJ7/L3+F3yluRxWgd79o/37wKeS3Elv7vD4GcPOBT5dVT85xOVpiXR/l78BnAx8A9hFby75SnpTDTcDXwd+SO+FSKpqB/BbwOXAI93zf6Oqnp7neFTVp4A/pPei45PAX9J7oRPgwu55jwPndfvm80fANuBzSZ4EbgFOG/LL/hC9H0qfA74P/BlwTFU9CfwDet/n36Q3ZfNB4JlDHlcrWKrG+x9cJFkHXF9VL0vyLGBHVc0M8f7xdwDvqKovHqoaJWk5OKyu0Kvq+8DXu7eW0f0q+sv793fz6c8F/npMJUrSYWusgZ7kGnrh/JLuQxvn0/tV9/wkXwHuBfrnJM8Ftta4f62QpMPQ2KdcJEmjcVhNuUiSFm9sHyw67rjjat26dYt67g9+8AOOPfbY0Ra0jNmPQfZjkP0YtNz7cfvttz9SVS+Ybd/YAn3dunXcdttti3ru9PQ0U1NToy1oGbMfg+zHIPsxaLn3I8lDc+1bcMolyZXdne7umWP/eUnuSnJ3ki/2vytFknToDDOHfhWwcZ79XwdeW1WnAO+n90lPSdIhtuCUS1Xd3H34Z679/R/wuQVYM4K6JEkHaKi3LfZ/mnOBce8GXlpVb5tj/2Z695FmYmLiVVu3bj3gggH27t3L6tWrFx64QtiPQfZjkP0YtNz7ccYZZ9xeVZOz7RvZi6JJzgDOp3dD/1lV1Ra6KZnJycla7AsTy/1FjVGzH4PsxyD7Majlfowk0JO8nN7d486sqmH/9xdJ0ggd9AeLkrwQ+Azwpqr66sGXJElajAWv0Lv7rUwBxyXZBbwPOBKgqj4GXAI8n97/ewiwb675HUnS0hnmXS6bFtj/NmDWF0ElSYfOuP9P0UW5e/cTvPXiz47l3A9+4A1jOa8kLcSbc0lSIwx0SWqEgS5JjTDQJakRBrokNcJAl6RGGOiS1AgDXZIaYaBLUiMMdElqhIEuSY0w0CWpEQa6JDXCQJekRhjoktQIA12SGmGgS1IjDHRJaoSBLkmNMNAlqREGuiQ1wkCXpEYY6JLUCANdkhphoEtSIwx0SWrEgoGe5Moke5LcM8f+JPnjJDuT3JXklaMvU5K0kGGu0K8CNs6z/0xgfffYDPzJwZclSTpQCwZ6Vd0MPDrPkHOAT1TPLcBzkhw/qgIlScNZNYJjnAA83Le+q9v2rZkDk2ymdxXPxMQE09PTizrhxDFw0Sn7FvXcg7XYmpfS3r17D8u6xsV+DLIfg1ruxygCfWhVtQXYAjA5OVlTU1OLOs7lV1/HZXcf0tL/1oPnTY3lvPOZnp5msb1skf0YZD8GtdyPUbzLZTewtm99TbdNknQIjSLQtwFv7t7tcjrwRFX9nekWSdLSWnDeIsk1wBRwXJJdwPuAIwGq6mPAduAsYCfwFPDPl6pYSdLcFgz0qtq0wP4C3jGyiiRJi+InRSWpEQa6JDXCQJekRhjoktQIA12SGmGgS1IjDHRJaoSBLkmNMNAlqREGuiQ1wkCXpEYY6JLUCANdkhphoEtSIwx0SWqEgS5JjTDQJakRBrokNcJAl6RGGOiS1AgDXZIaYaBLUiMMdElqhIEuSY0w0CWpEQa6JDViqEBPsjHJjiQ7k1w8y/4XJrkpyR1J7kpy1uhLlSTNZ8FAT3IEcAVwJrAB2JRkw4xhvwdcW1WvAM4FPjrqQiVJ8xvmCv1UYGdVPVBVTwNbgXNmjCngWd3ys4Fvjq5ESdIwUlXzD0jeCGysqrd1628CTquqC/rGHA98DngucCzw+qq6fZZjbQY2A0xMTLxq69atiyp6z6NP8J3/t6inHrRTTnj2eE48j71797J69epxl3HYsB+D7Meg5d6PM8444/aqmpxt36oRnWMTcFVVXZbk1cAnk7ysqn7aP6iqtgBbACYnJ2tqampRJ7v86uu47O5RlX5gHjxvaiznnc/09DSL7WWL7Mcg+zGo5X4MM+WyG1jbt76m29bvfOBagKr6a+Bo4LhRFChJGs4wgX4rsD7JSUmOovei57YZY74BvA4gyS/RC/TvjrJQSdL8Fgz0qtoHXADcANxP790s9ya5NMnZ3bCLgLcn+QpwDfDWWmhyXpI0UkNNRFfVdmD7jG2X9C3fB7xmtKVJkg6EnxSVpEYY6JLUCANdkhphoEtSIwx0SWqEgS5JjTDQJakRBrokNcJAl6RGGOiS1AgDXZIaYaBLUiMMdElqhIEuSY0w0CWpEQa6JDXCQJekRhjoktQIA12SGmGgS1IjDHRJaoSBLkmNMNAlqREGuiQ1wkCXpEYY6JLUiKECPcnGJDuS7Exy8RxjfjPJfUnuTfLnoy1TkrSQVQsNSHIEcAXw68Au4NYk26rqvr4x64H3AK+pqseS/NxSFSxJmt0wV+inAjur6oGqehrYCpwzY8zbgSuq6jGAqtoz2jIlSQsZJtBPAB7uW9/Vbev3YuDFSf5PkluSbBxVgZKk4Sw45XIAx1kPTAFrgJuTnFJVj/cPSrIZ2AwwMTHB9PT0ok42cQxcdMq+gyh38RZb81Lau3fvYVnXuNiPQfZjUMv9GCbQdwNr+9bXdNv67QK+VFU/Br6e5Kv0Av7W/kFVtQXYAjA5OVlTU1OLKvryq6/jsrtH9bPowDx43tRYzjuf6elpFtvLFtmPQfZjUMv9GGbK5VZgfZKTkhwFnAtsmzHmL+ldnZPkOHpTMA+MrkxJ0kIWDPSq2gdcANwA3A9cW1X3Jrk0ydndsBuA7yW5D7gJ+N2q+t5SFS1J+ruGmreoqu3A9hnbLulbLuBd3UOSNAZ+UlSSGmGgS1IjDHRJaoSBLkmNMNAlqREGuiQ1wkCXpEYY6JLUCANdkhphoEtSIwx0SWqEgS5JjTDQJakRBrokNcJAl6RGGOiS1AgDXZIaYaBLUiMMdElqhIEuSY0w0CWpEQa6JDXCQJekRhjoktQIA12SGmGgS1IjDHRJasRQgZ5kY5IdSXYmuXiecf84SSWZHF2JkqRhLBjoSY4ArgDOBDYAm5JsmGXczwIXAl8adZGSpIUNc4V+KrCzqh6oqqeBrcA5s4x7P/BB4IcjrE+SNKRVQ4w5AXi4b30XcFr/gCSvBNZW1WeT/O5cB0qyGdgMMDExwfT09AEXDDBxDFx0yr5FPfdgLbbmpbR3797Dsq5xsR+D7MeglvsxTKDPK8kzgA8Bb11obFVtAbYATE5O1tTU1KLOefnV13HZ3Qdd+qI8eN7UWM47n+npaRbbyxbZj0H2Y1DL/RhmymU3sLZvfU23bb+fBV4GTCd5EDgd2OYLo5J0aA0T6LcC65OclOQo4Fxg2/6dVfVEVR1XVeuqah1wC3B2Vd22JBVLkma1YKBX1T7gAuAG4H7g2qq6N8mlSc5e6gIlScMZaiK6qrYD22dsu2SOsVMHX5Yk6UD5SVFJaoSBLkmNMNAlqREGuiQ1wkCXpEYY6JLUCANdkhphoEtSIwx0SWqEgS5JjTDQJakRBrokNcJAl6RGGOiS1AgDXZIaYaBLUiMMdElqhIEuSY0w0CWpEQa6JDXCQJekRhjoktQIA12SGmGgS1IjDHRJaoSBLkmNGCrQk2xMsiPJziQXz7L/XUnuS3JXkhuTnDj6UiVJ81kw0JMcAVwBnAlsADYl2TBj2B3AZFW9HPg08B9HXagkaX7DXKGfCuysqgeq6mlgK3BO/4CquqmqnupWbwHWjLZMSdJCVg0x5gTg4b71XcBp84w/H/gfs+1IshnYDDAxMcH09PRwVc4wcQxcdMq+RT33YC225qW0d+/ew7KucbEfg+zHoJb7MUygDy3JbwGTwGtn219VW4AtAJOTkzU1NbWo81x+9XVcdvdISx/ag+dNjeW885menmaxvWyR/RhkPwa13I9hUnE3sLZvfU23bUCS1wPvBV5bVT8aTXmSpGENM4d+K7A+yUlJjgLOBbb1D0jyCuDjwNlVtWf0ZUqSFrJgoFfVPuAC4AbgfuDaqro3yaVJzu6G/SdgNfCpJHcm2TbH4SRJS2Soieiq2g5sn7Htkr7l14+4LknSAfKTopLUCANdkhphoEtSIwx0SWqEgS5JjTDQJakRBrokNcJAl6RGGOiS1AgDXZIaYaBLUiMMdElqhIEuSY0w0CWpEQa6JDXCQJekRhjoktQIA12SGmGgS1IjDHRJaoSBLkmNMNAlqREGuiQ1wkCXpEYY6JLUCANdkhphoEtSI4YK9CQbk+xIsjPJxbPsf2aSv+j2fynJupFXKkma14KBnuQI4ArgTGADsCnJhhnDzgceq6qTgQ8DHxx1oZKk+a0aYsypwM6qegAgyVbgHOC+vjHnAL/fLX8a+EiSVFWNsFaNybqLPzuW8z74gTeM5bzScjVMoJ8APNy3vgs4ba4xVbUvyRPA84FH+gcl2Qxs7lb3JtmxmKKB42Ye+1DJ4fm7x9j6sZQOotdN9uMg2I9By70fJ861Y5hAH5mq2gJsOdjjJLmtqiZHUFIT7Mcg+zHIfgxquR/DvCi6G1jbt76m2zbrmCSrgGcD3xtFgZKk4QwT6LcC65OclOQo4Fxg24wx24C3dMtvBL7g/LkkHVoLTrl0c+IXADcARwBXVtW9SS4FbquqbcCfAZ9MshN4lF7oL6WDnrZpjP0YZD8G2Y9BzfYjXkhLUhv8pKgkNcJAl6RGLLtAX+g2BC1KcmWSPUnu6dv2vCSfT/I33Z/P7bYnyR93/bkrySvHV/nSSLI2yU1J7ktyb5ILu+0rsidJjk7y5SRf6frx77vtJ3W34tjZ3ZrjqG5787fqSHJEkjuSXN+tr4heLKtAH/I2BC26Ctg4Y9vFwI1VtR64sVuHXm/Wd4/NwJ8cohoPpX3ARVW1ATgdeEf3fbBSe/Ij4Neq6peBXwE2Jjmd3i04PtzdkuMxerfogJVxq44Lgfv71ldGL6pq2TyAVwM39K2/B3jPuOs6RF/7OuCevvUdwPHd8vHAjm7548Cm2ca1+gCuA37dnhTAzwD/l96nuR8BVnXb//bfDr13rL26W17Vjcu4ax9hD9bQ+4H+a8D1QFZKL5bVFTqz34bghDHVMm4TVfWtbvnbwES3vKJ61P2K/ArgS6zgnnRTDHcCe4DPA18DHq+qfd2Q/q954FYdwP5bdbTiPwP/Gvhpt/58VkgvllugaxbVu7xYce8/TbIa+K/Ab1fV9/v3rbSeVNVPqupX6F2dngq8dLwVjUeSfwjsqarbx13LOCy3QB/mNgQrxXeSHA/Q/bmn274iepTkSHphfnVVfabbvKJ7AlBVjwM30ZtWeE53Kw4Y/JpbvlXHa4CzkzwIbKU37fJHrJBeLLdAH+Y2BCtF/+0W3kJvHnn/9jd37+w4HXiibxqiCUlC79PJ91fVh/p2rcieJHlBkud0y8fQez3hfnrB/sZu2Mx+NHmrjqp6T1Wtqap19PLhC1V1HiulF+OexF/ECx5nAV+lN0f43nHXc4i+5muAbwE/pjf/dz69eb4bgb8B/hfwvG5s6L0T6GvA3cDkuOtfgn78Kr3plLuAO7vHWSu1J8DLgTu6ftwDXNJtfxHwZWAn8Cngmd32o7v1nd3+F437a1iivkwB16+kXvjRf0lqxHKbcpEkzcFAl6RGGOiS1AgDXZIaYaBLUiMMdElqhIEuSY34/wNP/q9F+rtpAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "dataMl.hist(column='cooccurence')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* Coexpression attributes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "max_val_coe = dataMl['coexpression'].max()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "min_val_coe = dataMl['coexpression'].min()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "std_val_coe = dataMl['coexpression'].std()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "mean_val_coe = dataMl['coexpression'].mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "999"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "max_val_coe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "min_val_coe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "97.58623059797844"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "std_val_coe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "60.211073065127"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mean_val_coe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[<AxesSubplot:title={'center':'coexpression'}>]], dtype=object)"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXQAAAEICAYAAABPgw/pAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/Il7ecAAAACXBIWXMAAAsTAAALEwEAmpwYAAAUz0lEQVR4nO3ce7SddX3n8fdHIuAkFkHsKYZIsEU7WVAvnMW1M56MWgO1ZNaqHaGMAsVmdSrVUcYWRqsttrOwM+go0tZMi4xIiUgdyQIGmCJnnI4VISPKTTRAFCIQ5GZDtZD6nT/2k3ZzPMnZOdnJ5vzO+7XWXnkuv72f728/J5/znN9zSVUhSZr7njPqAiRJw2GgS1IjDHRJaoSBLkmNMNAlqREGuiQ1wkCXdpMkJye5btR1qF3xOnRJaoNH6NI0kiwYdQ3SjjLQNVJJliT5XJKHkzyS5ONJnpPkfUm+nWRTkk8l2afvPUcl+VKSx5N8LclEt/yYJN9LsqSbf0WSx5L8bDe/IcnZSe7oln8yyd7duokk9yf5nSQPAp/s6jgryd1dbZcl2a9rv3eST3fLH09yU5Kxbt2pSe5J8rdJ7k1yct/yv+7rxzHd+57o/j2mb91kkg8m+b/d51yXZP9dvDs0xxnoGpkkewBXAt8GlgKLgTXAqd1rOfBSYBHw8e49i4GrgD8A9gP+A/CXSV5UVV8CPgH89yTPAz4N/G5VfaNvsycDbwB+GngZ8L6+dT/VfeZBwCrgt4B/DbwGeDHwGHBB1/YUYB9gCfBC4DeAHyRZCHwMOK6qng8cA9wyTd/36/rxse79HwauSvLCvma/CpwG/CSwZ9dXaduqamQv4EJgE3DbAG0/Qu8/xi3AN4HHR1m7r6Hs/6OBh4EFU5ZfD/xm3/zLgaeBBcDvABdPaX8tcEo3/VxgHXArcA3deaJu3QbgN/rmjwfu7qYngKeAvfvW3wm8tm/+gL46fg34EvBzU2pZCDwO/DLwvCnrTgX+upt+C/CVKev/Bji1m54E3te37jeBa0a9z3w9u1+jPkK/CFgxSMOqeldVvbKqXgmcD3xuF9al3WMJ8O2q2jJl+YvpHbVv9W16ITpG7+j5V7phjseTPA78PL2wpaqepvdzdShwXlVNPet/35TPfXHf/MNV9cO++YOA/9G3nTuBf+jquJjeL5I1Sb6b5I+SPLeqngTeTO+I/YEkV20d8pmhj1vrWdw3/2Df9N/R+0tF2qaRBnpVfRF4tH9Zkp9Ock2SdUn+zzb+M5wEXLpbitSudB/wkmlOQH6XXphu9RJgC/BQ956Lq+oFfa+FVXUu/OOQzAeATwLnJdlrymcvmfK53+2bny78j5uyrb2ramNVPV1Vv19Vy+gNq7wReCtAVV1bVa+n90vmG8B/m6bvU/u4tZ6N07SVBjLqI/TprAZ+q6oOpzdm+Mf9K5McBBwMfGEEtWm4vgI8AJybZGF3ovFYer+s35Xk4CSLgP8EfKY7kv808EtJ3pBkj+49E0kOTBJ6R+d/DpzeffYHp2zz7V3b/YD3Ap/ZTn1/Cvxh9zNHkhclWdlNL09yWHce4Pv0hmJ+lGQsycpuLP3vgc3Aj6b57KuBlyX51SQLkrwZWEbvnII0K8+qQO/+8x4DfDbJLfROcB0wpdmJwOVV9Q+7uTwNWbcPfwn4GeA7wP30hisupDek8UXgXuCH9E5QUlX3ASuB/0hv/P0+4D30fpbfQe8E4u92Qy2nAacl+Rd9m/0L4DrgHuBueidXt+WjwFrguiR/C3wZOLJb91PA5fTC/E7gf3c1Pwd4N70j8EfpnVD9d9P0/RF6R/VnAo8Avw28saq+t90vTdqOkd9YlGQpcGVVHZrkJ4C7qmpqiPe3/yrw9upd0SANLMkG4G1V9VejrkXaFZ5VR+hV9X3g3iS/ApCeV2xd342n70vvagBJUp+RBnqSS+mF88u7mzpOp3ed8OlJvgbcTu/P661OBNZMc+WCJM17Ix9ykSQNx7NqyEWSNHszPoAoyYX0zsZvqqpDp1kfelcDHE/v5odTq+r/zfS5+++/fy1dunSHCwZ48sknWbhw4azeO1fZ5/nBPs8PO9PndevWfa+qXjTtypluJQX+JfBqtnF7Pr0g/59AgKOAGwe5RfXwww+v2brhhhtm/d65yj7PD/Z5ftiZPgM312xv/a9p7uacYiXwqW5bXwZekGSblx1KknaNYTzzeTHPfD7G/d2yB6Y2TLKK3lPsGBsbY3JyclYb3Lx586zfO1fZ5/nBPs8Pu6rPu/Uh/lW1mt6t/YyPj9fExMSsPmdycpLZvneuss/zg32eH3ZVn4dxlctGnvnAowPxAUOStNsNI9DXAm/t7uo8Cniiqn5suEWStGsNctnipfQe/r9/kvvpPZr0uQBV9af0nhp3PLCe3mWLp+2qYiVJ2zZjoFfVSTOsL+DtQ6tIkjQr3ikqSY0w0CWpEbv1ssVhuXXjE5x61lUj2faGc39xJNuVpJl4hC5JjTDQJakRBrokNcJAl6RGGOiS1AgDXZIaYaBLUiMMdElqhIEuSY0w0CWpEQa6JDXCQJekRhjoktQIA12SGmGgS1IjDHRJaoSBLkmNMNAlqREGuiQ1wkCXpEYY6JLUCANdkhphoEtSIwx0SWqEgS5JjTDQJakRBrokNcJAl6RGGOiS1AgDXZIaYaBLUiMGCvQkK5LclWR9krOmWf+SJDck+WqSryc5fvilSpK2Z8ZAT7IHcAFwHLAMOCnJsinN3gdcVlWvAk4E/njYhUqStm+QI/QjgPVVdU9VPQWsAVZOaVPAT3TT+wDfHV6JkqRBpKq23yB5E7Ciqt7Wzb8FOLKqzuhrcwBwHbAvsBB4XVWtm+azVgGrAMbGxg5fs2bNrIre9OgTPPSDWb11px22eJ+RbHfz5s0sWrRoJNseFfs8P9jnHbN8+fJ1VTU+3boFO1XVPzkJuKiqzktyNHBxkkOr6kf9japqNbAaYHx8vCYmJma1sfMvuYLzbh1W6Ttmw8kTI9nu5OQks/2+5ir7PD/Y5+EZZMhlI7Ckb/7Ablm/04HLAKrqb4C9gf2HUaAkaTCDBPpNwCFJDk6yJ72TnmuntPkO8FqAJP+cXqA/PMxCJUnbN2OgV9UW4AzgWuBOelez3J7knCQndM3OBH49ydeAS4FTa6bBeUnSUA00EF1VVwNXT1n2/r7pO4Bjh1uaJGlHeKeoJDXCQJekRhjoktQIA12SGmGgS1IjDHRJaoSBLkmNMNAlqREGuiQ1wkCXpEYY6JLUCANdkhphoEtSIwx0SWqEgS5JjTDQJakRBrokNcJAl6RGGOiS1AgDXZIaYaBLUiMMdElqhIEuSY0w0CWpEQa6JDXCQJekRhjoktQIA12SGmGgS1IjDHRJaoSBLkmNMNAlqREGuiQ1wkCXpEYMFOhJViS5K8n6JGdto82/SXJHktuT/MVwy5QkzWTBTA2S7AFcALweuB+4Kcnaqrqjr80hwNnAsVX1WJKf3FUFS5KmN8gR+hHA+qq6p6qeAtYAK6e0+XXggqp6DKCqNg23TEnSTFJV22+QvAlYUVVv6+bfAhxZVWf0tfk88E3gWGAP4Peq6pppPmsVsApgbGzs8DVr1syq6E2PPsFDP5jVW3faYYv3Gcl2N2/ezKJFi0ay7VGxz/ODfd4xy5cvX1dV49Otm3HIZUALgEOACeBA4ItJDquqx/sbVdVqYDXA+Ph4TUxMzGpj519yBefdOqzSd8yGkydGst3JyUlm+33NVfZ5frDPwzPIkMtGYEnf/IHdsn73A2ur6umqupfe0fohwylRkjSIQQL9JuCQJAcn2RM4EVg7pc3n6R2dk2R/4GXAPcMrU5I0kxkDvaq2AGcA1wJ3ApdV1e1JzklyQtfsWuCRJHcANwDvqapHdlXRkqQfN9BAdFVdDVw9Zdn7+6YLeHf3kiSNgHeKSlIjDHRJaoSBLkmNMNAlqREGuiQ1wkCXpEYY6JLUCANdkhphoEtSIwx0SWqEgS5JjTDQJakRBrokNcJAl6RGGOiS1AgDXZIaYaBLUiMMdElqhIEuSY0w0CWpEQa6JDXCQJekRhjoktQIA12SGmGgS1IjDHRJaoSBLkmNMNAlqREGuiQ1wkCXpEYY6JLUCANdkhphoEtSIwx0SWrEQIGeZEWSu5KsT3LWdtr9cpJKMj68EiVJg5gx0JPsAVwAHAcsA05Ksmyads8H3gncOOwiJUkzG+QI/QhgfVXdU1VPAWuAldO0+yDwIeCHQ6xPkjSgBQO0WQzc1zd/P3Bkf4MkrwaWVNVVSd6zrQ9KsgpYBTA2Nsbk5OQOFwww9jw487Ats3rvzpptzTtr8+bNI9v2qNjn+cE+D88ggb5dSZ4DfBg4daa2VbUaWA0wPj5eExMTs9rm+ZdcwXm37nTps7Lh5ImRbHdycpLZfl9zlX2eH+zz8Awy5LIRWNI3f2C3bKvnA4cCk0k2AEcBaz0xKkm71yCBfhNwSJKDk+wJnAis3bqyqp6oqv2ramlVLQW+DJxQVTfvkoolSdOaMdCragtwBnAtcCdwWVXdnuScJCfs6gIlSYMZaCC6qq4Grp6y7P3baDux82VJknaUd4pKUiMMdElqhIEuSY0w0CWpEQa6JDXCQJekRhjoktQIA12SGmGgS1IjDHRJaoSBLkmNMNAlqREGuiQ1wkCXpEYY6JLUCANdkhphoEtSIwx0SWqEgS5JjTDQJakRBrokNcJAl6RGGOiS1AgDXZIaYaBLUiMMdElqhIEuSY0w0CWpEQa6JDXCQJekRhjoktQIA12SGmGgS1IjDHRJasRAgZ5kRZK7kqxPctY069+d5I4kX09yfZKDhl+qJGl7Zgz0JHsAFwDHAcuAk5Ism9Lsq8B4Vf0ccDnwR8MuVJK0fYMcoR8BrK+qe6rqKWANsLK/QVXdUFV/181+GThwuGVKkmaSqtp+g+RNwIqqels3/xbgyKo6YxvtPw48WFV/MM26VcAqgLGxscPXrFkzq6I3PfoED/1gVm/daYct3mck2928eTOLFi0aybZHxT7PD/Z5xyxfvnxdVY1Pt27BTlU1RZJ/C4wDr5lufVWtBlYDjI+P18TExKy2c/4lV3DerUMtfWAbTp4YyXYnJyeZ7fc1V9nn+cE+D88gqbgRWNI3f2C37BmSvA54L/Caqvr74ZQnSRrUIGPoNwGHJDk4yZ7AicDa/gZJXgV8AjihqjYNv0xJ0kxmDPSq2gKcAVwL3AlcVlW3JzknyQlds/8MLAI+m+SWJGu38XGSpF1koIHoqroauHrKsvf3Tb9uyHVJknaQd4pKUiMMdElqhIEuSY0w0CWpEQa6JDXCQJekRozm/vk5bOlZV41kuxetWDiS7UqaOzxCl6RGGOiS1AgDXZIaYaBLUiMMdElqhIEuSY0w0CWpEQa6JDXCQJekRhjoktQIA12SGmGgS1IjDHRJaoSBLkmNMNAlqREGuiQ1wkCXpEYY6JLUCANdkhphoEtSIwx0SWqEgS5JjVgw6gI0mFs3PsGpZ101km1vOPcXR7JdSTvGI3RJaoSBLkmNMNAlqREGuiQ1YqCToklWAB8F9gD+rKrOnbJ+L+BTwOHAI8Cbq2rDcEvVqCwd0cnYi1YsHMl2pblqxiP0JHsAFwDHAcuAk5Ism9LsdOCxqvoZ4CPAh4ZdqCRp+wY5Qj8CWF9V9wAkWQOsBO7oa7MS+L1u+nLg40lSVTXEWjXPjPJSzVE587At9nk3afFy3EECfTFwX9/8/cCR22pTVVuSPAG8EPhef6Mkq4BV3ezmJHfNpmhg/6mf3bp32Od5wT7vPhntOMLO9Pmgba3YrTcWVdVqYPXOfk6Sm6tqfAglzRn2eX6wz/PDrurzIFe5bASW9M0f2C2btk2SBcA+9E6OSpJ2k0EC/SbgkCQHJ9kTOBFYO6XNWuCUbvpNwBccP5ek3WvGIZduTPwM4Fp6ly1eWFW3JzkHuLmq1gJ/DlycZD3wKL3Q35V2ethmDrLP84N9nh92SZ/jgbQktcE7RSWpEQa6JDVizgV6khVJ7kqyPslZo65nWJIsSXJDkjuS3J7knd3y/ZL8ryTf6v7dt1ueJB/rvoevJ3n1aHswO0n2SPLVJFd28wcnubHr12e6E/Ek2aubX9+tXzrSwmcpyQuSXJ7kG0nuTHL0PNjH7+p+pm9LcmmSvVvcz0kuTLIpyW19y3Z43yY5pWv/rSSnTLetbZlTgT7gYwjmqi3AmVW1DDgKeHvXt7OA66vqEOD6bh5638Eh3WsV8Ce7v+SheCdwZ9/8h4CPdI+ReIzeYyWgncdLfBS4pqp+FngFvb43u4+TLAbeAYxX1aH0Lqw4kTb380XAiinLdmjfJtkP+AC9mzePAD6w9ZfAQKpqzryAo4Fr++bPBs4edV27qK9XAK8H7gIO6JYdANzVTX8COKmv/T+2mysvevc0XA/8K+BKIPTunlswdX/Tu8rq6G56Qdcuo+7DDvZ3H+DeqXU3vo+33kW+X7ffrgTe0Op+BpYCt8123wInAZ/oW/6MdjO95tQROtM/hmDxiGrZZbo/M18F3AiMVdUD3aoHgbFuuoXv4r8Cvw38qJt/IfB4VW3p5vv79IzHSwBbHy8xlxwMPAx8shtm+rMkC2l4H1fVRuC/AN8BHqC339bR9n7ut6P7dqf2+VwL9OYlWQT8JfDvq+r7/euq9yu7ietMk7wR2FRV60Zdy260AHg18CdV9SrgSf7pT3CgrX0M0A0XrKT3y+zFwEJ+fFhiXtgd+3auBfogjyGYs5I8l16YX1JVn+sWP5TkgG79AcCmbvlc/y6OBU5IsgFYQ2/Y5aPAC7rHR8Az+9TC4yXuB+6vqhu7+cvpBXyr+xjgdcC9VfVwVT0NfI7evm95P/fb0X27U/t8rgX6II8hmJOShN4dt3dW1Yf7VvU/VuEUemPrW5e/tTtbfhTwRN+fds96VXV2VR1YVUvp7ccvVNXJwA30Hh8BP97fOf14iap6ELgvycu7Ra+l9xjqJvdx5zvAUUn+WfczvrXPze7nKXZ0314L/EKSfbu/bn6hWzaYUZ9EmMVJh+OBbwJ3A+8ddT1D7NfP0/tz7OvALd3reHrjh9cD3wL+Ctivax96V/zcDdxK7yqCkfdjln2fAK7spl8KfAVYD3wW2Ktbvnc3v75b/9JR1z3Lvr4SuLnbz58H9m19HwO/D3wDuA24GNirxf0MXErvPMHT9P4aO302+xb4ta7/64HTdqQGb/2XpEbMtSEXSdI2GOiS1AgDXZIaYaBLUiMMdElqhIEuSY0w0CWpEf8fSGCLZO6G7V4AAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "dataMl.hist(column='coexpression')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* Experimental attributes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "max_val_exp = dataMl['experimental'].max()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "min_val_exp = dataMl['experimental'].min()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "std_val_exp = dataMl['experimental'].std()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "mean_val_exp = dataMl['experimental'].mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "999"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "max_val_exp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "min_val_exp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "118.82915600957516"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "std_val_exp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "68.34279346643135"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mean_val_exp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[<AxesSubplot:title={'center':'experimental'}>]], dtype=object)"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAWoAAAEICAYAAAB25L6yAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/Il7ecAAAACXBIWXMAAAsTAAALEwEAmpwYAAAVYUlEQVR4nO3da5Rdd3nf8e8PGWzQEBnHMDWChZwCJtReXDwUCLSdMYEamyZ5QRI73ONWTVvApKSJCSVAW1JnFYe6DgEcAi6J8AQMxNSkXOowELKCQQKDfMHB2AqWAQkQCI+rBgxPX5w9MBYjz5nLOeevOd/PWrN0zr6c/TyztX6z57/3np2qQpLUrnuNugBJ0j0zqCWpcQa1JDXOoJakxhnUktQ4g1qSGmdQa8NK8uYkrxp1HctJMp1k76jrULuOGXUB0qBU1a+NYrtJXgM8vKqeO4rta+PxiFobUpJNo65BWi8GtQYiyYOTvCfJ15PcmuSlSU5IsjfJv+iWmUhyc5Lnd+8v64YrPpLkjiQfS/KwRZ/5qG7egSQ3JfmlRfMuS/KmJH+R5E5gppv2X7r50922fzPJ/iRfTfILSc5K8rfdZ/72os+7V5ILknwpyTeTvCvJCd28bUkqyQuSfDnJN5K8spt3JvDbwC8nmU/yuW76i5Lc2PV1S5J/PfCdoA3DoNa6S3Iv4H8BnwO2Ak8DXgY8AfhV4I+SPAh4A3BtVb1j0erPAf4zcCJwLbCj+8zNwEeAdwIPAs4B/jDJoxet+yvA64D7A59YorR/ABzX1fQ7wB8BzwVOB/4J8KokJ3fLvgT4BeCfAQ8GvgW88bDPeypwStff7yT56ar6IPC7wJ9V1URVPaZbdj/wLOAngBcBb0jy+CN8C6W7GVhQJ3lbd+RyXZ/L/1KSG5Jcn+Sdg6pLQ/EE4IFV9Z+q6rtVdQu9UDynqj4MvBu4GjgLOPzI8gNV9fGq+nvglcCTkzyUXsjtqaq3V9VdVfVZ4D3ALy5a98qq+uuq+kFV/b8l6voe8Lqq+h4wS++HwcVVdUdVXQ/cACwE668Br6yqvV0trwGenWTxeZ3XVtWhqvocvR9Kj+EIquoDVfWl6vkY8GF6PxykZQ3yZOJlwB8A71hmOZI8AngF8JSq+lZ3tKWj18OAByf59qJpm4C/6l5fCrwY+N2q+uZh69628KKq5pMcoHdE+zDgiYd95jHAnyy17hF8s6q+370+1P27b9H8Q8DEoh7el+QHi+Z/H5hc9P5ri17/30Xr/pgkzwReDTyS3gHS/YDdy9QrAQM8oq6qjwMHFk9L8g+TfDDJriR/leRR3ax/Bbyxqr7Vrbt/UHVpKG4Dbq2q4xd93b+qzupO8l1K7wf4v03y8MPWfejCiyQTwAnAV7rP/NhhnzlRVf9m0brr+acgbwOeedj2jquq2/tY9251JDmW3tH/64HJqjoe+Asg61ivNrBhj1FfCrykqk4HfgP4w276I4FHJvnrJJ/sTsjo6PUp4I4kv5Xkvkk2JTk1yRPonWgremPV/w14x2FXaJyV5KlJ7kNvrPqTVXUbcBW9/yPPS3Lv7usJSX56QD28GXjdwsnMJA9M8vN9rrsP2NaN1QPcBzgW+DpwV3d0/Yz1Llgb19CCujs6+hng3UmuBd4CnNTNPgZ4BDANnEvvZNPxw6pN66sbXngW8FjgVuAbwFuBM4B/Dzy/W+b36IX2BYtWfye9IYID9E7yPbf7zDvohds59I6wv9atf+yA2rgYeD/w4SR3AJ8Entjnuu/u/v1mks90tb8UeBe9k5K/0n221JcM8sEBSbYBV1XVqUl+Aripqk5aYrk3A9dU1du791cDF1TVpwdWnJqT5DJgb1X9x1HXIrVkaEfUVfUd4NYkvwiQnoWz5H9O72iaJCfSGwq5ZVi1SVLLBnl53uXA3wCndDcanEfvGtnzupsArgcWxvw+RO/XxBuAjwL/YYmrASRpLA106EOStHbemShJjRvIDS8nnnhibdu2bVXr3nnnnWzevHl9C2qcPY8He9741tLvrl27vlFVD1xq3kCCetu2bezcuXNV687NzTE9Pb2+BTXOnseDPW98a+k3yd8daZ5DH5LUOINakhpnUEtS4wxqSWqcQS1JjTOoJalxBrUkNc6glqTGGdSS1LhBPjNxVXbffpAXXvCBoW93z4VnD32bktQPj6glqXEGtSQ1zqCWpMYZ1JLUuL6COsmvJ7k+yXVJLk9y3KALkyT1LBvUSbbSe9T9VFWdCmwCzhl0YZKknn6HPo4B7pvkGOB+wFcGV5IkabG+Hm6b5HzgdcAh4MNV9ZwlltkObAeYnJw8fXZ2dlUF7T9wkH2HVrXqmpy2dcvwN9qZn59nYmJiZNsfBXseD+PW81r6nZmZ2VVVU0vNWzaokzwAeA/wy8C3gXcDV1TVnx5pnampqVrto7gu2XElF+0e/n04o7zhZdweVwT2PC7Grec1PorriEHdz9DHzwK3VtXXq+p7wHuBn1lVJZKkFesnqL8MPCnJ/ZIEeBpw42DLkiQtWDaoq+oa4ArgM8Dubp1LB1yXJKnT12BwVb0aePWAa5EkLcE7EyWpcQa1JDXOoJakxhnUktQ4g1qSGmdQS1LjDGpJapxBLUmNM6glqXEGtSQ1zqCWpMYZ1JLUOINakhpnUEtS4wxqSWqcQS1JjVs2qJOckuTaRV/fSfKyIdQmSaKPJ7xU1U3AYwGSbAJuB9432LIkSQtWOvTxNOBLVfV3gyhGkvTjUlX9L5y8DfhMVf3BEvO2A9sBJicnT5+dnV1VQfsPHGTfoVWtuianbd0y/I125ufnmZiYGNn2R8Gex8O49byWfmdmZnZV1dRS8/oO6iT3Ab4C/KOq2ndPy05NTdXOnTtXXCjAJTuu5KLdfT1zd13tufDsoW9zwdzcHNPT0yPb/ijY83gYt57X0m+SIwb1SoY+nknvaPoeQ1qStL5WEtTnApcPqhBJ0tL6Cuokm4GnA+8dbDmSpMP1NRhcVXcCPzngWiRJS/DORElqnEEtSY0zqCWpcQa1JDXOoJakxhnUktQ4g1qSGmdQS1LjDGpJapxBLUmNM6glqXEGtSQ1zqCWpMYZ1JLUOINakhpnUEtS4/p9wsvxSa5I8oUkNyZ58qALkyT19Pu474uBD1bVs7unkd9vgDVJkhZZNqiTbAH+KfBCgKr6LvDdwZYlSVqQqrrnBZLHApcCNwCPAXYB53fPUVy83HZgO8Dk5OTps7Ozqypo/4GD7Du0qlXX5LStW4a/0c78/DwTExMj2/4o2PN4GLee19LvzMzMrqqaWmpeP0E9BXwSeEpVXZPkYuA7VfWqI60zNTVVO3fuXFWxl+y4kot29zsis372XHj20Le5YG5ujunp6ZFtfxTseTyMW89r6TfJEYO6n5OJe4G9VXVN9/4K4PGrqkSStGLLBnVVfQ24Lckp3aSn0RsGkSQNQb9jDC8BdnRXfNwCvGhwJUmSFusrqKvqWmDJsRNJ0mB5Z6IkNc6glqTGGdSS1DiDWpIaZ1BLUuMMaklqnEEtSY0zqCWpcQa1JDXOoJakxhnUktQ4g1qSGmdQS1LjDGpJapxBLUmN6+vvUSfZA9wBfB+460jP9ZIkrb+VPEV2pqq+MbBKJElLcuhDkhrXb1AX8OEku5JsH2RBkqS7S1Utv1CytapuT/Ig4CPAS6rq44ctsx3YDjA5OXn67Ozsqgraf+Ag+w6tatU1OW3rluFvtDM/P8/ExMTItj8K9jwexq3ntfQ7MzOz60jn//oK6rutkLwGmK+q1x9pmampqdq5c+eKPnfBJTuu5KLdKxk6Xx97Ljx76NtcMDc3x/T09Mi2Pwr2PB7Gree19JvkiEG97NBHks1J7r/wGngGcN2qKpEkrVg/h66TwPuSLCz/zqr64ECrkiT90LJBXVW3AI8ZQi2SpCV4eZ4kNc6glqTGGdSS1DiDWpIaZ1BLUuMMaklqnEEtSY0zqCWpcQa1JDXOoJakxhnUktQ4g1qSGmdQS1LjDGpJapxBLUmNM6glqXF9B3WSTUk+m+SqQRYkSbq7lRxRnw/cOKhCJElL6yuokzwEOBt462DLkSQdLlW1/ELJFcB/Be4P/EZVPWuJZbYD2wEmJydPn52dXVVB+w8cZN+hVa26Jqdt3TL8jXbm5+eZmJgY2fZHwZ7Hw7j1vJZ+Z2ZmdlXV1FLzln24bZJnAfuraleS6SMtV1WXApcCTE1N1fT0ERe9R5fsuJKLdvfzcPT1tec500Pf5oK5uTlW+/06WtnzeBi3ngfVbz9DH08Bfi7JHmAWOCPJn657JZKkJS0b1FX1iqp6SFVtA84B/rKqnjvwyiRJgNdRS1LzVjQYXFVzwNxAKpEkLckjaklqnEEtSY0zqCWpcQa1JDXOoJakxhnUktQ4g1qSGmdQS1LjDGpJapxBLUmNM6glqXEGtSQ1zqCWpMYZ1JLUOINakhpnUEtS45YN6iTHJflUks8luT7Ja4dRmCSpp58nvPw9cEZVzSe5N/CJJP+7qj454NokSfQR1FVVwHz39t7dVw2yKEnSj6SXw8sslGwCdgEPB95YVb+1xDLbge0Ak5OTp8/Ozq6qoP0HDrLv0KpWXZPTtm4Z/kY78/PzTExMjGz7o2DP42Hcel5LvzMzM7uqamqpeX0F9Q8XTo4H3ge8pKquO9JyU1NTtXPnzpXWCcAlO67kot0reubuuthz4dlD3+aCubk5pqenR7b9UbDn8TBuPa+l3yRHDOoVXfVRVd8GPgqcuapKJEkr1s9VHw/sjqRJcl/g6cAXBlyXJKnTzxjDScD/7Map7wW8q6quGmxZkqQF/Vz18XngcUOoRZK0BO9MlKTGGdSS1DiDWpIaZ1BLUuOGf2dJo7Zd8IGRbfuyMzePbNuS2ucRtSQ1zqCWpMYZ1JLUOINakhpnUEtS4wxqSWqcQS1JjTOoJalxBrUkNc6glqTGGdSS1Lh+HsX10CQfTXJDkuuTnD+MwiRJPf38Uaa7gJdX1WeS3B/YleQjVXXDgGuTJNHHEXVVfbWqPtO9vgO4Edg66MIkST2pqv4XTrYBHwdOrarvHDZvO7AdYHJy8vTZ2dlVFbT/wEH2HVrVqketk7dsYmJiYtRlDNX8/Lw9j4Fx63kt/c7MzOyqqqml5vUd1EkmgI8Br6uq997TslNTU7Vz584VFwpwyY4ruWj3eP2Z7MvO3Mz09PSoyxiqubk5ex4D49bzWvpNcsSg7uuqjyT3Bt4D7FgupCVJ66ufqz4C/DFwY1X9/uBLkiQt1s8R9VOA5wFnJLm2+zprwHVJkjrLDgZX1SeADKEWSdISvDNRkhpnUEtS4wxqSWqcQS1JjTOoJalxBrUkNc6glqTGGdSS1DiDWpIaZ1BLUuMMaklqnEEtSY0zqCWpcQa1JDXOoJakxhnUktS4fh7F9bYk+5NcN4yCJEl3188R9WXAmQOuQ5J0BMsGdVV9HDgwhFokSUtIVS2/ULINuKqqTr2HZbYD2wEmJydPn52dXVVB+w8cZN+hVa161Dp5yyYmJiZGXcZQzc/P2/MYGLee19LvzMzMrqqaWmresg+37VdVXQpcCjA1NVXT09Or+pxLdlzJRbvXrayjwmVnbma136+j1dzcnD2PgXHreVD9etWHJDXOoJakxvVzed7lwN8ApyTZm+S8wZclSVqw7GBwVZ07jEIkSUtz6EOSGmdQS1LjDGpJapxBLUmNM6glqXEGtSQ1zqCWpMYZ1JLUOINakho3Xn+mrlG7bz/ICy/4wNC3u+fCs4e+TUkr5xG1JDXOoJakxhnUktQ4x6jH2LYRjIsvuOzMzSPbtnS08YhakhpnUEtS4/oK6iRnJrkpyc1JLhh0UZKkH1l2jDrJJuCNwNOBvcCnk7y/qm4YdHHauLx2XOpfPycT/zFwc1XdApBkFvh5wKDWUWeUJ1BfftpdI/nhNErj1vOgTpKnqu55geTZwJlV9S+7988DnlhVLz5sue3A9u7tKcBNq6zpROAbq1z3aGXP48GeN7619PuwqnrgUjPW7fK8qroUuHStn5NkZ1VNrUNJRw17Hg/2vPENqt9+TibeDjx00fuHdNMkSUPQT1B/GnhEkpOT3Ac4B3j/YMuSJC1Yduijqu5K8mLgQ8Am4G1Vdf0Aa1rz8MlRyJ7Hgz1vfAPpd9mTiZKk0fLORElqnEEtSY1rJqg36m3qSR6a5KNJbkhyfZLzu+knJPlIki92/z6gm54k/6P7Pnw+yeNH28HqJdmU5LNJruren5zkmq63P+tOTpPk2O79zd38bSMtfJWSHJ/kiiRfSHJjkidv9P2c5Ne7/9fXJbk8yXEbbT8neVuS/UmuWzRtxfs1yQu65b+Y5AUrqaGJoF50m/ozgUcD5yZ59GirWjd3AS+vqkcDTwL+XdfbBcDVVfUI4OruPfS+B4/ovrYDbxp+yevmfODGRe9/D3hDVT0c+BZwXjf9POBb3fQ3dMsdjS4GPlhVjwIeQ6/3Dbufk2wFXgpMVdWp9C42OIeNt58vA848bNqK9muSE4BXA0+kd7f3qxfCvS9VNfIv4MnAhxa9fwXwilHXNaBer6T3d1NuAk7qpp0E3NS9fgtw7qLlf7jc0fRF73r7q4EzgKuA0Ltj65jD9zm9K4qe3L0+plsuo+5hhf1uAW49vO6NvJ+BrcBtwAndfrsK+OcbcT8D24DrVrtfgXOBtyyafrfllvtq4oiaH+3wBXu7aRtK96ve44BrgMmq+mo362vAZPd6o3wv/jvwm8APuvc/CXy7qu7q3i/u64c9d/MPdssfTU4Gvg68vRvueWuSzWzg/VxVtwOvB74MfJXeftvFxt7PC1a6X9e0v1sJ6g0vyQTwHuBlVfWdxfOq9yN2w1wnmeRZwP6q2jXqWoboGODxwJuq6nHAnfzo12FgQ+7nB9D7A20nAw8GNvPjQwQb3jD2aytBvaFvU09yb3ohvaOq3ttN3pfkpG7+ScD+bvpG+F48Bfi5JHuAWXrDHxcDxydZuMlqcV8/7LmbvwX45jALXgd7gb1VdU33/gp6wb2R9/PPArdW1der6nvAe+nt+428nxesdL+uaX+3EtQb9jb1JAH+GLixqn5/0az3Awtnfl9Ab+x6Yfrzu7PHTwIOLvoV66hQVa+oqodU1TZ6+/Ivq+o5wEeBZ3eLHd7zwvfi2d3yR9WRZ1V9DbgtySndpKfR+1PAG3Y/0xvyeFKS+3X/zxd63rD7eZGV7tcPAc9I8oDuN5FndNP6M+pB+kWD62cBfwt8CXjlqOtZx76eSu/Xos8D13ZfZ9Ebm7sa+CLwf4ATuuVD7wqYLwG76Z1RH3kfa+h/Griqe/1TwKeAm4F3A8d204/r3t/czf+pUde9yl4fC+zs9vWfAw/Y6PsZeC3wBeA64E+AYzfafgYupzcG/z16vzmdt5r9Cvxq1/vNwItWUoO3kEtS41oZ+pAkHYFBLUmNM6glqXEGtSQ1zqCWpMYZ1JLUOINakhr3/wHspq4w8R67vgAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "dataMl.hist(column='experimental')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* Database Attributes: For this particular attribute, I will leave it as unique values with its frequencies, "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([  0, 900, 650, 800, 600, 540, 360, 720])"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataMl['database'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0      11053014\n",
       "900      604362\n",
       "600       30592\n",
       "540       17986\n",
       "650       17972\n",
       "800       16150\n",
       "720        9772\n",
       "360        9606\n",
       "Name: database, dtype: int64"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataMl['database'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* Textmining attributes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "max_val_txt = dataMl['textmining'].max()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [],
   "source": [
    "min_val_txt = dataMl['textmining'].min()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [],
   "source": [
    "std_val_txt = dataMl['textmining'].std()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "mean_val_txt = dataMl['textmining'].mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "993"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "max_val_txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "min_val_txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "149.22899651855138"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "std_val_txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "170.136070603278"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mean_val_txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[<AxesSubplot:title={'center':'textmining'}>]], dtype=object)"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXQAAAEICAYAAABPgw/pAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/Il7ecAAAACXBIWXMAAAsTAAALEwEAmpwYAAAX1klEQVR4nO3dfZRcdX3H8fdHkgJlbYJit+kSXSxRjyVHMVMexJ7OYtEIFP4oVjiohGK3ekSwJ7YktgWlp8fQ02gRqJgKopiyWFSSBhQR2CptRXYR8ii6alqSYhayEB0M1Oi3f8yNHdfZzJ3Z2b2Z33xe58zJffjNvb/v3s1n79y5D4oIzMys8z2v6A6YmVl7ONDNzBLhQDczS4QD3cwsEQ50M7NEONDNzBLhQDc7AEnnS/pyu9uazQT5PHQ7GEnaDrwjIr5yMCzHrBN4D93MLBEOdDvoSLoZeDHwr5Iqkv5C0kmS/kPS05IekVTO2r5W0pOSFmbjr5L0lKRXTLGcfkkh6UJJj2Vt3ynpdyRtzJZ/bU1flkm6v2Y8svbfydpeJ0kttD1E0uqs79+XdHHWfs7M/4QtWRHhl18H3QvYDvx+NtwH7AZOp7oTclo2/qJs/t8C9wKHA5uAi+stJxvvBwK4HjgMeAPwLHA78OvZusaB38vaLwPur3l/ABuA+VT/WDwBLG2h7TuBrcDRwJHAV7L2c4r+2fvVua9C99Al3ShpXNLmnO3/SNJWSVsk/fNM988OGm8F7oyIOyPiZxFxNzBCNeABPgDMA74B7ASuy7HMv4mIZyPiy8AzwC0RMR4RO4GvAccf4L2rIuLpiPhv4D7g1S20/SPg6ojYERFPAaty9NnsgIo+5HITsDRPQ0mLgJXAKRHx28B7Z65bdpB5CfDm7LDF05KeBl4HLACIiJ9Q/V06DlgdEXm+6d9VM7y3znjPAd77g5rhH7fY9jeBx2rm1Q6btaTQQI+IrwITtdMk/ZakL0kalfQ1Sa/IZv0JcF22N0NEjM9yd2121YbyY8DNETG/5nVERKwCkNQHXAF8Elgt6dAplnMweZzq4Zb9FhbVEUtH0Xvo9awB3hMRS4D3Af+YTX8Z8DJJ/y7p65Jy7dlbx9oFvDQb/gzwB5LemH2ZeJiksqSjsy8ZbwJuAC6iGpR/M8VyDiafBS6V1CdpPnBZwf2xBBxUgS6pB3gt8C+SHgY+TvaxGpgDLALKwHnAP2X/ESxNHwL+Kju88hbgbOD9VL9YfAz4c6q/v5dQ/TLzr7NDLRcCF0r63cnLkfS+2S3hgP4J+DKwEfgmcCewD/hpkZ2yzlb4hUWS+oENEXGcpF8DHo2IBXXaXQ88EBGfzMbvAVZExIOz2mGzGSDpTcD1EfGSovtineug2kOPiB8C35f0ZgBVvSqbfTvVvXMkHUX1EMz3Cuim2bRJOlzS6ZLm1HwH8IWi+2WdrejTFm8B/hN4uaQdki4CzgcukvQIsIXqR22Au4DdkrZSPf3rzyNidxH9NmsDAR8EnqJ6yGUbcHmhPbKOV/ghFzMza4+D6pCLmZm1rrD7Rhx11FHR39/f0nufeeYZjjjiiPZ2qAN0Y92uuXt0Y92t1Dw6OvpkRLyo3rzCAr2/v5+RkZGW3js8PEy5XG5vhzpAN9btmrtHN9bdSs2S/muqeT7kYmaWCAe6mVkiHOhmZolwoJuZJcKBbmaWCAe6mVkiHOhmZolwoJuZJcKBbmaWiMKuFJ2OTTv3sGzFHYWse/uqMwpZr5lZI95DNzNLhAPdzCwRDnQzs0Q40M3MEuFANzNLhAPdzCwRDnQzs0Q40M3MEuFANzNLRO5Al3SIpG9K2lBn3qGSbpU0JukBSf1t7aWZmTXUzB76pcC2KeZdBDwVEccCHwGumm7HzMysObkCXdLRwBnAJ6ZocjbwqWz4NuD1kjT97pmZWV6KiMaNpNuADwHPB94XEWdOmr8ZWBoRO7Lx7wInRsSTk9oNAoMAvb29S4aGhlrq9PjEHnbtbemt07a4b14xKwYqlQo9PT2Frb8Irrl7dGPdrdQ8MDAwGhGlevMa3m1R0pnAeESMSio3teZJImINsAagVCpFudza4q5Zu47Vm4q5UeT288uFrBdgeHiYVn9mnco1d49urLvdNec55HIKcJak7cAQcKqkz0xqsxNYCCBpDjAP2N22XpqZWUMNAz0iVkbE0RHRD5wL3BsRb53UbD1wQTZ8Ttam8bEcMzNrm5aPW0i6EhiJiPXADcDNksaACarBb2Zms6ipQI+IYWA4G768ZvqzwJvb2TEzM2uOrxQ1M0uEA93MLBEOdDOzRDjQzcwS4UA3M0uEA93MLBEOdDOzRDjQzcwS4UA3M0uEA93MLBEOdDOzRDjQzcwS4UA3M0uEA93MLBEOdDOzRDQMdEmHSfqGpEckbZH0wTptlkl6QtLD2esdM9NdMzObSp4HXDwHnBoRFUlzgfslfTEivj6p3a0RcXH7u2hmZnk0DPTs2aCVbHRu9vLzQs3MDjLK8yxnSYcAo8CxwHURcdmk+cuADwFPAN8G/iwiHquznEFgEKC3t3fJ0NBQS50en9jDrr0tvXXaFvfNK2bFQKVSoaenp7D1F8E1d49urLuVmgcGBkYjolRvXq5A/3ljaT7wBeA9EbG5ZvoLgUpEPCfpT4G3RMSpB1pWqVSKkZGR3Ouudc3adaze1PLzradl+6ozClkvwPDwMOVyubD1F8E1d49urLuVmiVNGejNPiT6aUn3AUuBzTXTd9c0+wTwd031sIP0r7ijkPUW+YfEzDpDnrNcXpTtmSPpcOA04FuT2iyoGT0L2NbGPpqZWQ559tAXAJ/KjqM/D/hsRGyQdCUwEhHrgUsknQXsAyaAZTPVYTMzqy/PWS4bgePrTL+8ZnglsLK9XTMzs2b4SlEzs0Q40M3MEuFANzNLhAPdzCwRDnQzs0Q40M3MEuFANzNLhAPdzCwRDnQzs0Q40M3MEuFANzNLhAPdzCwRDnQzs0Q40M3MEuFANzNLRJ4nFh0m6RuSHpG0RdIH67Q5VNKtksYkPSCpf0Z6a2ZmU8qzh/4ccGpEvAp4NbBU0kmT2lwEPBURxwIfAa5qay/NzKyhhoEeVZVsdG72iknNzgY+lQ3fBrxektrWSzMza0gRk7O5TqPq80RHgWOB6yLisknzNwNLI2JHNv5d4MSIeHJSu0FgEKC3t3fJ0NBQS50en9jDrr0tvbVjLe6bR6VSoaenp+iuzCrX3D26se5Wah4YGBiNiFK9eXkeEk1E/BR4taT5wBckHRcRm5vqRXU5a4A1AKVSKcrlcrOLAOCatetYvSlX15Ox/fwyw8PDtPoz61SuuXt0Y93trrmps1wi4mngPmDppFk7gYUAkuYA84DdbeifmZnllOcslxdle+ZIOhw4DfjWpGbrgQuy4XOAeyPPsRwzM2ubPMctFgCfyo6jPw/4bERskHQlMBIR64EbgJsljQETwLkz1mMzM6urYaBHxEbg+DrTL68ZfhZ4c3u7ZmZmzfCVomZmiXCgm5klwoFuZpYIB7qZWSIc6GZmiXCgm5klwoFuZpYIB7qZWSIc6GZmiXCgm5klwoFuZpYIB7qZWSIc6GZmiXCgm5klwoFuZpaIPE8sWijpPklbJW2RdGmdNmVJeyQ9nL0ur7csMzObOXmeWLQPWB4RD0l6PjAq6e6I2Dqp3dci4sz2d9HMzPJouIceEY9HxEPZ8I+AbUDfTHfMzMya09QxdEn9VB9H90Cd2SdLekTSFyX9djs6Z2Zm+Ski8jWUeoB/A/42Ij4/ad6vAT+LiIqk04GrI2JRnWUMAoMAvb29S4aGhlrq9PjEHnbtbemtHWtx3zwqlQo9PT1Fd2VWuebu0Y11t1LzwMDAaESU6s3LFeiS5gIbgLsi4sM52m8HShHx5FRtSqVSjIyMNFx3PdesXcfqTXkO/6dj+6ozGB4eplwuF92VWeWau0c31t1KzZKmDPQ8Z7kIuAHYNlWYS/qNrB2STsiWu7upXpqZ2bTk2c09BXgbsEnSw9m09wMvBoiI64FzgHdJ2gfsBc6NvMdyzMysLRoGekTcD6hBm2uBa9vVKTMza56vFDUzS4QD3cwsEQ50M7NEONDNzBLhQDczS0R3XZ3TwfpX3MHyxftYtuKOWV/39lVnzPo6zax53kM3M0uEA93MLBEOdDOzRDjQzcwS4UA3M0uEA93MLBEOdDOzRDjQzcwS4UA3M0uEA93MLBF5HkG3UNJ9krZK2iLp0jptJOmjksYkbZT0mpnprpmZTSXPvVz2Acsj4iFJzwdGJd0dEVtr2rwJWJS9TgQ+lv1rZmazpOEeekQ8HhEPZcM/ArYBfZOanQ18Oqq+DsyXtKDtvTUzsympmWc5S+oHvgocFxE/rJm+AViVPX8USfcAl0XEyKT3DwKDAL29vUuGhoZa6vT4xB527W3prR2t93AKqXtx37zZX2mmUqnQ09NT2PqL0I01Q3fW3UrNAwMDoxFRqjcv9+1zJfUAnwPeWxvmzYiINcAagFKpFOVyuZXFcM3adaze1H13/l2+eF8hdW8/vzzr69xveHiYVn9POlU31gzdWXe7a851loukuVTDfG1EfL5Ok53Awprxo7NpZmY2S/Kc5SLgBmBbRHx4imbrgbdnZ7ucBOyJiMfb2E8zM2sgz+f3U4C3AZskPZxNez/wYoCIuB64EzgdGAN+DFzY9p6amdkBNQz07ItONWgTwLvb1SkzM2uerxQ1M0uEA93MLBEOdDOzRDjQzcwS4UA3M0uEA93MLBEOdDOzRDjQzcwS4UA3M0uEA93MLBEOdDOzRDjQzcwS4UA3M0uEA93MLBEOdDOzROR5YtGNksYlbZ5iflnSHkkPZ6/L299NMzNrJM8Ti24CrgU+fYA2X4uIM9vSIzMza0nDPfSI+CowMQt9MTOzaVD16XENGkn9wIaIOK7OvDLwOWAH8D/A+yJiyxTLGQQGAXp7e5cMDQ211OnxiT3s2tvSWzta7+EUUvfivnmzv9JMpVKhp6ensPUXoRtrhu6su5WaBwYGRiOiVG9enkMujTwEvCQiKpJOB24HFtVrGBFrgDUApVIpyuVySyu8Zu06Vm9qR9c7y/LF+wqpe/v55Vlf537Dw8O0+nvSqbqxZujOuttd87TPcomIH0ZEJRu+E5gr6ahp98zMzJoy7UCX9BuSlA2fkC1z93SXa2ZmzWn4+V3SLUAZOErSDuAKYC5ARFwPnAO8S9I+YC9wbuQ5MG9mZm3VMNAj4rwG86+lelqjmZkVyFeKmpklwoFuZpYIB7qZWSIc6GZmiXCgm5klwoFuZpYIB7qZWSIc6GZmiXCgm5klwoFuZpYIB7qZWSIc6GZmiXCgm5klwoFuZpYIB7qZWSIaBrqkGyWNS9o8xXxJ+qikMUkbJb2m/d00M7NG8uyh3wQsPcD8N1F9KPQiYBD42PS7ZWZmzWoY6BHxVWDiAE3OBj4dVV8H5kta0K4OmplZPsrz+E9J/cCGiDiuzrwNwKqIuD8bvwe4LCJG6rQdpLoXT29v75KhoaGWOj0+sYdde1t6a0frPZxC6l7cN2/2V5qpVCr09PQUtv4idGPN0J11t1LzwMDAaESU6s1r+EzRdoqINcAagFKpFOVyuaXlXLN2Has3zWrXDwrLF+8rpO7t55dnfZ37DQ8P0+rvSafqxpqhO+tud83tOMtlJ7CwZvzobJqZmc2idgT6euDt2dkuJwF7IuLxNizXzMya0PDzu6RbgDJwlKQdwBXAXICIuB64EzgdGAN+DFw4U501M7OpNQz0iDivwfwA3t22HpmZWUt8paiZWSIc6GZmiXCgm5klwoFuZpYIB7qZWSK673JLa1r/ijsKWe/2VWcUsl6zTuU9dDOzRDjQzcwS4UA3M0uEA93MLBEOdDOzRDjQzcwS4UA3M0uEA93MLBEOdDOzROQKdElLJT0qaUzSijrzl0l6QtLD2esd7e+qmZkdSJ4nFh0CXAecBuwAHpS0PiK2Tmp6a0RcPAN9NDOzHPLsoZ8AjEXE9yLif4Eh4OyZ7ZaZmTUrT6D3AY/VjO/Ipk32h5I2SrpN0sK29M7MzHJT9ZGgB2ggnQMsjYh3ZONvA06sPbwi6YVAJSKek/SnwFsi4tQ6yxoEBgF6e3uXDA0NtdTp8Yk97Nrb0ls7Wu/hdFXdi/vmUalU6OnpKbors6oba4burLuVmgcGBkYjolRvXp5APxn4QES8MRtfCRARH5qi/SHARETMO9ByS6VSjIyM5Oj+L7tm7TpWb+q+O/8uX7yv6+ouquYib907PDxMuVwubP1F6ca6W6lZ0pSBnueQy4PAIknHSPoV4Fxg/aQVLKgZPQvY1lQPzcxs2hru+kTEPkkXA3cBhwA3RsQWSVcCIxGxHrhE0lnAPmACWDaDfTYzszpyfZaNiDuBOydNu7xmeCWwsr1dMzOzZvhKUTOzRDjQzcwS4UA3M0uEA93MLBEOdDOzRDjQzcwS4UA3M0tEd11HbpZT/4o7CllvkbccsM7nPXQzs0Q40M3MEuFANzNLhAPdzCwRDnQzs0T4LBezg0j/ijtYvngfy2b5LBufXZMG76GbmSXCgW5mlohch1wkLQWupvrEok9ExKpJ8w8FPg0sAXZTfUj09vZ21cxmSlEXUoEP97RTw0DPHvp8HXAasAN4UNL6iNha0+wi4KmIOFbSucBVwFtmosNmlpb9f0xm+7uDFP+Q5NlDPwEYi4jvAUgaAs4GagP9bOAD2fBtwLWSFBHRxr6ambVNip9K8gR6H/BYzfgO4MSp2mQPld4DvBB4sraRpEFgMButSHq0lU4DR01edje4pAvrds3do5vq1lU/H2yl5pdMNWNWT1uMiDXAmukuR9JIRJTa0KWO0o11u+bu0Y11t7vmPGe57AQW1owfnU2r20bSHGAe1S9HzcxsluQJ9AeBRZKOkfQrwLnA+klt1gMXZMPnAPf6+LmZ2exqeMglOyZ+MXAX1dMWb4yILZKuBEYiYj1wA3CzpDFggmroz6RpH7bpUN1Yt2vuHt1Yd1trlnekzczS4CtFzcwS4UA3M0tExwW6pKWSHpU0JmlF0f1pF0kLJd0naaukLZIuzaa/QNLdkr6T/XtkNl2SPpr9HDZKek2xFbRO0iGSvilpQzZ+jKQHstpuzb6MR9Kh2fhYNr+/0I5Pg6T5km6T9C1J2ySdnPq2lvRn2e/2Zkm3SDostW0t6UZJ45I210xrertKuiBr/x1JF9RbVz0dFeg1tyF4E/BK4DxJryy2V22zD1geEa8ETgLendW2ArgnIhYB92TjUP0ZLMpeg8DHZr/LbXMpsK1m/CrgIxFxLPAU1VtLQM0tJoCPZO061dXAlyLiFcCrqNaf7LaW1AdcApQi4jiqJ1jsv01IStv6JmDppGlNbVdJLwCuoHoB5wnAFfv/CDQUER3zAk4G7qoZXwmsLLpfM1TrOqr3z3kUWJBNWwA8mg1/HDivpv3P23XSi+p1DfcApwIbAFG9cm7O5G1O9Uyrk7PhOVk7FV1DCzXPA74/ue8pb2v+/2ryF2TbbgPwxhS3NdAPbG51uwLnAR+vmf4L7Q706qg9dOrfhqCvoL7MmOzj5fHAA0BvRDyezfoB0JsNp/Kz+AfgL4CfZeMvBJ6OiH3ZeG1dv3CLCWD/LSY6zTHAE8Ans0NNn5B0BAlv64jYCfw98N/A41S33Sjpb2tofru2vL07LdCTJ6kH+Bzw3oj4Ye28qP65TuY8U0lnAuMRMVp0X2bZHOA1wMci4njgGf7/YziQ5LY+kupN/I4BfhM4gl8+NJG8md6unRboeW5D0LEkzaUa5msj4vPZ5F2SFmTzFwDj2fQUfhanAGdJ2g4MUT3scjUwP7uFBPxiXancYmIHsCMiHsjGb6Ma8Clv698Hvh8RT0TET4DPU93+qW9raH67try9Oy3Q89yGoCNJEtUrbrdFxIdrZtXeVuECqsfW909/e/ZN+UnAnpqPdR0hIlZGxNER0U91W94bEecD91G9hQT8cs0df4uJiPgB8Jikl2eTXk/1dtTJbmuqh1pOkvSr2e/6/pqT3taZZrfrXcAbJB2ZfbJ5QzatsaK/QGjhC4fTgW8D3wX+suj+tLGu11H9KLYReDh7nU71uOE9wHeArwAvyNqL6hk/3wU2UT17oPA6plF/GdiQDb8U+AYwBvwLcGg2/bBsfCyb/9Ki+z2Nel8NjGTb+3bgyNS3NfBB4FvAZuBm4NDUtjVwC9XvCH5C9ZPYRa1sV+CPs9rHgAvzrt+X/puZJaLTDrmYmdkUHOhmZolwoJuZJcKBbmaWCAe6mVkiHOhmZolwoJuZJeL/AMBZr16mlK5UAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "dataMl.hist(column='textmining')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* Combined_score attributes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [],
   "source": [
    "max_val_comb = dataMl['combined_score'].max()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [],
   "source": [
    "min_val_comb = dataMl['combined_score'].min()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [],
   "source": [
    "std_val_comb = dataMl['combined_score'].std()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [],
   "source": [
    "mean_val_comb = dataMl['combined_score'].mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "999"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "max_val_comb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "150"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "min_val_comb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "197.15931208847724"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "std_val_comb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "295.8066422131504"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mean_val_comb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[<AxesSubplot:title={'center':'combined_score'}>]], dtype=object)"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAWoAAAEICAYAAAB25L6yAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/Il7ecAAAACXBIWXMAAAsTAAALEwEAmpwYAAAVJklEQVR4nO3da5QkZ33f8e9PWgFiB6+QBWOxCEZcrHDZGNBgILLNjAiyhAi8CA7oEEBYsCGJAZMltshJ7PicOFHOsYzJAROEuMQEGMw1HMlGIaAJwTFSdrmtbg5CEkiLkBCWBCMrhrX/edE1UmuY6enpnZ55Zvf7OafPdlU9VfXvp2t/U/1U9UyqCklSu47a7AIkSYMZ1JLUOINakhpnUEtS4wxqSWqcQS1JjTOoJalxBrWak+TcJF8csPxPk7xqDPudSlJJtq33tqVD4QGpLaeqztrsGqSN5Bm11JgkR292DWqLQa01S3JSkk8k+V6S7yd5e5KjkvzrJN9KcnuSP0qyo2u/OKTw6iQ3J7kzyeuSPDPJ15PcleTtP7mbvD3J3UmuS/K8vgXzSV7TPT83yReT/F633RuTnNXXdkeS9yS5NcmBJP9uMQiTHN2td0eSG4Czh3z95ya5IckPu/29vG/Za5Nc2y27JskzuvlP6uq+K8nVSV7Ut877k7wzyZ8kuQeYTfKoJB/v+vjGJG9Y8xulw0dVjeUBvBe4HbhqyPb/CLgGuBr40Ljq8nHI7+vRwNeAtwLbgYcAvwD8KnA98DhgAvgE8IFunSmggP/ctT8D+H/Ap4BHAju7Y+W5XftzgYPAm4BjgJcCdwPHd8vngdf0tf0x8Nqutn8KfAdIt/yTwLu6Wh8JXAn8k27Z64DrgJOA44HLuzq3DXj924EfAKd00ycCT+me/wpwAHgmEOAJwGO713A98K+ABwGnAz/s28b7u9d3Gr2Tp4cC+4Df6to/DrgB+OXNfv99bM5jfBuGXwKeMUxQA08EvgI8vJt+5GZ3jI8V36vnAN9bGmbA54B/1jd9Sheg2/qCemff8u8DL+2b/jjw693zc/vDtpt3JfCK7vnSoL6+r91Du339DDAJ/DVwbN/yc4DLu+efB17Xt+yMIYP6LuAf9m+3W3YZ8MZl1vlF4LvAUX3zPgz82+75+4E/6lv2LODbS7bxFuB9m/3++9icx9guJlbVF5JM9c9L8njgHcAjgL8CXltV19E7G3pHVd3ZrXv7uOrSITsJ+FZVHVwy/1HAt/qmv0UvpCf75t3W9/zeZaYn+qYPVFX/r3b8VreP5Xx38UlV/VUSum0dT+9s9tZuHvTOWG/uq/nm+zfzgPqXVVX3JHkp8GbgPUn+DNjTHccnAd9cZrVHATdX1d8u2dfOvun+Oh4LPCrJXX3zjgb+12r16fC00WPUFwGvr6pT6R3of9jN/1ngZ5P8WZIvJTlzg+vS8G4GHrPMLWzfoRcwix5Db/jiNkazM33p2m3vO2vcxs30zqhPqKrjusdPVdVTuuW30gvX/n2sqqouq6rn0xv2uA54d9/+Hr/MKt8BTkrS///tMfSGSe7b7JK6b+yr+biqelhVvWCY+nT42bCgTjIB/D3go0m+Sm/c8MRu8TZ6wx8z9D6avjvJcRtVm9bkSnoBd0GS7UkekuQ0eh/l35Tk5O69/vfAR5Y58x7WI4E3JDkmya8ATwL+ZC0bqKpbgf8OXJjkp7oLno9P8tyuyR93+3h0kocD56+2zSSTSV6cZDu9HwILwOKZ8sXAm5Ocmp4nJHkscAW9T5C/0b2eGeAfAHMr7OZK4IdJfjPJsd1Fz6cmeeZaXr8OHxt5Rn0UcFdVPa3v8aRu2S3Ap6vqx1V1I/B/6QW3GlNVf0MvZJ4AfJvee/dSehePPwB8AbiR3sXC1x/Crq6gdwzcAfwu8JKq+v4I23klvQty1wB3Ah/j/hOEd9MbV/4a8GV6F0BXcxTwL+idJf8l8Fx6FzCpqo92tX6I3sXCT9G7APojen12Vvd6/hB4ZTdc8hO6Pn4h8DR6fXkHvR8CO4Z90Tq85IHDgOu88d4Y9SVV9dRu+n8Db62qj3Yfa/9uVX2tG+o4p6peleQEehcWnzbif0xJOqyM7Yw6yYeBPwdOSXJLkvOAlwPnJfkavdvwXtw1vwz4fpJr6N0i9S8NaUnqGesZtbRVJVlYYdFZVeXdF9pQqwZ1klOAj/TNehzwW1X1B2OsS5LUWdMZdffV2wPAs6pqxXtOTzjhhJqamjr06g7BPffcw/bt2ze1hlbZN4PZP4PZP4ON2j/79u27o6oesdyytX7h5XnANweFNMDU1BR79+5d46bX1/z8PDMzM5taQ6vsm8Hsn8Hsn8FG7Z8kK+bqWs+o3wt8uaqW/gIdkuwGdgNMTk6eOje30i2iG2NhYYGJiYnVGx6B7JvB7J/B7J/BRu2f2dnZfVU1vdyyoYM6yYPo3Tv6lKoa+G2z6enp8oy6XfbNYPbPYPbPYIdwRr1iUK/l9ryz6J1Nj/qVYEnSCNYS1OfQ+5qwJGkDDRXU3e81eD7DfcVWkrSOhrrro6ruAX56zLVIkpbhn+KSpMYZ1JLUOINakho3tj/FNaqp8y9dl+3s2XWQc9ewrZsuGOoPUEvShvOMWpIaZ1BLUuMMaklqnEEtSY0zqCWpcQa1JDXOoJakxhnUktQ4g1qSGmdQS1LjDGpJapxBLUmNM6glqXEGtSQ1zqCWpMYZ1JLUOINakho3VFAnOS7Jx5Jcl+TaJM8Zd2GSpJ5h/xTX24DPVNVLkjwIeOgYa5Ik9Vk1qJPsAH4JOBegqn4E/Gi8ZUmSFg0z9HEy8D3gfUm+kuTiJNvHXJckqZOqGtwgmQa+BJxWVVckeRvwg6r6N0va7QZ2A0xOTp46Nzc3UkH7D9w90npLTR4Lt907fPtdO3esy363goWFBSYmJja7jGbZP4PZP4ON2j+zs7P7qmp6uWXDBPXPAF+qqqlu+heB86vq7JXWmZ6err179665UICp8y8dab2l9uw6yIX7hx2Ch5suWPHlHHbm5+eZmZnZ7DKaZf8MZv8MNmr/JFkxqFcd+qiq7wI3Jzmlm/U84Jo1VyFJGsmwp5yvBz7Y3fFxA/Dq8ZUkSeo3VFBX1VeBZU/JJUnj5TcTJalxBrUkNc6glqTGGdSS1DiDWpIaZ1BLUuMMaklqnEEtSY0zqCWpcQa1JDXOoJakxhnUktQ4g1qSGmdQS1LjDGpJapxBLUmNM6glqXEGtSQ1zqCWpMYZ1JLUOINakhpnUEtS47YN0yjJTcAPgb8BDlbV9DiLkiTdb6ig7sxW1R1jq0SStCyHPiSpcamq1RslNwJ3AgW8q6ouWqbNbmA3wOTk5Klzc3MjFbT/wN0jrbfU5LFw273Dt9+1c8e67HcrWFhYYGJiYrPLaJb9M5j9M9io/TM7O7tvpWHlYYN6Z1UdSPJI4LPA66vqCyu1n56err179665UICp8y8dab2l9uw6yIX7hx/ZuemCs9dlv1vB/Pw8MzMzm11Gs+yfweyfwUbtnyQrBvVQQx9VdaD793bgk8DPr7kKSdJIVg3qJNuTPGzxOXAGcNW4C5Mk9QwzNjAJfDLJYvsPVdVnxlqVJOk+qwZ1Vd0A/NwG1CJJWoa350lS4wxqSWqcQS1JjTOoJalxBrUkNc6glqTGGdSS1DiDWpIaZ1BLUuMMaklqnEEtSY0zqCWpcQa1JDXOoJakxhnUktQ4g1qSGmdQS1LjDGpJapxBLUmNM6glqXEGtSQ1zqCWpMYNHdRJjk7ylSSXjLMgSdIDreWM+o3AteMqRJK0vKGCOsmjgbOBi8dbjiRpqVTV6o2SjwH/AXgY8OaqeuEybXYDuwEmJydPnZubG6mg/QfuHmm9pSaPhdvuHb79rp071mW/W8HCwgITExObXUaz7J/B7J/BRu2f2dnZfVU1vdyybautnOSFwO1VtS/JzErtquoi4CKA6enpmplZselA555/6UjrLbVn10Eu3L/qy7vPTS+fWZf9bgXz8/OM+v4cCeyfweyfwcbRP8MMfZwGvCjJTcAccHqS/7quVUiSVrRqUFfVW6rq0VU1BbwM+HxV/eOxVyZJAryPWpKaN/wgLlBV88D8WCqRJC3LM2pJapxBLUmNM6glqXEGtSQ1zqCWpMYZ1JLUOINakhpnUEtS4wxqSWqcQS1JjTOoJalxBrUkNc6glqTGGdSS1DiDWpIaZ1BLUuMMaklqnEEtSY0zqCWpcQa1JDXOoJakxhnUktS4VYM6yUOSXJnka0muTvI7G1GYJKln2xBt/ho4vaoWkhwDfDHJn1bVl8ZcmySJIYK6qgpY6CaP6R41zqIkSfdLL4dXaZQcDewDngC8o6p+c5k2u4HdAJOTk6fOzc2NVND+A3ePtN5Sk8fCbfcO337Xzh3rst+tYGFhgYmJic0uo1n2z2D2z2Cj9s/s7Oy+qppebtlQQX1f4+Q44JPA66vqqpXaTU9P1969e9daJwBT51860npL7dl1kAv3DzOy03PTBWevy363gvn5eWZmZja7jGbZP4PZP4ON2j9JVgzqNd31UVV3AZcDZ665CknSSIa56+MR3Zk0SY4Fng9cN+a6JEmdYcYGTgT+SzdOfRTwx1V1yXjLkiQtGuauj68DT9+AWiRJy/CbiZLUOINakhpnUEtS4wxqSWqcQS1JjTOoJalxBrUkNc6glqTGGdSS1DiDWpIaZ1BLUuMMaklqnEEtSY0zqCWpcQa1JDXOoJakxhnUktQ4g1qSGmdQS1LjDGpJapxBLUmNW/WvkB8pps6/dNP2fdMFZ2/aviW1b9Uz6iQnJbk8yTVJrk7yxo0oTJLUM8wZ9UFgT1V9OcnDgH1JPltV14y5NkkSQ5xRV9WtVfXl7vkPgWuBneMuTJLUk6oavnEyBXwBeGpV/WDJst3AboDJyclT5+bmRipo/4G7R1pvqclj4bZ712VTY7dr544N3d/CwgITExMbus+txP4ZzP4ZbNT+mZ2d3VdV08stGzqok0wA/xP43ar6xKC209PTtXfv3jUXCut3UW/ProNcuH9rXCvd6IuJ8/PzzMzMbOg+txL7ZzD7Z7BR+yfJikE91O15SY4BPg58cLWQliStr2Hu+gjwHuDaqvr98ZckSeo3zBn1acArgNOTfLV7vGDMdUmSOqsO4lbVF4FsQC2SpGX4FXJJapxBLUmNM6glqXEGtSQ1zqCWpMYZ1JLUOINakhpnUEtS4wxqSWqcQS1JjTOoJalxBrUkNc6glqTGGdSS1DiDWpIaZ1BLUuMMaklqnEEtSY0zqCWpcQa1JDXOoJakxhnUktS4VYM6yXuT3J7kqo0oSJL0QMOcUb8fOHPMdUiSVrBqUFfVF4C/3IBaJEnLSFWt3iiZAi6pqqcOaLMb2A0wOTl56tzc3EgF7T9w90jrLTV5LNx277psaux27dyxoftbWFhgYmJiQ/e5ldg/g9k/g43aP7Ozs/uqanq5ZesW1P2mp6dr7969aypy0dT5l4603lJ7dh3kwv3b1mVbh5vFvrnpgrM3u5Qmzc/PMzMzs9llNMv+GWzU/kmyYlB714ckNc6glqTGDXN73oeBPwdOSXJLkvPGX5YkadGqg7hVdc5GFCJJWp5DH5LUOINakhpnUEtS4wxqSWqcQS1JjTOoJalxBrUkNc6glqTG+VuLjmDr9QuwRuEvhJKG5xm1JDXOoJakxhnUktQ4x6gljYXXQNaPZ9SS1DiDWpIa59CHNsVmfSw+3D4S68hgUEs67BxuJwIOfUhS4wxqSWqcQx86ogzzkXjProOcu84fnR0b16HwjFqSGucZtbQB/PKHDsVQQZ3kTOBtwNHAxVV1wVirkrRu1vuHxDiGhjTYqkMfSY4G3gGcBTwZOCfJk8ddmCSpZ5gx6p8Hrq+qG6rqR8Ac8OLxliVJWpSqGtwgeQlwZlW9ppt+BfCsqvq1Je12A7u7yVOAv1j/ctfkBOCOTa6hVfbNYPbPYPbPYKP2z2Or6hHLLVi3i4lVdRFw0Xpt71Al2VtV05tdR4vsm8Hsn8Hsn8HG0T/DDH0cAE7qm350N0+StAGGCer/AzwxyclJHgS8DPj0eMuSJC1adeijqg4m+TXgMnq35723qq4ee2WHrplhmAbZN4PZP4PZP4Ote/+sejFRkrS5/Aq5JDXOoJakxm3JoE5yUpLLk1yT5Ookb+zmH5/ks0m+0f378G5+kvynJNcn+XqSZ2zuKxi/JEcn+UqSS7rpk5Nc0fXBR7oLwyR5cDd9fbd8alML3wBJjkvysSTXJbk2yXM8du6X5E3d/6urknw4yUOO5OMnyXuT3J7kqr55az5ekryqa/+NJK9aSw1bMqiBg8Ceqnoy8Gzgn3dfaz8f+FxVPRH4XDcNva+/P7F77AbeufElb7g3Atf2Tf9H4K1V9QTgTuC8bv55wJ3d/Ld27Q53bwM+U1V/B/g5ev3ksQMk2Qm8AZiuqqfSu4HgZRzZx8/7gTOXzFvT8ZLkeOC3gWfR+7b3by+G+1Cqass/gP8GPJ/etyFP7OadCPxF9/xdwDl97e9rdzg+6N3r/jngdOASIPS+KbWtW/4c4LLu+WXAc7rn27p22ezXMMa+2QHcuPQ1euzc9/p2AjcDx3fHwyXALx/pxw8wBVw16vECnAO8q2/+A9qt9tiqZ9T36T5qPR24Apisqlu7Rd8FJrvniwffolu6eYerPwB+A/jbbvqngbuq6mA33f/67+ubbvndXfvD1cnA94D3dUNDFyfZjscOAFV1APg94NvArfSOh314/Cy11uPlkI6jLR3USSaAjwO/XlU/6F9WvR9bR9y9h0leCNxeVfs2u5ZGbQOeAbyzqp4O3MP9H1uBI/fYAeg+jr+Y3g+0RwHb+cmP/eqzEcfLlg3qJMfQC+kPVtUnutm3JTmxW34icHs3/0j6GvxpwIuS3ETvNx2eTm9M9rgki19w6n/99/VNt3wH8P2NLHiD3QLcUlVXdNMfoxfcHjs9fx+4saq+V1U/Bj5B75jy+HmgtR4vh3QcbcmgThLgPcC1VfX7fYs+DSxeTX0VvbHrxfmv7K7IPhu4u+9jy2Glqt5SVY+uqil6F4E+X1UvBy4HXtI1W9o3i332kq79YXs2WVXfBW5Ocko363nANXjsLPo28OwkD+3+ny32j8fPA631eLkMOCPJw7tPLWd084az2YP0Iw7s/wK9jxpfB77aPV5Ab2zsc8A3gP8BHN+1D70/fvBNYD+9K9qb/jo2oJ9mgEu6548DrgSuBz4KPLib/5Bu+vpu+eM2u+4N6JenAXu74+dTwMM9dh7QP78DXAdcBXwAePCRfPwAH6Y3Xv9jep/IzhvleAF+teun64FXr6UGv0IuSY3bkkMfknQkMaglqXEGtSQ1zqCWpMYZ1JLUOINakhpnUEtS4/4/+fTV1g+9EWkAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "dataMl.hist(column='combined_score')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* Neighbors attributes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [],
   "source": [
    "max_val_neigh = dataMl['neighbors'].max()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [],
   "source": [
    "min_val_neigh = dataMl['neighbors'].min()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [],
   "source": [
    "std_val_neigh = dataMl['neighbors'].std()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [],
   "source": [
    "mean_val_neigh = dataMl['neighbors'].mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "7645"
      ]
     },
     "execution_count": 90,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "max_val_neigh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "min_val_neigh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "790.0866039459232"
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "std_val_neigh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1067.936830230383"
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mean_val_neigh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[<AxesSubplot:title={'center':'neighbors'}>]], dtype=object)"
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXUAAAEICAYAAACgQWTXAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/Il7ecAAAACXBIWXMAAAsTAAALEwEAmpwYAAATkElEQVR4nO3dfbRldX3f8fdHHhTn0gHF3sJAHa2GLOskKLc+lNTeWGNQGP3HuHARK4lxmrRmaTPWQtMnu1Za01XSRGoqEyWxDXJ9iAQKSdQqE6pNUEZBnivKGEBgRB50KHE55Ns/9h65Tgfuvefpbn+8X2vddffZZ5+9P+uccz9n39/e55xUFZKkNjxpvQNIkibHUpekhljqktQQS12SGmKpS1JDLHVJaoilrh86SW5IsrjKZXcnecVjXLeY5I5JZpPW26HrHUBaq6r62+udQRoq99SlCUjiDpIGwVLXuumHRt6R5MtJHkzy4SRP6a87Pck1SR5I8r+T/NgBt3tFP31Ekg8muT/JTUneeZAhlZMOto1l6/sXSe7t13vmsvkbk/y3JN9M8vUk/zLJk/rrzkryuST/Ocm3gH+b5DlJ/rTfzr1JPjyt+056LO5daL29HjgV+Evgc8BZSa4CLgC2AlcDPwtcmuTEqvruAbf/N8Bm4NnABuCPVrMN4H39dX8DOAbYBLwE+KMkV1fVLcB5wMZ+3U8HPgncBXygv+2LgSVgHjisz/xJ4CeBw4GFEe4PaSxT21NPckGSPUmuX+Xyr09yY38Q7EPTyqXBeU9VfaOq7gP+B3ASsA04v6quqqpHquqDwHfpSvdArwf+fVXdX1V3AO9Z5TaW+1dV9d2q+lPgcuD1SQ4BzgDOqarvVNVu4Fzgjctu942qOq+q9lXVw8D3gGcCx1XVX1bVZ0e5Q6RxTHP45ffo9o5WlOS5wDnAKf1BsLdPL5YG5u5l0/8XmKMrxu390MsDSR4ATgCOO8jtjwNuX3b59oMsc7Bt7Hd/VT207PLX+3UeQ7f3/fUDrtv0ONt6JxDg8/3Oyc8fJIs0VVMr9aq6Erhv+bwkfyvJnyTZleR/JfnR/qq3AO+tqvv72+6ZVi79ULgd+LWqOmrZz1Or6qKDLHsXcPyyyyescVtHJ9mw7PLfBL4B3Muje97Lr7tz2eUf+IjTqrq7qt5SVccB/wj47STPWWMeaSyzPlC6A/jlqjoZeAfw2/38HwF+pD/w9OdJVrWHr2b9DvCLSV6czoYkpyU58iDLfgQ4J8nRSTYBbx1he+9KcniSvwecDny0qh7p1/1rSY5M8kzgV4Dff6yVJPmZJPtfYO6nK/2/GiGPNLKZHShNMgf8XeCjSfbPfvKyHM8FFun2uq5MsqWqHphVPg1HVV2d5C3Af6F7XjwMfBa48iCL/zu6g5630e21Xwj83Bo2dzddAX+DbmjmF6vq5v66X6Y7WPo1uoOsv0N3MPSx/B3gN5NsBO4B3lZVX1tDFmlsmeaXZCTZDFxWVc9P8teAW6rq2IMs9z7gqqr63f7yp4Gzq+oLUwunJiX5JeCMqvr7651FWg8zG36pqm8DtyX5GYD+3+of76/+Q7q9dJIcQzcc4x6OVpTk2CSnJHlSkhOB7cDF651LWi/TPKXxIuDPgBOT3JHkzcCZwJuTXAvcALy2X/wTwLeS3AhcAfyzqvrWtLKpKYcD5wPfAT4DXMKjx2qkJ5ypDr9IkmbLjwmQpIZM5eyXY445pjZv3jzSbR966CE2bNiw8oLrZMj5hpwNzDcu841uyNmgy3fzzTffW1XPGHtlVTXxn5NPPrlGdcUVV4x821kYcr4hZ6sy37jMN7ohZ6vq8gFX1wT61+EXSWqIpS5JDbHUJakhlrokNWRVZ78k2U335o5HgH1V5Yf/S9IAreWUxp+sqnunlkSSNDaHXySpIav6mIAkt/Ho50OfX1U7DrLMNrqvIWN+fv7kpaWlkQLt3buXubm5lRdcJ0PON+RsYL5xmW90Q84GXb6tW7fumsjQ9mpOZgc29b//OnAt8LLHW943H62PIWerMt+4zDe6IWermuybj1Y1pl5Vd/a/9yS5GHgRB//CgrFdd+eDnHX25dNY9ePa/e7TZr5NSZq0FcfU+68SO3L/NPBK4PppB5Mkrd1q9tTngYv7r6A7FPhQVf3JVFNJkkayYqlX9x2LP77ScpKk9ecpjZLUEEtdkhpiqUtSQyx1SWqIpS5JDbHUJakhlrokNcRSl6SGWOqS1BBLXZIaYqlLUkMsdUlqiKUuSQ2x1CWpIZa6JDXEUpekhljqktQQS12SGrKa7yh9Qth89uWrWm77ln2ctcplV2v3u0+b6PokPXG5py5JDbHUJakhlrokNcRSl6SGWOqS1BBLXZIaYqlLUkMsdUlqiKUuSQ2x1CWpIZa6JDXEUpekhljqktQQS12SGrLqUk9ySJIvJblsmoEkSaNby57624CbphVEkjS+VZV6kuOB04D3TzeOJGkcqaqVF0o+BvwH4EjgHVV1+kGW2QZsA5ifnz95aWlppEB77nuQex4e6aYzMX8EE8+3ZdPGiaxn7969zM3NTWRd02C+8ZhvdEPOBl2+rVu37qqqhXHXteLX2SU5HdhTVbuSLD7WclW1A9gBsLCwUIuLj7no4zrvwks497rhfsve9i37Jp5v95mLE1nPzp07GfV+nwXzjcd8oxtyNujyTcpqhl9OAV6TZDewBLw8ye9PLIEkaWJWLPWqOqeqjq+qzcAZwGeq6mennkyStGaepy5JDVnT4HBV7QR2TiWJJGls7qlLUkMsdUlqiKUuSQ2x1CWpIZa6JDXEUpekhljqktQQS12SGmKpS1JDLHVJaoilLkkNsdQlqSGWuiQ1xFKXpIZY6pLUEEtdkhpiqUtSQyx1SWqIpS5JDbHUJakhlrokNcRSl6SGWOqS1BBLXZIaYqlLUkMsdUlqiKUuSQ2x1CWpIZa6JDXEUpekhljqktQQS12SGrJiqSd5SpLPJ7k2yQ1J3jWLYJKktTt0Fct8F3h5Ve1Nchjw2SR/XFV/PuVskqQ1WrHUq6qAvf3Fw/qfmmYoSdJo0nX2CgslhwC7gOcA762qf36QZbYB2wDm5+dPXlpaGinQnvse5J6HR7rpTMwfwcTzbdm0cSLr2bt3L3NzcxNZ1zSYbzzmG92Qs0GXb+vWrbuqamHcda1m+IWqegQ4KclRwMVJnl9V1x+wzA5gB8DCwkItLi6OFOi8Cy/h3OtWFWtdbN+yb+L5dp+5OJH17Ny5k1Hv91kw33jMN7ohZ4Mu36Ss6eyXqnoAuAI4dWIJJEkTs5qzX57R76GT5Ajgp4Cbp5xLkjSC1YwjHAt8sB9XfxLwkaq6bLqxJEmjWM3ZL18GXjCDLJKkMfmOUklqiKUuSQ2x1CWpIZa6JDXEUpekhljqktQQS12SGmKpS1JDLHVJaoilLkkNsdQlqSGWuiQ1xFKXpIZY6pLUEEtdkhpiqUtSQyx1SWqIpS5JDbHUJakhlrokNcRSl6SGWOqS1BBLXZIaYqlLUkMsdUlqiKUuSQ2x1CWpIZa6JDXEUpekhljqktQQS12SGmKpS1JDLHVJasiKpZ7khCRXJLkxyQ1J3jaLYJKktTt0FcvsA7ZX1ReTHAnsSvKpqrpxytkkSWu04p56Vd1VVV/sp78D3ARsmnYwSdLapapWv3CyGbgSeH5VffuA67YB2wDm5+dPXlpaGinQnvse5J6HR7rpTMwfwcTzbdm0cSLr2bt3L3NzcxNZ1zSYbzzmG92Qs0GXb+vWrbuqamHcda1m+AWAJHPAHwBvP7DQAapqB7ADYGFhoRYXF0cKdN6Fl3DudauONXPbt+ybeL7dZy5OZD07d+5k1Pt9Fsw3HvONbsjZoMs3Kas6+yXJYXSFfmFVfXxiW5ckTdRqzn4J8AHgpqr6jelHkiSNajV76qcAbwRenuSa/ufVU84lSRrBioPDVfVZIDPIIkkak+8olaSGWOqS1BBLXZIaYqlLUkMsdUlqiKUuSQ2x1CWpIcP9kJUnkM1nXz6R9Wzfso+z1rCu3e8+bSLblTQc7qlLUkMsdUlqiKUuSQ2x1CWpIZa6JDXEUpekhljqktQQS12SGmKpS1JDLHVJaoilLkkNsdQlqSGWuiQ1xFKXpIZY6pLUEEtdkhpiqUtSQyx1SWqIpS5JDbHUJakhlrokNcRSl6SGWOqS1BBLXZIaYqlLUkNWLPUkFyTZk+T6WQSSJI1uNXvqvwecOuUckqQJWLHUq+pK4L4ZZJEkjSlVtfJCyWbgsqp6/uMssw3YBjA/P3/y0tLSSIH23Pcg9zw80k1nYv4IBptvrdm2bNo4vTAHsXfvXubm5ma6zbUw33iGnG/I2aDLt3Xr1l1VtTDuug6dRCCAqtoB7ABYWFioxcXFkdZz3oWXcO51E4s1cdu37BtsvrVm233m4vTCHMTOnTsZ9XkxC+Ybz5DzDTkbdPkmxbNfJKkhlrokNWQ1pzReBPwZcGKSO5K8efqxJEmjWHEAtqreMIsgkqTxOfwiSQ2x1CWpIZa6JDXEUpekhljqktQQS12SGmKpS1JDLHVJaoilLkkNsdQlqSGWuiQ1xFKXpIZY6pLUEEtdkhpiqUtSQyx1SWqIpS5JDbHUJakhlrokNcRSl6SGrPjF02rX5rMvn+n2tm/Zx1n9Nne/+7SZblt6onBPXZIaYqlLUkMsdUlqiKUuSQ2x1CWpIZa6JDXEUpekhljqktQQ33ykdTHrNz7t55ue1Dr31CWpIZa6JDXEUpekhqxqTD3JqcBvAYcA76+qd081lTQljzeWv/wDxybNsXzNyoqlnuQQ4L3ATwF3AF9IcmlV3TjtcFIrJnFgeNQXHV9QnlhWs6f+IuDWqvoaQJIl4LWApS79EJjVmUbT/E9nXLPKNoQX0FTV4y+QvA44tap+ob/8RuDFVfXWA5bbBmzrL54I3DJipmOAe0e87SwMOd+Qs4H5xmW+0Q05G3T5NlTVM8Zd0cTOU6+qHcCOcdeT5OqqWphApKkYcr4hZwPzjct8oxtyNvh+vs2TWNdqzn65Ezhh2eXj+3mSpIFZTal/AXhukmclORw4A7h0urEkSaNYcfilqvYleSvwCbpTGi+oqhummGnsIZwpG3K+IWcD843LfKMbcjaYYL4VD5RKkn54+I5SSWqIpS5JDRlMqSc5NcktSW5NcvYMt3tBkj1Jrl8272lJPpXkK/3vo/v5SfKePuOXk7xw2W3e1C//lSRvmmC+E5JckeTGJDckedtQMiZ5SpLPJ7m2z/aufv6zklzVZ/hwf4CdJE/uL9/aX7952brO6effkuSnx812QM5DknwpyWVDy5dkd5LrklyT5Op+3ro/tsvWe1SSjyW5OclNSV46lHxJTuzvt/0/307y9gHl+6f938X1SS7q/16m/9yrqnX/oTsA+1Xg2cDhwLXA82a07ZcBLwSuXzbvPwJn99NnA7/eT78a+GMgwEuAq/r5TwO+1v8+up8+ekL5jgVe2E8fCfwf4HlDyNhvY66fPgy4qt/mR4Az+vnvA36pn/7HwPv66TOAD/fTz+sf8ycDz+qfC4dM8DH+FeBDwGX95cHkA3YDxxwwb90f22VZPgj8Qj99OHDUkPIty3kIcDfwzCHkAzYBtwFHLHvOnTWL597E7tQx74CXAp9Ydvkc4JwZbn8zP1jqtwDH9tPHArf00+cDbzhwOeANwPnL5v/AchPOegnd5/AMKiPwVOCLwIvp3rl36IGPLd0ZVC/tpw/tl8uBj/fy5SaQ63jg08DLgcv67Q0p327+/1IfxGMLbKQrpgwx3wGZXgl8bij56Er9droXikP7595Pz+K5N5Thl/13wH539PPWy3xV3dVP3w3M99OPlXMm+ft/yV5At0c8iIz90MY1wB7gU3R7Eg9U1b6DbOf7GfrrHwSePq1svd8E3gn8VX/56QPLV8Ank+xK91EbMJDHlm7P8JvA7/bDV+9PsmFA+ZY7A7ion173fFV1J/CfgL8A7qJ7Lu1iBs+9oZT6YFX38rju530mmQP+AHh7VX17+XXrmbGqHqmqk+j2iF8E/Oh65DiYJKcDe6pq13pneRw/UVUvBF4F/JMkL1t+5To//w6lG5r8r1X1AuAhuuGM7xvC30c/Lv0a4KMHXrde+fpx/NfSvTAeB2wATp3FtodS6kP7KIJ7khwL0P/e089/rJxTzZ/kMLpCv7CqPj7EjFX1AHAF3b+URyXZ/8a25dv5fob++o3At6aY7RTgNUl2A0t0QzC/NaB8+/foqKo9wMV0L4xDeWzvAO6oqqv6yx+jK/mh5NvvVcAXq+qe/vIQ8r0CuK2qvllV3wM+Tvd8nPpzbyilPrSPIrgU2H8E/E1049j75//D/ij6S4AH+3/zPgG8MsnR/Sv0K/t5Y0sS4APATVX1G0PKmOQZSY7qp4+gG+u/ia7cX/cY2fZnfh3wmX5P6lLgjP4MgGcBzwU+P042gKo6p6qOr+6Dks7ot3fmUPIl2ZDkyP3TdI/J9QzgsQWoqruB25Oc2M/6B3QfuT2IfMu8gUeHXvbnWO98fwG8JMlT+7/h/ffd9J97kzxYMeaBhVfTndnxVeBXZ7jdi+jGvL5Ht2fyZrqxrE8DXwH+J/C0ftnQfWHIV4HrgIVl6/l54Nb+5+cmmO8n6P59/DJwTf/z6iFkBH4M+FKf7XrgX/fzn90/8W6l+5f4yf38p/SXb+2vf/aydf1qn/kW4FVTeJwXefTsl0Hk63Nc2//csP95P4THdtl6TwKu7h/jP6Q7O2RI+TbQ7dFuXDZvEPmAdwE3938b/53uDJapP/f8mABJashQhl8kSRNgqUtSQyx1SWqIpS5JDbHUJakhlrokNcRSl6SG/D/c0+Vp9WOFgwAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "dataMl.hist(column='neighbors')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Feature Engineering "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Giving the descriptive analytics from above, I will consider the next situation. I do have good variability in my data on the feature \"neighbors\", \"combined_score\" and \"textmining\". I decided to extract the IQR from these features and perform a substraction operation to locate how far the register is from the IQR and use it as an extra feature to detect whether the register belongs to cancer positive label."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [],
   "source": [
    "Q1_neigh = dataMl['neighbors'].quantile(0.25)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [],
   "source": [
    "Q3_neigh = dataMl['neighbors'].quantile(0.75)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {},
   "outputs": [],
   "source": [
    "IQR_neigh = Q3_neigh - Q1_neigh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "859.0"
      ]
     },
     "execution_count": 116,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "IQR_neigh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {},
   "outputs": [],
   "source": [
    "neighbors_dev = Q3_neigh = dataMl['neighbors'] - IQR_neigh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {},
   "outputs": [],
   "source": [
    "Q1_comb = dataMl['combined_score'].quantile(0.25)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {},
   "outputs": [],
   "source": [
    "Q3_comb = dataMl['combined_score'].quantile(0.75)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {},
   "outputs": [],
   "source": [
    "IQR_comb = Q3_comb - Q1_comb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "140.0"
      ]
     },
     "execution_count": 121,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "IQR_comb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_score_dev = dataMl['combined_score'] - IQR_comb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {},
   "outputs": [],
   "source": [
    "Q1_txt = dataMl['textmining'].quantile(0.25)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {},
   "outputs": [],
   "source": [
    "Q3_txt = dataMl['textmining'].quantile(0.75)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {},
   "outputs": [],
   "source": [
    "IQR_txt = Q3_txt - Q1_txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "174.0"
      ]
     },
     "execution_count": 126,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "IQR_txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [],
   "source": [
    "textmining_dev = dataMl['textmining'] - IQR_comb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {},
   "outputs": [],
   "source": [
    "predictor = dataMl['cancer']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "metadata": {},
   "outputs": [],
   "source": [
    "del dataMl['cancer']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "metadata": {},
   "outputs": [],
   "source": [
    "dataMl['neighbors_dev'] = neighbors_dev"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "metadata": {},
   "outputs": [],
   "source": [
    "dataMl['combined_score_dev'] = combined_score_dev"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "metadata": {},
   "outputs": [],
   "source": [
    "dataMl['textmining_dev'] = textmining_dev"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "metadata": {},
   "outputs": [],
   "source": [
    "dataMl['cancer'] = predictor"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Apply Label encoding and normalization"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Given that I do have around 20000 different kind of protein names, I decided to apply a label encoder instead of one hot encoder, due to my computational issues, it will make my data to grow a lot. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "metadata": {},
   "outputs": [],
   "source": [
    "label_prot1 = preprocessing.LabelEncoder() \n",
    "dataMl['protein1']= label_prot1.fit_transform(dataMl['protein1']) \n",
    "#print(block['Gender'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "metadata": {},
   "outputs": [],
   "source": [
    "label_prot2 = preprocessing.LabelEncoder() \n",
    "dataMl['protein2'] = label_prot2.fit_transform(dataMl['protein2'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-152-594595a18b85>:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  encoder_columns['index_'] = encoder_columns.index\n"
     ]
    }
   ],
   "source": [
    "encoder_columns = dataMl[['protein1','protein2']]\n",
    "encoder_columns['index_'] = encoder_columns.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "metadata": {},
   "outputs": [],
   "source": [
    "number_columns = dataMl.iloc[:,2:-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "metadata": {},
   "outputs": [],
   "source": [
    "scaler = MinMaxScaler()\n",
    "numerical_data = pd.DataFrame(scaler.fit_transform(number_columns),columns=number_columns.columns, index=number_columns.index) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "metadata": {},
   "outputs": [],
   "source": [
    "numerical_data['index_'] = numerical_data.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 161,
   "metadata": {},
   "outputs": [],
   "source": [
    "dataMl_prep = encoder_columns.join(numerical_data.set_index('index_'), on='index_')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 165,
   "metadata": {},
   "outputs": [],
   "source": [
    "del dataMl_prep['index_']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 166,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein1</th>\n",
       "      <th>protein2</th>\n",
       "      <th>neighborhood</th>\n",
       "      <th>fusion</th>\n",
       "      <th>cooccurence</th>\n",
       "      <th>coexpression</th>\n",
       "      <th>experimental</th>\n",
       "      <th>database</th>\n",
       "      <th>textmining</th>\n",
       "      <th>combined_score</th>\n",
       "      <th>neighbors</th>\n",
       "      <th>neighbors_dev</th>\n",
       "      <th>combined_score_dev</th>\n",
       "      <th>textmining_dev</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>3469</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.741071</td>\n",
       "      <td>0.062062</td>\n",
       "      <td>0.181181</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.125881</td>\n",
       "      <td>0.400471</td>\n",
       "      <td>0.217295</td>\n",
       "      <td>0.217295</td>\n",
       "      <td>0.400471</td>\n",
       "      <td>0.125881</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>1759</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.186186</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.056395</td>\n",
       "      <td>0.056537</td>\n",
       "      <td>0.217295</td>\n",
       "      <td>0.217295</td>\n",
       "      <td>0.056537</td>\n",
       "      <td>0.056395</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>15599</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.159159</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.010601</td>\n",
       "      <td>0.217295</td>\n",
       "      <td>0.217295</td>\n",
       "      <td>0.010601</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>16415</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.061061</td>\n",
       "      <td>0.158158</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.545821</td>\n",
       "      <td>0.537102</td>\n",
       "      <td>0.217295</td>\n",
       "      <td>0.217295</td>\n",
       "      <td>0.537102</td>\n",
       "      <td>0.545821</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "      <td>7401</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.088088</td>\n",
       "      <td>0.078078</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.089627</td>\n",
       "      <td>0.020024</td>\n",
       "      <td>0.217295</td>\n",
       "      <td>0.217295</td>\n",
       "      <td>0.020024</td>\n",
       "      <td>0.089627</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11759449</th>\n",
       "      <td>19353</td>\n",
       "      <td>6045</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.167167</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.020024</td>\n",
       "      <td>0.012951</td>\n",
       "      <td>0.012951</td>\n",
       "      <td>0.020024</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11759450</th>\n",
       "      <td>19353</td>\n",
       "      <td>8702</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.175175</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.029446</td>\n",
       "      <td>0.012951</td>\n",
       "      <td>0.012951</td>\n",
       "      <td>0.029446</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11759451</th>\n",
       "      <td>19353</td>\n",
       "      <td>9464</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.195195</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.053004</td>\n",
       "      <td>0.012951</td>\n",
       "      <td>0.012951</td>\n",
       "      <td>0.053004</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11759452</th>\n",
       "      <td>19353</td>\n",
       "      <td>12477</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.883392</td>\n",
       "      <td>0.012951</td>\n",
       "      <td>0.012951</td>\n",
       "      <td>0.883392</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11759453</th>\n",
       "      <td>19353</td>\n",
       "      <td>9429</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.213213</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.074205</td>\n",
       "      <td>0.012951</td>\n",
       "      <td>0.012951</td>\n",
       "      <td>0.074205</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>11759454 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          protein1  protein2  neighborhood  fusion  cooccurence  coexpression  \\\n",
       "0                0      3469           0.0     0.0     0.741071      0.062062   \n",
       "1                0      1759           0.0     0.0     0.000000      0.000000   \n",
       "2                0     15599           0.0     0.0     0.000000      0.000000   \n",
       "3                0     16415           0.0     0.0     0.000000      0.061061   \n",
       "4                0      7401           0.0     0.0     0.000000      0.088088   \n",
       "...            ...       ...           ...     ...          ...           ...   \n",
       "11759449     19353      6045           0.0     0.0     0.000000      0.167167   \n",
       "11759450     19353      8702           0.0     0.0     0.000000      0.175175   \n",
       "11759451     19353      9464           0.0     0.0     0.000000      0.195195   \n",
       "11759452     19353     12477           0.0     0.0     0.000000      0.000000   \n",
       "11759453     19353      9429           0.0     0.0     0.000000      0.213213   \n",
       "\n",
       "          experimental  database  textmining  combined_score  neighbors  \\\n",
       "0             0.181181       0.0    0.125881        0.400471   0.217295   \n",
       "1             0.186186       0.0    0.056395        0.056537   0.217295   \n",
       "2             0.159159       0.0    0.000000        0.010601   0.217295   \n",
       "3             0.158158       0.0    0.545821        0.537102   0.217295   \n",
       "4             0.078078       0.0    0.089627        0.020024   0.217295   \n",
       "...                ...       ...         ...             ...        ...   \n",
       "11759449      0.000000       0.0    0.000000        0.020024   0.012951   \n",
       "11759450      0.000000       0.0    0.000000        0.029446   0.012951   \n",
       "11759451      0.000000       0.0    0.000000        0.053004   0.012951   \n",
       "11759452      0.000000       1.0    0.000000        0.883392   0.012951   \n",
       "11759453      0.000000       0.0    0.000000        0.074205   0.012951   \n",
       "\n",
       "          neighbors_dev  combined_score_dev  textmining_dev  \n",
       "0              0.217295            0.400471        0.125881  \n",
       "1              0.217295            0.056537        0.056395  \n",
       "2              0.217295            0.010601        0.000000  \n",
       "3              0.217295            0.537102        0.545821  \n",
       "4              0.217295            0.020024        0.089627  \n",
       "...                 ...                 ...             ...  \n",
       "11759449       0.012951            0.020024        0.000000  \n",
       "11759450       0.012951            0.029446        0.000000  \n",
       "11759451       0.012951            0.053004        0.000000  \n",
       "11759452       0.012951            0.883392        0.000000  \n",
       "11759453       0.012951            0.074205        0.000000  \n",
       "\n",
       "[11759454 rows x 14 columns]"
      ]
     },
     "execution_count": 166,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataMl_prep"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 168,
   "metadata": {},
   "outputs": [],
   "source": [
    "dataMl_prep['cancer'] = predictor"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 171,
   "metadata": {},
   "outputs": [],
   "source": [
    "#dataMl_prep.to_csv('cancer_data_prep.csv') #save dataset to use it the next time. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Dealing with Imbalanced Data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In order to make sure that the predictions are made properly, it is needed to know how our data looks. For example, if we have too few cancer properties, we will need to create a homogeneous dataset in order to test the algorithm. Otherwise, it might be biased due to the skewed data. The same case occurs in the anomaly detection for credit cards fraud, cases where the anomalies turn out to be around 5% or 2% of the total amount of data. \n",
    "Therefore, from the Machine Learning Dataset we want to identify how many positive cancer proteins do we have. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "After filtering the cancer proteins, I found out that only 129 are cancer out of 11474579. That makes our data highly imbalanced, only 1.6327288664932914e-05% of the proteins represent cancer. There will be a need to use SMOTE (Synthetic Minority Over-sampling Technique) to give homogeneity to our dataset. According to [Imbalanced Learning](https://imbalanced-learn.org/en/stable/auto_examples/combine/plot_comparison_combine.html#sphx-glr-auto-examples-combine-plot-comparison-combine-py) we will need to apply the oversampling method to increase our skewed cancer data and an undersampling method to clean the noise generated on data. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "dataMl_prep = pd.read_csv('cancer_data_prep.csv', sep=',', header='infer')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "del dataMl_prep['Unnamed: 0']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "x_res = dataMl_prep.iloc[:,:-1]\n",
    "y_res = dataMl_prep.iloc[:,-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Perform resample minority\n",
    "#Increase the skewed cancer data to 20%.\n",
    "smote = SMOTE(sampling_strategy = 0.2)\n",
    "x_smt, y_smt = smote.fit_resample(x_res,y_res)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Data Exploration: Variable Visualization"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here I would like to demonstrate the visualization of the feature values against the proteins with a heatmap."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "dfVisualization = pd.DataFrame(dataMl.iloc[:,0])\n",
    "dfNormalization = pd.DataFrame(StandardScaler().fit_transform(dataMl.iloc[:,1:].values), columns = ['neighborhood','neighborhoodTransferred','fusion','coocurence','homology','coexpression','coexpressionTransferred','experiments','experimentsTransferred','delete','database','databaseTransferred','combinedScores','neighbors','inDegree','outDegree'\n",
    "])\n",
    "dfVisual = pd.concat([dfVisualization,dfNormalization], axis=1, sort=False).dropna()\n",
    "del dfVisual[\"delete\"]\n",
    "#net = Network(clustergrammer_widget)\n",
    "vis = dfVisual.iloc[:,1:].values\n",
    "cols = np.array(dfVisual['id'])\n",
    "ind = ['neighborhood','neighborhoodTransferred','fusion','coocurence','homology','coexpression','coexpressionTransferred','experiments','experimentsTransferred','database','databaseTransferred','combinedScores','neighbors','inDegree','outDegree'\n",
    "]\n",
    "vs = pd.DataFrame(vis.T, index = ind, columns = cols)\n",
    "vs.T.to_csv('vs_explore.csv')\n",
    "plt.pcolor(vs)\n",
    "plt.yticks(np.arange(0.5, len(vs.index), 1), vs.index)\n",
    "plt.xticks(np.arange(0.5, len(vs.columns), 1), vs.columns, rotation=90)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The visualization above displays the 7000 proteins with its corresponding values. The reader cannot really infer from the plot, how the cancer looks. Thus, below I sliced information in chunks of 25 proteins to be able to compare cancer against regular proteins. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dfVisualization_z = pd.DataFrame(np.array(dataMl.iloc[475:500,0]))\n",
    "dfNormalization_z = pd.DataFrame(StandardScaler().fit_transform(dataMl.iloc[475:500,1:].values), columns = ['neighborhood','neighborhoodTransferred','fusion','coocurence','homology','coexpression','coexpressionTransferred','experiments','experimentsTransferred','delete','database','databaseTransferred','combinedScores','neighbors','inDegree','outDegree'\n",
    "])\n",
    "dfVisual_z = pd.concat([dfVisualization_z,dfNormalization_z], axis=1, sort=False).dropna()\n",
    "del dfVisual_z[\"delete\"]\n",
    "#net = Network(clustergrammer_widget)\n",
    "vis_z = dfVisual_z.iloc[:,1:].values\n",
    "cols = np.array(dfVisual_z[0])\n",
    "ind = ['neighborhood','neighborhoodTransferred','fusion','coocurence','homology','coexpression','coexpressionTransferred','experiments','experimentsTransferred','database','databaseTransferred','combinedScores','neighbors','inDegree','outDegree'\n",
    "]\n",
    "vs_z = pd.DataFrame(vis_z.T, index = ind, columns = cols)\n",
    "plt.pcolor(vs_z)\n",
    "plt.yticks(np.arange(0.5, len(vs_z.index), 1), vs_z.index)\n",
    "plt.xticks(np.arange(0.5, len(vs_z.columns), 1), vs_z.columns, rotation=90)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The first chunk of 25 proteins shows clearly the influence of the outDegree feature to detect cancer. In this case protein \"9606.ENSP00000276480\" was detected as cancer. However, one slice of information does not confirm the visual exploration. Due to this reason, I plotted another two chunks for verification. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dfVisualization_m = pd.DataFrame(np.array(dataMl.iloc[575:600,0]))\n",
    "dfNormalization_m = pd.DataFrame(StandardScaler().fit_transform(dataMl.iloc[575:600,1:].values), columns = ['neighborhood','neighborhoodTransferred','fusion','coocurence','homology','coexpression','coexpressionTransferred','experiments','experimentsTransferred','delete','database','databaseTransferred','combinedScores','neighbors','inDegree','outDegree'\n",
    "])\n",
    "dfVisual_m = pd.concat([dfVisualization_m,dfNormalization_m], axis=1, sort=False).dropna()\n",
    "del dfVisual_m[\"delete\"]\n",
    "#net = Network(clustergrammer_widget)\n",
    "vis_m = dfVisual_m.iloc[:,1:].values\n",
    "cols = np.array(dfVisual_m[0])\n",
    "ind = ['neighborhood','neighborhoodTransferred','fusion','coocurence','homology','coexpression','coexpressionTransferred','experiments','experimentsTransferred','database','databaseTransferred','combinedScores','neighbors','inDegree','outDegree'\n",
    "]\n",
    "vs_m = pd.DataFrame(vis_m.T, index = ind, columns = cols)\n",
    "plt.pcolor(vs_m)\n",
    "plt.yticks(np.arange(0.5, len(vs_m.index), 1), vs_m.index)\n",
    "plt.xticks(np.arange(0.5, len(vs_m.columns), 1), vs_m.columns, rotation=90)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The second chunk of 25 proteins confirms the influence of the outDegree feature to detect cancer. In this case protein \"9606.ENSP00000263083\" was detected as cancer. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "dfVisualization_m = pd.DataFrame(np.array(dataMl.iloc[751:775,0]))\n",
    "dfNormalization_m = pd.DataFrame(StandardScaler().fit_transform(dataMl.iloc[751:775,1:].values), columns = ['neighborhood','neighborhoodTransferred','fusion','coocurence','homology','coexpression','coexpressionTransferred','experiments','experimentsTransferred','delete','database','databaseTransferred','combinedScores','neighbors','inDegree','outDegree'\n",
    "])\n",
    "dfVisual_m = pd.concat([dfVisualization_m,dfNormalization_m], axis=1, sort=False).dropna()\n",
    "del dfVisual_m[\"delete\"]\n",
    "#net = Network(clustergrammer_widget)\n",
    "vis_m = dfVisual_m.iloc[:,1:].values\n",
    "cols = np.array(dfVisual_m[0])\n",
    "ind = ['neighborhood','neighborhoodTransferred','fusion','coocurence','homology','coexpression','coexpressionTransferred','experiments','experimentsTransferred','database','databaseTransferred','combinedScores','neighbors','inDegree','outDegree'\n",
    "]\n",
    "vs_m = pd.DataFrame(vis_m.T, index = ind, columns = cols)\n",
    "plt.pcolor(vs_m)\n",
    "plt.yticks(np.arange(0.5, len(vs_m.index), 1), vs_m.index)\n",
    "plt.xticks(np.arange(0.5, len(vs_m.columns), 1), vs_m.columns, rotation=90)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The third chunk of 25 proteins again confirms the influence of the outDegree feature to detect cancer. In this case protein \"9606.ENSP00000291906\" was detected as cancer. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Overall from the visualization exploration, it is possible to infer the importance of analyzing from the network approach interaction between proteins. In this case I confirm the importance of the outDegree feature to train the machine learning algorithm. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# initialize network object\n",
    "net = Network(clustergrammer_widget)\n",
    "# load dataframe\n",
    "net.load_df(vs_m)\n",
    "# cluster using default parameters\n",
    "net.cluster(enrichrgram=True)\n",
    "# make the visualization\n",
    "net.widget()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Please check the Tableau Report for a complete detail of the data visualization exploration* "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### PCA Component Decision"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this part I am going to verify how many of the features affecting proteins are useful to detect cancer. Thus, the next step is to apply the principal component analysis to create feature reduction. In order to proceed with the PCA, I will need to know which independent variables are more significant in terms of variance. Thanks to the function \"explained_variance_ratio\", I will check the percentage of variance is represented by each variable. From my personal point of view, I like to reduce the variables up to the point where I am using around 70% of the system variance. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "network_characteristics = x_smt.iloc[:,2:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>neighborhood</th>\n",
       "      <th>fusion</th>\n",
       "      <th>cooccurence</th>\n",
       "      <th>coexpression</th>\n",
       "      <th>experimental</th>\n",
       "      <th>database</th>\n",
       "      <th>textmining</th>\n",
       "      <th>combined_score</th>\n",
       "      <th>neighbors</th>\n",
       "      <th>neighbors_dev</th>\n",
       "      <th>combined_score_dev</th>\n",
       "      <th>textmining_dev</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.741071</td>\n",
       "      <td>0.062062</td>\n",
       "      <td>0.181181</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.125881</td>\n",
       "      <td>0.400471</td>\n",
       "      <td>0.217295</td>\n",
       "      <td>0.217295</td>\n",
       "      <td>0.400471</td>\n",
       "      <td>0.125881</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.186186</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.056395</td>\n",
       "      <td>0.056537</td>\n",
       "      <td>0.217295</td>\n",
       "      <td>0.217295</td>\n",
       "      <td>0.056537</td>\n",
       "      <td>0.056395</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.159159</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.010601</td>\n",
       "      <td>0.217295</td>\n",
       "      <td>0.217295</td>\n",
       "      <td>0.010601</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.061061</td>\n",
       "      <td>0.158158</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.545821</td>\n",
       "      <td>0.537102</td>\n",
       "      <td>0.217295</td>\n",
       "      <td>0.217295</td>\n",
       "      <td>0.537102</td>\n",
       "      <td>0.545821</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.088088</td>\n",
       "      <td>0.078078</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.089627</td>\n",
       "      <td>0.020024</td>\n",
       "      <td>0.217295</td>\n",
       "      <td>0.217295</td>\n",
       "      <td>0.020024</td>\n",
       "      <td>0.089627</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14111109</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.026769</td>\n",
       "      <td>0.179329</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.102868</td>\n",
       "      <td>0.162911</td>\n",
       "      <td>0.230631</td>\n",
       "      <td>0.230631</td>\n",
       "      <td>0.162911</td>\n",
       "      <td>0.102868</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14111110</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.123221</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.103280</td>\n",
       "      <td>0.089110</td>\n",
       "      <td>0.369309</td>\n",
       "      <td>0.369309</td>\n",
       "      <td>0.089110</td>\n",
       "      <td>0.103280</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14111111</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.243215</td>\n",
       "      <td>0.107789</td>\n",
       "      <td>0.036192</td>\n",
       "      <td>0.036192</td>\n",
       "      <td>0.107789</td>\n",
       "      <td>0.243215</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14111112</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.021707</td>\n",
       "      <td>0.094532</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.343593</td>\n",
       "      <td>0.290696</td>\n",
       "      <td>0.369309</td>\n",
       "      <td>0.369309</td>\n",
       "      <td>0.290696</td>\n",
       "      <td>0.343593</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14111113</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.041154</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.156345</td>\n",
       "      <td>0.039437</td>\n",
       "      <td>0.342491</td>\n",
       "      <td>0.342491</td>\n",
       "      <td>0.039437</td>\n",
       "      <td>0.156345</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>14111114 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          neighborhood  fusion  cooccurence  coexpression  experimental  \\\n",
       "0                  0.0     0.0     0.741071      0.062062      0.181181   \n",
       "1                  0.0     0.0     0.000000      0.000000      0.186186   \n",
       "2                  0.0     0.0     0.000000      0.000000      0.159159   \n",
       "3                  0.0     0.0     0.000000      0.061061      0.158158   \n",
       "4                  0.0     0.0     0.000000      0.088088      0.078078   \n",
       "...                ...     ...          ...           ...           ...   \n",
       "14111109           0.0     0.0     0.000000      0.026769      0.179329   \n",
       "14111110           0.0     0.0     0.000000      0.000000      0.123221   \n",
       "14111111           0.0     0.0     0.000000      0.000000      0.000000   \n",
       "14111112           0.0     0.0     0.000000      0.021707      0.094532   \n",
       "14111113           0.0     0.0     0.000000      0.000000      0.041154   \n",
       "\n",
       "          database  textmining  combined_score  neighbors  neighbors_dev  \\\n",
       "0              0.0    0.125881        0.400471   0.217295       0.217295   \n",
       "1              0.0    0.056395        0.056537   0.217295       0.217295   \n",
       "2              0.0    0.000000        0.010601   0.217295       0.217295   \n",
       "3              0.0    0.545821        0.537102   0.217295       0.217295   \n",
       "4              0.0    0.089627        0.020024   0.217295       0.217295   \n",
       "...            ...         ...             ...        ...            ...   \n",
       "14111109       0.0    0.102868        0.162911   0.230631       0.230631   \n",
       "14111110       0.0    0.103280        0.089110   0.369309       0.369309   \n",
       "14111111       0.0    0.243215        0.107789   0.036192       0.036192   \n",
       "14111112       0.0    0.343593        0.290696   0.369309       0.369309   \n",
       "14111113       0.0    0.156345        0.039437   0.342491       0.342491   \n",
       "\n",
       "          combined_score_dev  textmining_dev  \n",
       "0                   0.400471        0.125881  \n",
       "1                   0.056537        0.056395  \n",
       "2                   0.010601        0.000000  \n",
       "3                   0.537102        0.545821  \n",
       "4                   0.020024        0.089627  \n",
       "...                      ...             ...  \n",
       "14111109            0.162911        0.102868  \n",
       "14111110            0.089110        0.103280  \n",
       "14111111            0.107789        0.243215  \n",
       "14111112            0.290696        0.343593  \n",
       "14111113            0.039437        0.156345  \n",
       "\n",
       "[14111114 rows x 12 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "network_characteristics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([5.86936960e-01, 1.91123027e-01, 9.24260645e-02, 6.01359350e-02,\n",
       "       3.84248186e-02, 1.50993525e-02, 1.14172542e-02, 4.19167626e-03,\n",
       "       2.44912793e-04, 5.27854014e-31, 3.29672669e-31])"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Amount of components decision: run pca with all the elements in the x_reduced and analyze the percentage of \n",
    "# variance over the dataset with the percentage of variance.\n",
    "pca = PCA(n_components = 11)\n",
    "principal = pca.fit_transform(np.array(network_characteristics))\n",
    "x_org_pca = principal \n",
    "percentageVarianceVector = pca.explained_variance_ratio_\n",
    "percentageVarianceVector"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Exploring the components\n",
    "df_comp = pd.DataFrame(pca.components_)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.000811</td>\n",
       "      <td>0.000243</td>\n",
       "      <td>0.004895</td>\n",
       "      <td>0.039088</td>\n",
       "      <td>0.063123</td>\n",
       "      <td>0.465083</td>\n",
       "      <td>0.198296</td>\n",
       "      <td>0.591339</td>\n",
       "      <td>0.008785</td>\n",
       "      <td>0.008785</td>\n",
       "      <td>0.591339</td>\n",
       "      <td>0.198296</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.000085</td>\n",
       "      <td>-0.000111</td>\n",
       "      <td>0.001277</td>\n",
       "      <td>-0.066877</td>\n",
       "      <td>-0.114334</td>\n",
       "      <td>-0.494870</td>\n",
       "      <td>0.607270</td>\n",
       "      <td>-0.000745</td>\n",
       "      <td>0.001279</td>\n",
       "      <td>0.001279</td>\n",
       "      <td>-0.000745</td>\n",
       "      <td>0.607270</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>-0.000921</td>\n",
       "      <td>-0.000331</td>\n",
       "      <td>0.001807</td>\n",
       "      <td>0.064325</td>\n",
       "      <td>0.244133</td>\n",
       "      <td>-0.116344</td>\n",
       "      <td>-0.022277</td>\n",
       "      <td>0.027982</td>\n",
       "      <td>0.678293</td>\n",
       "      <td>0.678293</td>\n",
       "      <td>0.027982</td>\n",
       "      <td>-0.022277</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.011073</td>\n",
       "      <td>0.000907</td>\n",
       "      <td>0.007562</td>\n",
       "      <td>0.315448</td>\n",
       "      <td>0.707737</td>\n",
       "      <td>-0.482109</td>\n",
       "      <td>-0.111818</td>\n",
       "      <td>0.181739</td>\n",
       "      <td>-0.194842</td>\n",
       "      <td>-0.194842</td>\n",
       "      <td>0.181739</td>\n",
       "      <td>-0.111818</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.056038</td>\n",
       "      <td>0.001517</td>\n",
       "      <td>0.019042</td>\n",
       "      <td>0.740365</td>\n",
       "      <td>-0.574044</td>\n",
       "      <td>-0.235747</td>\n",
       "      <td>-0.109268</td>\n",
       "      <td>0.134822</td>\n",
       "      <td>0.038844</td>\n",
       "      <td>0.038844</td>\n",
       "      <td>0.134822</td>\n",
       "      <td>-0.109268</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         0         1         2         3         4         5         6   \\\n",
       "0  0.000811  0.000243  0.004895  0.039088  0.063123  0.465083  0.198296   \n",
       "1  0.000085 -0.000111  0.001277 -0.066877 -0.114334 -0.494870  0.607270   \n",
       "2 -0.000921 -0.000331  0.001807  0.064325  0.244133 -0.116344 -0.022277   \n",
       "3  0.011073  0.000907  0.007562  0.315448  0.707737 -0.482109 -0.111818   \n",
       "4  0.056038  0.001517  0.019042  0.740365 -0.574044 -0.235747 -0.109268   \n",
       "\n",
       "         7         8         9         10        11  \n",
       "0  0.591339  0.008785  0.008785  0.591339  0.198296  \n",
       "1 -0.000745  0.001279  0.001279 -0.000745  0.607270  \n",
       "2  0.027982  0.678293  0.678293  0.027982 -0.022277  \n",
       "3  0.181739 -0.194842 -0.194842  0.181739 -0.111818  \n",
       "4  0.134822  0.038844  0.038844  0.134822 -0.109268  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_comp.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>neighborhood</th>\n",
       "      <th>fusion</th>\n",
       "      <th>cooccurence</th>\n",
       "      <th>coexpression</th>\n",
       "      <th>experimental</th>\n",
       "      <th>database</th>\n",
       "      <th>textmining</th>\n",
       "      <th>combined_score</th>\n",
       "      <th>neighbors</th>\n",
       "      <th>neighbors_dev</th>\n",
       "      <th>combined_score_dev</th>\n",
       "      <th>textmining_dev</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>-0.009411</td>\n",
       "      <td>-0.000152</td>\n",
       "      <td>0.738561</td>\n",
       "      <td>0.005691</td>\n",
       "      <td>0.112569</td>\n",
       "      <td>-0.053535</td>\n",
       "      <td>-0.055720</td>\n",
       "      <td>0.227182</td>\n",
       "      <td>0.070235</td>\n",
       "      <td>0.070235</td>\n",
       "      <td>0.227182</td>\n",
       "      <td>-0.055720</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>-0.009411</td>\n",
       "      <td>-0.000152</td>\n",
       "      <td>-0.002511</td>\n",
       "      <td>-0.056371</td>\n",
       "      <td>0.117574</td>\n",
       "      <td>-0.053535</td>\n",
       "      <td>-0.125206</td>\n",
       "      <td>-0.116752</td>\n",
       "      <td>0.070235</td>\n",
       "      <td>0.070235</td>\n",
       "      <td>-0.116752</td>\n",
       "      <td>-0.125206</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>-0.009411</td>\n",
       "      <td>-0.000152</td>\n",
       "      <td>-0.002511</td>\n",
       "      <td>-0.056371</td>\n",
       "      <td>0.090547</td>\n",
       "      <td>-0.053535</td>\n",
       "      <td>-0.181601</td>\n",
       "      <td>-0.162689</td>\n",
       "      <td>0.070235</td>\n",
       "      <td>0.070235</td>\n",
       "      <td>-0.162689</td>\n",
       "      <td>-0.181601</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>-0.009411</td>\n",
       "      <td>-0.000152</td>\n",
       "      <td>-0.002511</td>\n",
       "      <td>0.004690</td>\n",
       "      <td>0.089546</td>\n",
       "      <td>-0.053535</td>\n",
       "      <td>0.364220</td>\n",
       "      <td>0.363813</td>\n",
       "      <td>0.070235</td>\n",
       "      <td>0.070235</td>\n",
       "      <td>0.363813</td>\n",
       "      <td>0.364220</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>-0.009411</td>\n",
       "      <td>-0.000152</td>\n",
       "      <td>-0.002511</td>\n",
       "      <td>0.031717</td>\n",
       "      <td>0.009466</td>\n",
       "      <td>-0.053535</td>\n",
       "      <td>-0.091973</td>\n",
       "      <td>-0.153266</td>\n",
       "      <td>0.070235</td>\n",
       "      <td>0.070235</td>\n",
       "      <td>-0.153266</td>\n",
       "      <td>-0.091973</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   neighborhood    fusion  cooccurence  coexpression  experimental  database  \\\n",
       "0     -0.009411 -0.000152     0.738561      0.005691      0.112569 -0.053535   \n",
       "1     -0.009411 -0.000152    -0.002511     -0.056371      0.117574 -0.053535   \n",
       "2     -0.009411 -0.000152    -0.002511     -0.056371      0.090547 -0.053535   \n",
       "3     -0.009411 -0.000152    -0.002511      0.004690      0.089546 -0.053535   \n",
       "4     -0.009411 -0.000152    -0.002511      0.031717      0.009466 -0.053535   \n",
       "\n",
       "   textmining  combined_score  neighbors  neighbors_dev  combined_score_dev  \\\n",
       "0   -0.055720        0.227182   0.070235       0.070235            0.227182   \n",
       "1   -0.125206       -0.116752   0.070235       0.070235           -0.116752   \n",
       "2   -0.181601       -0.162689   0.070235       0.070235           -0.162689   \n",
       "3    0.364220        0.363813   0.070235       0.070235            0.363813   \n",
       "4   -0.091973       -0.153266   0.070235       0.070235           -0.153266   \n",
       "\n",
       "   textmining_dev  \n",
       "0       -0.055720  \n",
       "1       -0.125206  \n",
       "2       -0.181601  \n",
       "3        0.364220  \n",
       "4       -0.091973  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#compute the mean centered data\n",
    "X_mean = network_characteristics - network_characteristics.mean()\n",
    "X_mean.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "-0.20620175244869882"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#compute pca for row 1\n",
    "np.dot(df_comp.loc[0, :], X_mean.loc[1, :])  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[5.8694e-01 1.9112e-01 9.2430e-02 6.0140e-02 3.8420e-02 1.5100e-02\n",
      " 1.1420e-02 4.1900e-03 2.4000e-04 0.0000e+00 0.0000e+00]\n"
     ]
    }
   ],
   "source": [
    "#getting the variance ratio up to 5 decimals\n",
    "print(pca.explained_variance_ratio_.round(5)[:13])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[0.587 0.191 0.092 0.06  0.038 0.015 0.011 0.004 0.    0.    0.   ]\n"
     ]
    }
   ],
   "source": [
    "#getting the variance ratio up to 50 decimals\n",
    "print(pca.explained_variance_ratio_.round(3)[:11])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As I first analyse the dataset x_smt, I noticed that the protein 1 and 2 had about the 99% influence over the variance. Thus, I leave them out. Because, those two attributes describe the relation itself, but not the network behavior. The result of the analysis of the network properties can be appreciate above. I am going to leave out from my ml algorithm the Protein 1 and Protein 2 in order to be able to predict my outcome based on the network behavior. The selected features at the end turn out to be neighborhood, fusion, coocurrence and coexpression. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein1</th>\n",
       "      <th>protein2</th>\n",
       "      <th>neighborhood</th>\n",
       "      <th>fusion</th>\n",
       "      <th>cooccurence</th>\n",
       "      <th>coexpression</th>\n",
       "      <th>experimental</th>\n",
       "      <th>database</th>\n",
       "      <th>textmining</th>\n",
       "      <th>combined_score</th>\n",
       "      <th>neighbors</th>\n",
       "      <th>neighbors_dev</th>\n",
       "      <th>combined_score_dev</th>\n",
       "      <th>textmining_dev</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>3469</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.741071</td>\n",
       "      <td>0.062062</td>\n",
       "      <td>0.181181</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.125881</td>\n",
       "      <td>0.400471</td>\n",
       "      <td>0.217295</td>\n",
       "      <td>0.217295</td>\n",
       "      <td>0.400471</td>\n",
       "      <td>0.125881</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>1759</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.186186</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.056395</td>\n",
       "      <td>0.056537</td>\n",
       "      <td>0.217295</td>\n",
       "      <td>0.217295</td>\n",
       "      <td>0.056537</td>\n",
       "      <td>0.056395</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>15599</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.159159</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.010601</td>\n",
       "      <td>0.217295</td>\n",
       "      <td>0.217295</td>\n",
       "      <td>0.010601</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>16415</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.061061</td>\n",
       "      <td>0.158158</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.545821</td>\n",
       "      <td>0.537102</td>\n",
       "      <td>0.217295</td>\n",
       "      <td>0.217295</td>\n",
       "      <td>0.537102</td>\n",
       "      <td>0.545821</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "      <td>7401</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.088088</td>\n",
       "      <td>0.078078</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.089627</td>\n",
       "      <td>0.020024</td>\n",
       "      <td>0.217295</td>\n",
       "      <td>0.217295</td>\n",
       "      <td>0.020024</td>\n",
       "      <td>0.089627</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14111109</th>\n",
       "      <td>16870</td>\n",
       "      <td>12919</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.026769</td>\n",
       "      <td>0.179329</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.102868</td>\n",
       "      <td>0.162911</td>\n",
       "      <td>0.230631</td>\n",
       "      <td>0.230631</td>\n",
       "      <td>0.162911</td>\n",
       "      <td>0.102868</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14111110</th>\n",
       "      <td>16417</td>\n",
       "      <td>12290</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.123221</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.103280</td>\n",
       "      <td>0.089110</td>\n",
       "      <td>0.369309</td>\n",
       "      <td>0.369309</td>\n",
       "      <td>0.089110</td>\n",
       "      <td>0.103280</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14111111</th>\n",
       "      <td>14375</td>\n",
       "      <td>206</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.243215</td>\n",
       "      <td>0.107789</td>\n",
       "      <td>0.036192</td>\n",
       "      <td>0.036192</td>\n",
       "      <td>0.107789</td>\n",
       "      <td>0.243215</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14111112</th>\n",
       "      <td>16417</td>\n",
       "      <td>9086</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.021707</td>\n",
       "      <td>0.094532</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.343593</td>\n",
       "      <td>0.290696</td>\n",
       "      <td>0.369309</td>\n",
       "      <td>0.369309</td>\n",
       "      <td>0.290696</td>\n",
       "      <td>0.343593</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14111113</th>\n",
       "      <td>2912</td>\n",
       "      <td>12386</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.041154</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.156345</td>\n",
       "      <td>0.039437</td>\n",
       "      <td>0.342491</td>\n",
       "      <td>0.342491</td>\n",
       "      <td>0.039437</td>\n",
       "      <td>0.156345</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>14111114 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          protein1  protein2  neighborhood  fusion  cooccurence  coexpression  \\\n",
       "0                0      3469           0.0     0.0     0.741071      0.062062   \n",
       "1                0      1759           0.0     0.0     0.000000      0.000000   \n",
       "2                0     15599           0.0     0.0     0.000000      0.000000   \n",
       "3                0     16415           0.0     0.0     0.000000      0.061061   \n",
       "4                0      7401           0.0     0.0     0.000000      0.088088   \n",
       "...            ...       ...           ...     ...          ...           ...   \n",
       "14111109     16870     12919           0.0     0.0     0.000000      0.026769   \n",
       "14111110     16417     12290           0.0     0.0     0.000000      0.000000   \n",
       "14111111     14375       206           0.0     0.0     0.000000      0.000000   \n",
       "14111112     16417      9086           0.0     0.0     0.000000      0.021707   \n",
       "14111113      2912     12386           0.0     0.0     0.000000      0.000000   \n",
       "\n",
       "          experimental  database  textmining  combined_score  neighbors  \\\n",
       "0             0.181181       0.0    0.125881        0.400471   0.217295   \n",
       "1             0.186186       0.0    0.056395        0.056537   0.217295   \n",
       "2             0.159159       0.0    0.000000        0.010601   0.217295   \n",
       "3             0.158158       0.0    0.545821        0.537102   0.217295   \n",
       "4             0.078078       0.0    0.089627        0.020024   0.217295   \n",
       "...                ...       ...         ...             ...        ...   \n",
       "14111109      0.179329       0.0    0.102868        0.162911   0.230631   \n",
       "14111110      0.123221       0.0    0.103280        0.089110   0.369309   \n",
       "14111111      0.000000       0.0    0.243215        0.107789   0.036192   \n",
       "14111112      0.094532       0.0    0.343593        0.290696   0.369309   \n",
       "14111113      0.041154       0.0    0.156345        0.039437   0.342491   \n",
       "\n",
       "          neighbors_dev  combined_score_dev  textmining_dev  \n",
       "0              0.217295            0.400471        0.125881  \n",
       "1              0.217295            0.056537        0.056395  \n",
       "2              0.217295            0.010601        0.000000  \n",
       "3              0.217295            0.537102        0.545821  \n",
       "4              0.217295            0.020024        0.089627  \n",
       "...                 ...                 ...             ...  \n",
       "14111109       0.230631            0.162911        0.102868  \n",
       "14111110       0.369309            0.089110        0.103280  \n",
       "14111111       0.036192            0.107789        0.243215  \n",
       "14111112       0.369309            0.290696        0.343593  \n",
       "14111113       0.342491            0.039437        0.156345  \n",
       "\n",
       "[14111114 rows x 14 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x_smt "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "x_smt_pca = network_characteristics.iloc[:,:3]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For this particular case I am going to work with the features up to the neighbors."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Use it to check whether the matrix was working properly.\n",
    "#np.linalg.svd(np.array(x_reduced))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Machine Learning Algorithm (Supervised Learning) : SVM for the binary classification"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I selected the support vector machine algorithm (SVM) to perform the binary classification over the data set. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "After applying resampling and spliting the dataset into training set and test set, I will apply the principal component analysis to reduce the independent variables to 4."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "k = 5\n",
    "kf = KFold(n_splits=k, random_state=None)\n",
    "clf = svm.SVC(kernel = 'linear', C=0.01)\n",
    " \n",
    "acc_score = []\n",
    " \n",
    "for train_index , test_index in kf.split(x_smt_pca):\n",
    "    X_train , X_test = x_smt_pca.iloc[train_index,:],x_smt_pca.iloc[test_index,:]\n",
    "    y_train , y_test = y_smt[train_index] , y_smt[test_index]\n",
    "    \n",
    "\n",
    "    clf.fit(X_train,y_train)\n",
    "    pred_values = clf.predict(X_test)\n",
    "     \n",
    "    acc = accuracy_score(pred_values , y_test)\n",
    "    acc_score.append(acc)\n",
    "     \n",
    "avg_acc_score = sum(acc_score)/k\n",
    " \n",
    "print('accuracy of each fold - {}'.format(acc_score))\n",
    "print('Avg accuracy : {}'.format(avg_acc_score))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
